---
title: "R Notebook"
output:
  html_document:
    df_print: paged
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code.

Try executing this chunk by clicking the Run button within the chunk or by placing your cursor inside it and pressing Cmd+Shift+Enter.

```{r}
gc() 
rm(list = ls()) #Remove variables from previous runs
```

```{r}
#install devtools package
# if (!requireNamespace("devtools", quietly = TRUE)) {
# install.packages("devtools")
# }
```

```{r}
# update the cdr3tools package
# devtools::install_github("caparks2/cdr3tools", force = TRUE)
# devtools::install_github("tidyverse/tidyverse", force = TRUE)
```

```{r}
# install.packages("remotes")
# remotes::install_github("agentlans/KneeArrower")
# install.packages("akmedoids")
# install.packages("vistime")
# install.packages("ggseqlogo")
#install.packages("docstring")
```

```{r}
# install.packages("lvplot")
#install.packages('ggbeeswarm')

```

```{r}
# attach package libraries
#library(akmedoids)
library(cdr3tools)
library(tidyverse)
library(eulerr)
library(pheatmap)
library(factoextra)
library(KneeArrower)
library(ggplot2)
library(gridExtra)
#library(vistime)
library(lvplot)
#library(ggseqlogo)
library(MGLM) 
library(readxl)
# library(Biostrings)
library(ggpubr)
library(purrr)
library(dplyr)
library(patchwork)
library(ggbeeswarm)
library(docstring)
library(ggalluvial)
library(plotly)
library(rstatix)
library(tidymodels)
library(cowplot)
library(ggtext)
```

```{r}
# set path to file
path ="/Users/lingtingshi/Documents/GVHD_project/Annotated_v4/" 
setwd(path)
filenames <-list.files(path) #Get list of all folders of patient data
print(filenames)


```

```{r}

combine_immuno <- function(immdata, sample_meta_data, comb_idx, unique_identifiers){ 
  #' Combine immunoseq files from teh same timepoint
  #' 
  #' @description This function combines n immunoseq 
  #' files from the same timepoint that are given from Adaptive Imunno Sequencing.
  #' The file names must be in format PatientID_timepoint_celltype
  #' 
  #' @param immdata List of dataframes containing the data from immunoseq files.
  #' Dataframes are obtained through the function read_immunoseq
  #' @param sample_meta_data 3 by sample dataframe that contains PatientID in first row,
  #' timepoint in second row, and cell type in the third row
  #' @param comb_idx the indices of the samples you would like to combine
  #' @param unique_identifiers Columns in immdata that identify unique clones (eg. cdr3_amino_acid, v_resolved, j_resolved)
  #' @usage combine_immuno(immdata, sample_meta_data, comb_idx, unique_identifiers)
  #' @return updated: list(immdata, sample_meta_data)
  
  
  
  cd4 <- immdata[[comb_idx[1]]] #initialize dataframe to be combined into
  cd4$cell_type<- sample_meta_data[3,comb_idx[1]] #set the cell_type of that dataframe
  for(i in seq(2,length(comb_idx),1)){ #for every other datafram identify by the indices combine into the first dataframe
    cd <- immdata[[comb_idx[i]]]
    cd$cell_type <- sample_meta_data[3, comb_idx[i]]
    cd4 <- dplyr::full_join(cd4,cd, by = c(unique_identifiers, "PatientID", "timepoint"))%>%
      replace_na(list(frequency.x = 0, frequency.y = 0,templates.x = 0, templates.y = 0, cell_type.x = "", cell_type.y = ""))
    cd4$templates <- cd4$templates.x+ cd4$templates.y #Add the templates of the same clones together
    cd4$frequency <- cd4$templates/sum(cd4$templates) #Recalculate the frequency
    cd4$cell_type <- gsub(" ", "",paste(gsub(" ", "",cd4$cell_type.x), gsub(" ", "", cd4$cell_type.y)))
    cd4 <- cd4 %>% select(all_of(c(unique_identifiers, "frequency", "templates", "PatientID", "timepoint", "cell_type")))
    immdata[[comb_idx[1]]] <- cd4
  }
  sample_meta_data[3,comb_idx[1]] <- "both" #change the first dataframes description to be both
  comb_idx <- comb_idx[-1] #remove the first indices from the indices
  immdata <- immdata[-comb_idx] #remove the dataframes that were combined from immdata
  sample_meta_data <- sample_meta_data[,-comb_idx] #remove the samples that were combined from sample_meta_data
  
  return(list(immdata, sample_meta_data))
  
}
```

```{r}
determine_cell_type <- function(cell_type.unstim, cell_type.CFSElo){
  if(cell_type.unstim == "CD4CD8" & cell_type.CFSElo != "CD4CD8"){
    return (cell_type.CFSElo)
  } else if(cell_type.CFSElo == "CD4CD8" & cell_type.unstim != "CD4CD8"){
    return (cell_type.unstim)
  } else if(cell_type.unstim != "CD4CD8" & cell_type.unstim != "CD4CD8"){
    if(cell_type.CFSElo == cell_type.unstim)
    {
      return(cell_type.CFSElo)
    } 
    else
    {
      return("CD4CD8")
    }
  }
  
  return("CD4CD8")
}

```

```{r}

process_data <- function(filename, path, unique_identifiers = c("cdr3_amino_acid", "v_resolved", "j_resolved")){
  #' Process patient immunoseq folder
  #' 
  #' @description This function takes a patients immunoseq folder and iterates
  #' through each file extracting the needed information and combining files that 
  #' are from the same timepoint. Once files are combined alloreactive clones
  #' are determined and returned.
  #' 
  #' @param filename name of patient folder
  #' @param path to folder that contains all patients
  #' @param unique_identifiers Columns in immdata that identify unique clones (eg. cdr3_amino_acid, v_resolved, j_resolved)
  #' @return list(immdata2, sample_meta_data, allo, nonallo)
  #' 

  # load data
  print(filename)
  
  immdata <- read_immunoseq(paste(path, filename,"/",sep = ""))
  immdata2 <- lapply(immdata, function(x) {
  x %>%
    dplyr::group_by(across(all_of(unique_identifiers))) %>% #Combine any duplicates and drop any unnecessary rows 
    dplyr::summarise(templates = sum(templates), .groups = "drop") %>%
    mutate(frequency = templates / sum(templates)) %>%
    dplyr::arrange(dplyr::desc(frequency))  
    })
  

  immdata1 <- map2( #Map PatientID to clones
    immdata2,
    data.frame(strsplit(names(immdata), split = "_"))[1,],
    ~ mutate(.x, PatientID = .y)
  )
  immdata2 <- map2( #map time to each sample
    immdata1,
    data.frame(strsplit(names(immdata), split = "_"))[2,],
    ~ mutate(.x, timepoint = .y)
  )
  
  # get metadata
  sample_names = rownames(summary(immdata))
  sample_meta_data <-data.frame(strsplit(sample_names, split = "_"))
  sample_meta_data
  
  
  Treg_idx <- grep('Treg', sample_meta_data[3,]) #Combine Treg
  if(length(Treg_idx) > 1){
    result <- combine_immuno(immdata2, sample_meta_data, Treg_idx, unique_identifiers)
    immdata2 <- result[[1]]
    sample_meta_data <- result[[2]]
  }
  
  MAX_ITER <-50 #may need to change if you are combining over 50 samples
  iter <- 0
  j <- 1
  while(j <= length(sample_meta_data[2,]) & iter < MAX_ITER){ #Combine and CD4 and CD8 that have the same timepoints

    time_idx <- which(sample_meta_data[2,] == sample_meta_data[2,j])
    if (length(time_idx) > 1){ #If there is one CD4 or CD8 Change to both
      result <- combine_immuno(immdata2, sample_meta_data, time_idx, unique_identifiers)
      immdata2 <- result[[1]]
      sample_meta_data <- result[[2]]
      #Ensure Cell type labeling is good
      # immdata2[[j]] <- immdata2[[j]] %>%
      #   mutate(cell_type = ifelse(grepl("CD3", cell_type), gsub("CD3 & CD8")))
      
    }else{
      if (sample_meta_data[3,j] != "both") {
        cd48 <- immdata2[[j]]
        cd48 <- cd48 %>%
          subset(select = c(unique_identifiers, "frequency", "templates", "PatientID", "timepoint"))
        cd48$cell_type <- sample_meta_data[3,j]
        immdata2[[j]] <- cd48
        sample_meta_data[3,j] = "both"
      }else{
        #print("Already Both")
        immdata2[[j]]$cell_type = "CD4CD8"
      }
    }
    iter <- iter + 1
    j <- j + 1
  }
  
  unstim_idx <- which(sample_meta_data[2,] == 'unstimulated')
  MLRCF_idx <- which(sample_meta_data[2,] == 'MLRCFSElo')
  
  allo_inner <- dplyr::inner_join(immdata2[[MLRCF_idx]],immdata2[[unstim_idx]],
                           by = c(unique_identifiers, "PatientID"), 
                           suffix = c(".CFSELo", ".unstim")
  ) %>%
    replace_na(list(frequency.unstim = 0, frequency.CFSELo = 0)) %>%
    filter(frequency.CFSELo >= 2 * frequency.unstim)%>%
    select(unique_identifiers, "frequency.CFSELo", "frequency.unstim", "templates.CFSELo", "templates.unstim", "PatientID", "cell_type.CFSELo")
  allo_inner['tag'] <- 'inner'
  colnames(allo_inner)[colnames(allo_inner) == "cell_type.CFSELo"] <- "cell_type"
  
  allo_outer <- dplyr::anti_join(immdata2[[MLRCF_idx]],immdata2[[unstim_idx]],
                          by = c(unique_identifiers), 
  ) %>%
    replace_na(list(frequency = 0)) %>%
    filter(frequency >= 1e-5)%>%
    select(unique_identifiers, "frequency", "templates", "PatientID", "cell_type")
  names(allo_outer)[names(allo_outer) == 'frequency'] <- 'frequency.CFSELo'
  names(allo_outer)[names(allo_outer) == 'templates'] <- 'templates.CFSELo'
  allo_outer['frequency.unstim'] <- 0
  allo_outer['templates.unstim'] <- 0
  allo_outer['tag'] <- 'outer'
  allo <- bind_rows(allo_inner, allo_outer)
  
  nonallo_unstim <- dplyr::anti_join(immdata2[[unstim_idx]], allo, by = c(unique_identifiers))%>%
    replace_na(list(frequency = 0))%>%
    select(unique_identifiers, "frequency", "templates", "PatientID", "cell_type")

  nonallo_mlr <- dplyr::anti_join(immdata2[[MLRCF_idx]], allo, by = c(unique_identifiers))%>%
    replace_na(list(frequency = 0))%>%
    select(unique_identifiers, "frequency", "templates", "PatientID", "cell_type")

  nonallo <- dplyr::full_join(nonallo_unstim, nonallo_mlr, 
                              by = c(unique_identifiers, "PatientID"),
                              suffix = c(".unstim", ".CFSElo"))%>%
    replace_na(list(frequency.unstim = 0, frequency.CFSElo = 0, 
                    templates.unstim = 0, templates.CFSElo = 0))%>%
    mutate(cell_type.unstim = gsub("CD3", "", cell_type.unstim), cell_type.CFSElo = gsub("CD3", "", cell_type.CFSElo))%>%
    mutate(cell_type.unstim = ifelse(is.na(cell_type.unstim), cell_type.CFSElo, cell_type.unstim), cell_type.CFSElo = ifelse(is.na(cell_type.CFSElo), cell_type.unstim, cell_type.CFSElo))%>%
    mutate(cell_type.CFSElo = gsub("CD3", "", cell_type.CFSElo ), cell_type.unstim = gsub("CD3","", cell_type.unstim))%>%
    mutate(cell_type.CFSElo = gsub("both", "CD4CD8", cell_type.CFSElo ), cell_type.unstim = gsub("both","CD4CD8", cell_type.unstim))%>%
    mutate(cell_type.CFSElo = gsub("tsv", "CD4CD8", cell_type.CFSElo ), cell_type.unstim = gsub("tsv","CD4CD8", cell_type.unstim))%>%
    mutate(cell_type.CFSElo = ifelse(cell_type.CFSElo == "", "CD4CD8", cell_type.CFSElo), cell_type.unstim = ifelse(cell_type.unstim == "", "CD4CD8", cell_type.unstim))%>%
    rowwise()%>%
    mutate(cell_type = determine_cell_type(cell_type.CFSElo, cell_type.unstim))
  # %>%
  #   mutate(cell_type = paste(ifelse((cell_type.unstim), "", cell_type.unstim), ifelse(is.na(cell_type.CFSElo), "", cell_type.CFSElo)))%>%
  #   mutate(cell_type = paste(ifelse(grepl("8",cell_type), "CD8", ""),ifelse(grepl("4",cell_type), "CD4", ""), ifelse(grepl("3",cell_type), "CD3", "") ))#%>%
  #   #select(all_of(c(unique_identifiers, "cell_type")))

  return(list(immdata2, sample_meta_data, allo, nonallo))
}

```

Here we are processing the immunoseq files. The process data function: - Combines duplicate clones - Combines timepoints that are split into CD4 and CD8 and recalculates frequencies - Removes tissue timeopints - Determine Alloreactive and Nonalloreactive clones

```{r}
data <- vector("list", length(filenames))
data_meta<- vector("list", length(filenames))
all_allo <- vector("list", length(filenames))
all_nonallo <- vector("list",  length(filenames))
unique_identifiers <- c("cdr3_amino_acid", "v_resolved", "j_resolved") #Set to what you want to classify as unique clone

#Read all data and map to corresponding list
for (file_num in seq(1,length(filenames), 1)) {
  results <- process_data(filenames[file_num], path, unique_identifiers)
  data[[file_num]] <- results[[1]]
  data_meta[[file_num]] <- results[[2]]
  all_allo[[file_num]] <- results[[3]]
  all_nonallo[[file_num]] <- results[[4]]
}
# nonallo_csv <- bind_rows(all_nonallo)
# write.csv(nonallo_csv, file = "/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/nonallo.csv")
# allo_csv <- bind_rows(all_allo)
# write.csv(allo_csv, file = "/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/allo.csv")


```

```{r}
all_full_allo <- vector("list", length(filenames))
for(i in seq(1,length(data),1)){ #Get list of Allo and Nonallo clones
  immdata2 <- data[[i]]
  allo <- all_allo[[i]]
  nonallo <- all_nonallo[[i]]
  sample_meta_data <- data_meta[[i]]
  
  full_allo_inner <- subset(allo, allo$tag == 'inner')
  allo_inner <- full_allo_inner%>%
    subset(select =  unique_identifiers)
  
  full_allo_outer <- subset(allo, allo$tag == 'outer')
  allo_outer <- full_allo_outer%>%
    subset(select =  unique_identifiers)
  
  all_full_allo[[i]] <- bind_rows(full_allo_inner, full_allo_outer)
}
```

To asses how clones diversity, we use the repertoire diversity function to calculate diversity at each timepoint.
```{r}
#We are calculating the diversity metrics for alloreactive, nonalloreactive and all clones
data_all <- vector("list", length(filenames))
data_allo <- vector("list", length(filenames))
data_allo_w0 <- vector("list", length(filenames)) #Storing when alloreactive clones are 0. Can't calculate diversity metric with 0s.
data_allo_inner <- vector("list", length(filenames))
data_allo_outer <- vector("list", length(filenames))
data_allo_10 <- vector("list", length(filenames))
data_nonallo <- vector("list", length(filenames))
all_nonallo_div_CD4 <- vector("list", length(filenames))
all_nonallo_div_CD8 <- vector("list", length(filenames))
data_nonallo_w0 <- vector("list", length(filenames)) #Storing when alloreactive clones are 0. Can't calculate diversity metric with 0s.
all_allo_div <- vector("list", length(filenames))
all_allo_div_CD4 <- vector("list", length(filenames))
all_allo_div_CD8 <- vector("list", length(filenames))
all_allo_inner_div <- vector("list", length(filenames))
all_allo_outer_div <- vector("list", length(filenames))
all_nonallo_div <- vector("list", length(filenames))
all_rep_div <- vector("list", length(filenames))



for(i in seq(1,length(data),1)){ #iterate through each patient and calculate diveristy metrics
  immdata2 <- data[[i]]
  allo <- all_allo[[i]]
  nonallo <- all_nonallo[[i]]
  sample_meta_data <- data_meta[[i]]
  
  
  
  full_allo_inner <- subset(allo, allo$tag == 'inner')
  allo_inner <- full_allo_inner%>%
    subset(select =  c(unique_identifiers, "PatientID"))
  
  full_allo_outer <- subset(allo, allo$tag == 'outer')
  allo_outer <- full_allo_outer%>%
    subset(select =  c(unique_identifiers, "PatientID"))
  
  
  
  allo <- allo %>% subset(select =  c(unique_identifiers, "cell_type")) # Separate out CD4 and CD8 to calculate metric on each.
  allo_CD4 <- allo[grepl("4", allo$cell_type) & !grepl("8", allo$cell_type),]
  allo_CD8 <- allo[grepl("8", allo$cell_type) & !grepl("4", allo$cell_type),]
  
  nonallo <- nonallo%>%
    subset(select =c(unique_identifiers, "cell_type", "PatientID"))
  nonallo_CD4 <- nonallo[grepl("4", nonallo$cell_type) & !grepl("8", nonallo$cell_type),]
  nonallo_CD8 <- nonallo[grepl("8", nonallo$cell_type) & !grepl("4", nonallo$cell_type),]
  
  all <- bind_rows(allo, nonallo)
  immdata_all <- lapply(immdata2, function(x){
    x%>%
      dplyr::inner_join(all, by = unique_identifiers)%>%
      mutate(cell_type = cell_type.y)%>%
      subset(select = -c(cell_type.x, cell_type.y))
  })
  
  all_rep_div[[i]] <-repertoire_diversity(immdata_all)
  
  immdata_alloreactives <- lapply(immdata2, function(x) {
      x%>%
      dplyr::inner_join(allo, by =  unique_identifiers)%>%
      mutate(cell_type = cell_type.y)%>%
      subset(select = -c(cell_type.x, cell_type.y))
  })
  
  immdata_alloreactives_w0 <- lapply(immdata2, function(x) {
      curr_timepoint <- unique(x$timepoint)
      x%>%
      dplyr::right_join(allo, by =  unique_identifiers)%>%
      mutate(cell_type = ifelse(is.na(cell_type.y), "CD4 CD8", cell_type.y),
             templates = ifelse(is.na(templates), 0, templates),
             frequency = ifelse(is.na(frequency), 0, frequency),
             PatientID = sample_meta_data[1,1],
             timepoint = curr_timepoint)%>%
      subset(select = -c(cell_type.x, cell_type.y))
  })
  
  immdata_alloreactives_CD4 <- lapply(immdata2, function(x) {
      x%>%
      dplyr::inner_join(allo_CD4, by =  unique_identifiers)%>%
      mutate(cell_type = cell_type.y)%>%
      subset(select = -c(cell_type.x, cell_type.y))
  })
  immdata_alloreactives_CD8 <- lapply(immdata2, function(x) {
      x%>%
      dplyr::inner_join(allo_CD8, by =  unique_identifiers)%>%
      mutate(cell_type = cell_type.y)%>%
      subset(select = -c(cell_type.x, cell_type.y))
  })

  all_allo_div[[i]] <-repertoire_diversity(immdata_alloreactives)
  all_allo_div_CD4[[i]] <-repertoire_diversity(immdata_alloreactives_CD4)
  all_allo_div_CD8[[i]] <-repertoire_diversity(immdata_alloreactives_CD8)
  
  
  #List of alloreactives
  immdata_alloreactives_inner <- lapply(immdata2, function(x) {
    x <- dplyr::inner_join(allo_inner, x, by =  unique_identifiers)
  })
  
  
  # repertoire diversity of allo-reactives
  all_allo_inner_div[[i]] <-repertoire_diversity(immdata_alloreactives_inner)
  
  #List of alloreactives
  immdata_alloreactives_outer <- lapply(immdata2, function(x) {
    x <- dplyr::inner_join(allo_outer, x, by =  unique_identifiers)
  })
  
  # repertoire diversity of allo-reactives
  all_allo_outer_div[[i]] <-repertoire_diversity(immdata_alloreactives_outer)
  
  
  
  # List of non-alloreactives
  immdata_nonalloreactives <- lapply(immdata2, function(x) {
    x %>% 
      dplyr::inner_join(nonallo, by =  unique_identifiers)%>%
      mutate(cell_type = cell_type.y)%>%
      subset(select = -c(cell_type.x, cell_type.y))
  })
  
  immdata_nonalloreactives_CD4 <- lapply(immdata2, function(x) {
      x%>%
      dplyr::inner_join(nonallo_CD4, by =  unique_identifiers)%>%
      mutate(cell_type = cell_type.y)%>%
      subset(select = -c(cell_type.x, cell_type.y))
  })
  immdata_nonalloreactives_CD8 <- lapply(immdata2, function(x) {
      x%>%
      dplyr::inner_join(nonallo_CD8, by =  unique_identifiers)%>%
      mutate(cell_type = cell_type.y)%>%
      subset(select = -c(cell_type.x, cell_type.y))
  })
  
  immdata_nonalloreactives_w0 <- lapply(seq_along(immdata2), function(x) {
      immdata2[[x]]%>%
      dplyr::right_join(nonallo, by =  c(unique_identifiers, "PatientID"))%>%
      mutate(cell_type = cell_type.y,
             templates = ifelse(is.na(templates), 0, templates),
             frequency = ifelse(is.na(frequency), 0, frequency),
             PatientID = strsplit(names(immdata2)[x], "_")[[1]][1],
             timepoint = strsplit(names(immdata2)[x], "_")[[1]][2])%>%
      subset(select = -c(cell_type.x, cell_type.y))
  })
  names(immdata_nonalloreactives_w0) = names(immdata2)
  
  # repertoire diversity of non-allo-reactives
  all_nonallo_div[[i]]<-repertoire_diversity(immdata_nonalloreactives)
  all_nonallo_div_CD4[[i]]<-repertoire_diversity(immdata_nonalloreactives_CD4)
  all_nonallo_div_CD8[[i]]<-repertoire_diversity(immdata_nonalloreactives_CD8)
  
  # List of top 10 non-alloreactives
  immdata_nonalloreactives_top10 <- lapply(immdata2, function(x) {
    x %>% 
      dplyr::anti_join(allo, by =  unique_identifiers) %>% 
      arrange(desc(frequency)) %>% 
      slice_head(n = 10)
  })
  
  # List of top 10 alloreactives
  immdata_alloreactives_top10 <- lapply(immdata2, function(x) {
    x %>% 
      dplyr::inner_join(allo, by =  unique_identifiers) %>%
      arrange(desc(frequency)) %>% 
      slice_head(n = 10)
  })
  data_all[[i]] <- immdata_all
  data_nonallo[[i]] <- immdata_nonalloreactives
  data_nonallo_w0[[i]] <- immdata_nonalloreactives_w0
  data_allo[[i]] <- immdata_alloreactives
  data_allo_w0[[i]] <- immdata_alloreactives_w0
  data_allo_inner[[i]] <- immdata_alloreactives_inner
  data_allo_outer[[i]] <- immdata_alloreactives_outer
  data_allo_10[[i]]<- immdata_alloreactives_top10
}
```

```{r}
meta_data <- read_excel("/Users/lingtingshi/Documents/GVHD_project/From_Michael/TCR_Project.xlsx", sheet = 'Patient Metadata')%>%
  mutate(PatientID = gsub("[^0-9.-]", "R", PatientID), PTCy = ifelse(grepl("PTCy", PTCy), "PTCy", "No PTCy"))

# meta_data <- read_excel("/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/TCR_Project.xlsx", sheet = 'Patient Metadata')
meta_data <- meta_data[1:29, ]
meta_data$PatientID <- paste("R", sprintf("%03d", as.integer(substring(meta_data$PatientID, 2))), sep = "")

```


```{r}
full_allo_input <- all_full_allo #determine the expanded clones
elbow <- vector("list", length(filenames))
expanded_clones <- vector("list", length(filenames))
for(i in seq(1,length(data),1)){
  full_allo <- full_allo_input[[i]]
  allo_cd48 = full_allo
  
  sorted_CD48_allo <-allo_cd48[order(-allo_cd48$frequency.CFSELo),] 
  simple_cutoff <- data.frame(findCutoff(as.numeric(rownames(sorted_CD48_allo)), sorted_CD48_allo$frequency.CFSELo))
  lo <- loess(sorted_CD48_allo$frequency.CFSELo~as.numeric(rownames(sorted_CD48_allo)), span = 0.4, degree = 1)
  line_data <- data.frame(y = predict(lo), x = seq_along(predict(lo)))
  lo_cutoff <- data.frame(findCutoff(as.numeric(rownames(sorted_CD48_allo)), predict(lo,as.numeric(rownames(sorted_CD48_allo))), method = "curvature"))
  
  expanded_plot <- ggplot(sorted_CD48_allo, aes(x = seq_along(frequency.CFSELo), y = frequency.CFSELo))+
    geom_point()+
    geom_line(data = line_data, color = "red", size = 1, aes(x = x, y = y))+
    geom_point(data = simple_cutoff, color = "black", size = 2, aes(x = x, y = y))+
    geom_point(data = lo_cutoff, color = "red", size = 2, aes(x = x, y = y))+
    theme_pubr()+
    labs(title = filenames[i])
  # plot(as.numeric(rownames(sorted_CD48_allo)), sorted_CD48_allo$frequency.CFSELo, main = filenames[i])

  plot(expanded_plot)
  
  ggsave(filename = paste0("/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig//expanded_plot_", filenames[i], ".png"), 
         plot = expanded_plot, width = 4, height = 3, units = "in", dpi = 300)
  ggsave(filename = paste0("/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig//expanded_plot_", filenames[i], ".pdf"), 
         plot = expanded_plot, width = 4, height = 3, units = "in", dpi = 300)

  
  
  elbow[[i]] <-findCutoff(as.numeric(rownames(sorted_CD48_allo)), predict(lo,as.numeric(rownames(sorted_CD48_allo))),  method = "curvature")
  top_allo_48 <-allo_cd48[allo_cd48$frequency.CFSELo >= elbow[[i]]$y,]
  expanded_clones[[i]] <- top_allo_48
  

}

```
```{r}
# Function to count unique MLRCFSElo tags by patient ID
count_unique_tags_by_patient <- function(data_allo) {
  # Create a data frame to store results
  results <- data.frame(
    patient_id = character(),
    num_unique_tags = integer(),
    stringsAsFactors = FALSE
  )
  
  # Loop through each element in data_allo
  for (i in seq_along(data_allo)) {
    # Get all names at this level that end with "_MLRCFSElo_both"
    mlrcfselo_names <- grep("MLRCFSElo", names(data_allo[[i]]), value = TRUE)
    
    for (mlrcfselo_name in mlrcfselo_names) {
      # Extract patient ID from the name (assuming format like "R051_MLRCFSElo_both")
      patient_id <- sub("^([^_]+)_.*$", "\\1", mlrcfselo_name)
      
      # Get the data
      mlrcfselo_data <- data_allo[[i]][[mlrcfselo_name]]
      
      # Count unique tags based on data type
      num_unique <- 0
      if (is.data.frame(mlrcfselo_data) || is.matrix(mlrcfselo_data)) {
        # For tibbles/data frames, look for tag columns
        if ("tag" %in% colnames(mlrcfselo_data)) {
          num_unique <- length(unique(mlrcfselo_data$tag))
        } else {
          # If no obvious tag column, count rows as a fallback
          num_unique <- nrow(mlrcfselo_data)
        }
      } else if (is.vector(mlrcfselo_data)) {
        # For vectors, count unique elements
        num_unique <- length(unique(mlrcfselo_data))
      }
      
      # Add to results
      results <- rbind(results, data.frame(
        patient_id = patient_id,
        num_unique_tags = num_unique,
        stringsAsFactors = FALSE
      ))
    }
  }
  
  # Aggregate by patient_id (in case the same patient appears multiple times)
  if (nrow(results) > 0) {
    agg_results <- aggregate(num_unique_tags ~ patient_id, data = results, FUN = sum)
    return(agg_results)
  } else {
    return(results)
  }
}

# Run the function
patient_tag_counts <- count_unique_tags_by_patient(data_allo)

# Define the custom order with R prefix and 3-digit format
custom_order <- c("R051", "R066", "R067", "R068", "R069", "R070", "R076", "R081", "R086", 
                  "R090", "R093", "R101", "R143", "R137", "R150", "R153", "R091", 
                  "R098", "R160", "R179")

# Create a factor with the custom order
patient_tag_counts$patient_id <- factor(patient_tag_counts$patient_id, 
                                        levels = custom_order)

# Sort the data frame by the factor
ordered_counts <- patient_tag_counts[order(patient_tag_counts$patient_id), ]

# Print results in requested order
print(ordered_counts)

# Create a bar plot of the results in the custom order
if (requireNamespace("ggplot2", quietly = TRUE)) {
  library(ggplot2)
  ggplot(ordered_counts, aes(x = patient_id, y = num_unique_tags)) +
    geom_bar(stat = "identity", fill = "steelblue") +
    labs(title = "Number of Unique Tags by Patient ID",
         x = "Patient ID",
         y = "Number of Unique Tags") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
}
```
```{r}
# Calculate number of unique clones (based on CDR3 amino acid + V + J) per patient and timepoint
# Uses data_all_squished as input

# Load required libraries
library(dplyr)
library(tidyr)

# Function to count unique clones
count_unique_clones <- function(data_all_squished) {
  # Ensure data is properly formatted
  if (!all(c("cdr3_amino_acid", "v_resolved", "j_resolved", "PatientID", "timepoint") %in% colnames(data_all_squished))) {
    stop("Required columns not found in dataset")
  }
  
  # 1. Count unique clones per patient
  clones_per_patient <- data_all_squished %>%
    group_by(PatientID) %>%
    summarise(
      unique_clones = n_distinct(paste(cdr3_amino_acid, v_resolved, j_resolved)),
      .groups = "drop"
    ) %>%
    arrange(PatientID)
  
  # 2. Count unique clones per timepoint
  clones_per_timepoint <- data_all_squished %>%
    group_by(timepoint) %>%
    summarise(
      unique_clones = n_distinct(paste(cdr3_amino_acid, v_resolved, j_resolved)),
      .groups = "drop"
    ) %>%
    arrange(timepoint)
  
  # 3. Count unique clones per patient and timepoint
  clones_per_patient_timepoint <- data_all_squished %>%
    group_by(PatientID, timepoint) %>%
    summarise(
      unique_clones = n_distinct(paste(cdr3_amino_acid, v_resolved, j_resolved)),
      .groups = "drop"
    ) %>%
    arrange(PatientID, timepoint)
  
  # Return all results as a list
  return(list(
    by_patient = clones_per_patient,
    by_timepoint = clones_per_timepoint,
    by_patient_timepoint = clones_per_patient_timepoint
  ))
}

# Example usage
# results <- count_unique_clones(data_all_squished)
# 
# # Print results
# print("Unique clones per patient:")
# print(results$by_patient)
# 
# print("Unique clones per timepoint:")
# print(results$by_timepoint)
# 
# print("Unique clones per patient and timepoint:")
# print(results$by_patient_timepoint)
# 
# # Optional: Export results to CSV files
# write.csv(results$by_patient, "unique_clones_by_patient.csv", row.names = FALSE)
# write.csv(results$by_timepoint, "unique_clones_by_timepoint.csv", row.names = FALSE)
# write.csv(results$by_patient_timepoint, "unique_clones_by_patient_timepoint.csv", row.names = FALSE)

# Alternative approach with additional metrics
calculate_clone_metrics <- function(data_all_squished) {
  # Basic counts
  clone_metrics <- data_all_squished %>%
    group_by(PatientID, timepoint) %>%
    summarise(
      total_templates = sum(templates, na.rm = TRUE),
      unique_clones = n_distinct(paste(cdr3_amino_acid, v_resolved, j_resolved)),
      unique_cdr3 = n_distinct(cdr3_amino_acid),
      unique_v_genes = n_distinct(v_resolved),
      unique_j_genes = n_distinct(j_resolved),
      .groups = "drop"
    )
  
  # Calculate summary per patient (across all timepoints)
  patient_summary <- data_all_squished %>%
    group_by(PatientID) %>%
    summarise(
      total_templates = sum(templates, na.rm = TRUE),
      unique_clones = n_distinct(paste(cdr3_amino_acid, v_resolved, j_resolved)),
      unique_cdr3 = n_distinct(cdr3_amino_acid),
      unique_v_genes = n_distinct(v_resolved),
      unique_j_genes = n_distinct(j_resolved),
      timepoints = n_distinct(timepoint),
      .groups = "drop"
    )
  
  return(list(
    detailed = clone_metrics,
    patient_summary = patient_summary
  ))
}

# Example of how to execute this function:
# results <- calculate_clone_metrics(data_all_squished)
# print(results$detailed)
# print(results$patient_summary)

results <- count_unique_clones(data_all_squished)

# View results
print(results$by_patient)
print(results$by_timepoint)
print(results$by_patient_timepoint)
```

```{r}
clones_per_patient_timepoint <- data_squished %>%
    group_by(PatientID, timepoint) %>%
    summarise(
      unique_clones = n_distinct(paste(cdr3_amino_acid, v_resolved, j_resolved)),
      .groups = "drop"
    ) %>%
    arrange(PatientID, timepoint)
```




```{r}
# Function to sum template values by patient ID
sum_template_by_patient <- function(data_all) {
  # Create a data frame to store results
  results <- data.frame(
    patient_id = character(),
    sum_template = numeric(),
    stringsAsFactors = FALSE
  )
  
  # Loop through each element in data_allo
  for (i in seq_along(data_allo)) {
    # Get all names at this level that end with "_MLRCFSElo_both"
    mlrcfselo_names <- grep("MLRCFSElo", names(data_allo[[i]]), value = TRUE)
    
    for (mlrcfselo_name in mlrcfselo_names) {
      # Extract patient ID from the name (assuming format like "R051_MLRCFSElo_both")
      patient_id <- sub("^([^_]+)_.*$", "\\1", mlrcfselo_name)
      
      # Get the data
      mlrcfselo_data <- data_allo[[i]][[mlrcfselo_name]]
      
      # Sum template values based on data type
      template_sum <- 0
      if (is.data.frame(mlrcfselo_data) || is.matrix(mlrcfselo_data)) {
        # For tibbles/data frames, look for template column
        if ("templates" %in% colnames(mlrcfselo_data)) {
          # Sum the template column, handling NA values
          template_sum <- sum(mlrcfselo_data$templates, na.rm = TRUE)
        }
      } 
      
      # Add to results
      results <- rbind(results, data.frame(
        patient_id = patient_id,
        sum_template = template_sum,
        stringsAsFactors = FALSE
      ))
    }
  }
  
  # Aggregate by patient_id (in case the same patient appears multiple times)
  if (nrow(results) > 0) {
    agg_results <- aggregate(sum_template ~ patient_id, data = results, FUN = sum)
    return(agg_results)
  } else {
    return(results)
  }
}

# Run the function
patient_template_sums <- sum_template_by_patient(data_allo)

# Define the custom order with R prefix and 3-digit format
custom_order <- c("R051", "R066", "R067", "R068", "R069", "R070", "R076", "R081", "R086", 
                  "R090", "R093", "R101", "R143", "R137", "R150", "R153", "R091", 
                  "R098", "R160", "R179")

# Create a factor with the custom order
patient_template_sums$patient_id <- factor(patient_template_sums$patient_id, 
                                        levels = custom_order)

# Sort the data frame by the factor
ordered_sums <- patient_template_sums[order(patient_template_sums$patient_id), ]

# Print results in requested order
print(ordered_sums)

# Create a bar plot of the results in the custom order
if (requireNamespace("ggplot2", quietly = TRUE)) {
  library(ggplot2)
  ggplot(ordered_sums, aes(x = patient_id, y = sum_template)) +
    geom_bar(stat = "identity", fill = "steelblue") +
    labs(title = "Sum of Template Values by Patient ID",
         x = "Patient ID",
         y = "Sum of Template Values") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
}
```

```{r}
# Assuming "template" is a specific column in each tibble:
sum_templates <- 0

# Loop through all tibbles in data_all[[1]]
for(i in 1:length(data_all[[1]])) {
  # Check if the current item is a tibble/data frame
  if(is.data.frame(data_all[[1]][[i]])) {
    # Check if the tibble has a column called "template" (or whatever your column name is)
    if("templates" %in% colnames(data_all[[1]][[i]])) {
      # Sum the values in the "template" column and add to our running total
      sum_templates <- sum_templates + sum(data_all[[1]][[i]]$templates, na.rm = TRUE)
    }
  }
}

# Print the result
print(sum_templates)
```


```{r}

# add cdr3vj 
data_allo <- map(data_allo, function(list1) {
  map(list1, ~ mutate(.x, tag = paste(cdr3_amino_acid, v_resolved, j_resolved, sep = " ")))
})

data_allo_w0 <- map(data_allo_w0, function(list1) {
  map(list1, ~ mutate(.x, tag = paste(cdr3_amino_acid, v_resolved, j_resolved, sep = " ")))
})

expanded_clones <- map(expanded_clones, ~ mutate(.x, tag = paste(cdr3_amino_acid, v_resolved, j_resolved, sep = " ")))
```

```{r}
# map data_allo with expanded_clones
for (i in 1:length(data_allo)) {
  for (j in 1:length(data_allo[[i]])) {
    data_allo[[i]][[j]] <- data_allo[[i]][[j]] %>%
      mutate(ifexpanded = ifelse(tag %in% expanded_clones[[i]]$tag, TRUE, FALSE))
  }
}

for (i in 1:length(data_allo_w0)) {
  for (j in 1:length(data_allo_w0[[i]])) {
    data_allo_w0[[i]][[j]] <- data_allo_w0[[i]][[j]] %>%
      mutate(ifexpanded = ifelse(tag %in% expanded_clones[[i]]$tag, TRUE, FALSE))
  }
}
```


```{r}
cd48_allo <- vector("list", length(filenames))

for (file_num in seq_along(filenames)) {
  cd48_allo[[file_num]] <- vector("list", length(data_allo_w0[[file_num]]))
  
  for (i in seq_along(data_allo_w0[[file_num]])) {
    cd48_allo[[file_num]][[i]] <- data_allo_w0[[file_num]][[i]] %>%
      mutate(ifexpanded = as.character(ifexpanded))
  }
}

data_CD48_all <- bind_rows(do.call("c", cd48_allo))
data_CD48_all <- dplyr::full_join(data_CD48_all, meta_data, by = 'PatientID')
data_CD48_all_expand <- data_CD48_all[data_CD48_all$ifexpanded==TRUE,]
data_CD48_all_expand <- data_CD48_all_expand[!is.na(data_CD48_all_expand$cdr3_amino_acid),]

data_CD48_all <- data_CD48_all

data_CD48_all_expand <- data_CD48_all_expand
```

```{r}
clinical_data <- read_excel("/Users/lingtingshi/Documents/GVHD_project/From_Michael/TCR_Project.xlsx", sheet = "Patient Metadata cont.")
clinical_data <- clinical_data[c('PatientID', 'Age_R', 'HLA_mis', 'DLI_PTD', 'Relapse', 'Off Immunosuppresion', 'GVHD_DX_Post_Transplant', 'GVHD Max Grade Post Transplant', 'EBV +', 'HHV6 +', 'Systemic_Steroids', 'CMV +', "Chronic")]

clinical_data$PatientID <- as.character(clinical_data$PatientID)
# clinical_data$PatientID <- paste("R", sprintf("%03d", as.integer(clinical_data$PatientID)), sep = "")
meta_data_clin <- dplyr::full_join(meta_data, clinical_data, by = "PatientID")%>%
  filter(!is.na(DonorType))
```

```{r}
data_allo_squished <- lapply(data_allo, function(x){
    bind_rows(x)
})
data_allo_squished <- bind_rows(data_allo_squished)
data_allo_squished$allo <- TRUE

data_nonallo_squished <- lapply(data_nonallo, function(x){
    bind_rows(x)
})
data_nonallo_squished <- bind_rows(data_nonallo_squished)
data_nonallo_squished$allo <- FALSE
data_nonallo_squished$ifexpanded <- FALSE

data_squished <- bind_rows(data_allo_squished, data_nonallo_squished)
data_squished_meta <- dplyr::inner_join(data_squished, meta_data_clin, by = "PatientID")


# write.csv(data_squished_meta, file = "/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/AllClones.csv")

```

```{r}
data_allo_squished_w0 <- lapply(data_allo_w0, function(x){
    bind_rows(x)
})
data_allo_squished_w0 <- bind_rows(data_allo_squished_w0)
data_allo_squished_w0$type <- "allo"

data_nonallo_squished_w0 <- lapply(data_nonallo_w0, function(x){
    bind_rows(x)
})
data_nonallo_squished_w0 <- bind_rows(data_nonallo_squished_w0)
data_nonallo_squished_w0$type <- "nonallo"
data_nonallo_squished_w0$ifexpanded <- FALSE

data_squished_w0 <- bind_rows(data_allo_squished_w0, data_nonallo_squished_w0)
data_squished_meta_w0 <- dplyr::inner_join(data_squished_w0, meta_data_clin, by = "PatientID")

length(data_squished_meta_w0$cdr3_amino_acid)

# mapped_cells_all <- read_csv("/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/tcr_sel_all_mapped.csv",)
# mapped_cells_all <- mapped_cells_all%>%
#   mutate(cdr3_amino_acid = cdr3, PatientID = ifelse(nchar(gsub("C","",PatientID)) == 3, gsub("C","R",PatientID), gsub("C","R0",PatientID)), mapped = TRUE,
#          v_resolved = paste0("TRBV",V1, "-", V2), j_resolved = paste0("TRBJ", J1, "-", J2))
# mapped_cells <- mapped_cells_all%>%
#   select(PatientID, cdr3_amino_acid, v_resolved, j_resolved, mapped, Tissue_Location, Migration_all, leiden_0.5)
# 
# mapped_cells <- mapped_cells[!duplicated(mapped_cells, by = c("PatientID","cdr3_amino_acid", "v_resolved", "j_resolved")),]
# data_squished_meta_w0 <- data_squished_meta_w0%>%
#   left_join(mapped_cells, by = c("PatientID","cdr3_amino_acid", "v_resolved", "j_resolved"), multiple = "any")%>%
#   replace_na(list(mapped = FALSE))%>%
#   dplyr::mutate(tag = paste(cdr3_amino_acid, v_resolved, j_resolved), mapped = ifelse(mapped, "Mapped", "Not Mapped"))%>%
#   dplyr::mutate(type = paste(type, mapped))



# write.csv(data_squished_meta_w0, file = "/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/AllClones.csv")

```

```{r}
all_allo_squished <- bind_rows(data_allo_squished_w0)%>%
  mutate(tag = paste(cdr3_amino_acid, v_resolved, j_resolved), type = "Allo")%>%
  select(PatientID, tag, type)%>%
  distinct()
all_nonallo_squished <- bind_rows(data_nonallo_squished_w0)%>%
  mutate(tag = paste(cdr3_amino_acid, v_resolved, j_resolved), type = "Non-Allo")%>%
  select(PatientID, tag, type)%>%
  distinct()

expanded_clones_squished <- bind_rows(expanded_clones)
data_all_squished <- bind_rows(lapply(data, bind_rows))%>%
  mutate(tag = paste(cdr3_amino_acid, v_resolved, j_resolved))

data_all_squished_res <- data.frame()
for(patient in unique(data_squished_w0$PatientID)){
nonallo_pat <- all_nonallo_squished[all_nonallo_squished$PatientID == patient,]
allo_pat <- all_allo_squished[all_allo_squished$PatientID == patient,]
temp_squished<- data_squished_w0[data_squished_w0$PatientID == patient,]%>%
  mutate(ifexpanded = ifelse(tag %in% expanded_clones_squished[expanded_clones_squished$PatientID == patient,]$tag, TRUE, FALSE))%>%
  left_join(allo_pat, by = c("PatientID","tag"), relationship = "many-to-one")%>%
  left_join(nonallo_pat, by = c("PatientID","tag"), relationship = "many-to-one")%>%
  mutate(type = ifelse(is.na(type.x), ifelse(is.na(type.y), "Other", type.y), type.x))%>%
  select(-c("type.y", "type.x"))

data_all_squished_res <- bind_rows(data_all_squished_res,temp_squished)
}

data_all_squished <- data_all_squished_res
data_all_squished_w0_meta <- dplyr::left_join(data_all_squished_res, meta_data_clin, by = "PatientID", multiple = "all", relationship = "many-to-one")

# write.csv(data_all_squished_w0_meta, file ="/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/AllClones_all.csv")

```

```{r}
expand_counts <- function(data) {
  
  # Create an empty list to store the expanded data frames
  expanded_data_list <- list()
  
  # Loop through each row of the data frame
  for (i in seq_len(nrow(data))) {
    # Extract the id and count for the current row
    cdr3 <- data$cdr3_amino_acid[i]
    v <- data$v_resolved[i]
    j <- data$j_resolved[i]
    count <- data$templates[i]
    
    # Create a data frame with 'count' rows and a single 1 in the 'count' column
    expanded_data <- data.frame(cdr3_amino_acid = rep(cdr3, count),v_resolved = rep(v, count),j_resolved = rep(j, count), templates = rep(1, count))
    
    # Add the expanded data frame to the list
    expanded_data_list[[i]] <- expanded_data
  }
  
  # Combine all the expanded data frames into a single data frame
  expanded_data <- do.call(rbind, expanded_data_list)
  
  # Return the expanded data frame
  return(expanded_data)
}

downsample <- function(immdata, num){
  immdata_exp <- expand_counts(immdata)
  immdata_exp_samp <- sample_n(immdata_exp, num)%>%
    group_by(cdr3_amino_acid, v_resolved, j_resolved)%>%
    summarise(templates = sum(templates), .groups="drop")%>%
    mutate(frequency = templates/sum(templates))
  
  return(immdata_exp_samp)
  
}

# immdata <- data_all[[i]]
# immdata_1 <- immdata[[1]]
# 
# res <- downsample(immdata_1, 1000)
# print(sum(res$templates))
```


```{r}

  str_list = c("unstimulated","MLRCFSElo")
  for (idx in sort(na.omit(data.frame(strtoi(unique(data_all_squished_w0_meta$timepoint))))[,1]))
  {
    str_list <-append(str_list, toString(idx))
  }

avg_timepoints <- data_all_squished_w0_meta %>%
  group_by(PatientID,timepoint)%>%
  summarise(freq = sum(frequency))%>%
  filter(timepoint %in% str_list)%>%
  ungroup()%>%
  group_by(PatientID)%>%
  summarise(ntp = n())

min_ntp <- min(avg_timepoints$ntp)
max_ntp <- max(avg_timepoints$ntp)
avg_ntp <- mean(avg_timepoints$ntp)

print(paste("Min # tp = ", min_ntp, ", Max # tp = ", max_ntp, ", Average # tp = ", avg_ntp))
```

```{r}
jensen_shannon_all <- vector(mode = "list", length(data))
jensen_shannon_all_vj <- vector(mode = "list", length(data))
jensen_shannon_all_delta <- vector(mode = "list", length(data))


# Calculate Jensen Shannon Divergence between each timepoint in each patient. 0 is the same and 1 is different. 

for(i in seq(1,length(data),1)){ 
  immdata <- data[[i]]
  names(immdata) <- sapply(strsplit(names(immdata),"_"), function(x){x[[2]]})
  
  str_list = c("unstimulated","MLRCFSElo")
  for (idx in sort(na.omit(data.frame(strtoi(names(immdata))))[,1]))
  {
    str_list <-append(str_list, toString(idx))
  }
  
  immdata <- immdata[str_list]
  # immdata_all <- bind_rows(immdata)%>%
  #   group_by(cdr3_amino_acid,v_resolved, j_resolved)%>%
  #   summarise(frequency = sum(frequency))%>%
  #   arrange(desc(frequency))%>%
  #   select(cdr3_amino_acid,v_resolved, j_resolved)%>%
  #   head(20000)
  
  immdata_top <- bind_rows(immdata)%>%
    group_by(cdr3_amino_acid, v_resolved, j_resolved, PatientID)%>%
    # filter(timepoint != "unstimulated", timepoint != "MLRCFSElo")%>%
    summarise(mean_freq = mean(frequency))%>%
    arrange(desc(mean_freq))%>%
    head(500)%>%
    select(cdr3_amino_acid, v_resolved, j_resolved)
  
  immdata <- lapply(immdata, function(x) {
    x %>%
      dplyr::group_by(cdr3_amino_acid,v_resolved, j_resolved) %>% #Combine any duplicates and drop any unnecessary rows 
      # dplyr::inner_join(immdata_top, by = c("cdr3_amino_acid","v_resolved", "j_resolved"))%>%
      dplyr::summarise(templates = sum(templates), .groups = "drop") %>%
      mutate(frequency = templates / sum(templates)) %>%
      arrange(desc(frequency))  
      })
  
  
  immdata_vj <- lapply(immdata, function(x) {
    x %>%
      dplyr::group_by(v_resolved, j_resolved) %>% #Combine any duplicates and drop any unnecessary rows 
      dplyr::summarise(templates = sum(templates), .groups = "drop") %>%
      mutate(frequency = templates / sum(templates)) %>%
      arrange(desc(frequency))  
      })
  
  mat <- matrix(nrow = length(immdata), ncol = length(immdata))
  colnames(mat) <- names(immdata)
  rownames(mat) <- names(immdata)
  mat_vj <- matrix(nrow = length(immdata), ncol = length(immdata))
  colnames(mat_vj) <- names(immdata_vj)
  rownames(mat_vj) <- names(immdata_vj)
  
  for(k in seq(1,length(immdata),1)){
    for(j in seq(k,length(immdata),1)){
      if(j != k){
        
        k_num <- sum(immdata[[k]]$templates)
        j_num <- sum(immdata[[j]]$templates)
        # print(k_num)
        # print(j_num)
        # if(k_num < 100 | j_num < 100){
        #   # print("GO NEXT")
        #   next
        # }
        # if (k_num > j_num & j_num > 100){
        #   temp_k <- immdata[[k]]#downsample(immdata[[k]], j_num)
        #   temp_j <- immdata[[j]]
        # }else if (k_num > 100){
        #   temp_k <- immdata[[k]]
        #   temp_j <- immdata[[j]] #downsample(immdata[[j]], k_num)
        # }
        # else{
        #   print("skipping")
        #   next
        # }
        # print(sum(temp_j$templates))
        # print(sum(temp_k$templates))
        temp_k <- immdata[[k]]
        temp_j <- immdata[[j]]
        immdata_comp <- dplyr::full_join(temp_k, temp_j, by=unique_identifiers)%>%
          replace_na(list(templates.x = 0, templates.y = 0, frequency.x = 0, frequency.y = 0))
        res <- try(jensen_shannon(immdata_comp$frequency.x, immdata_comp$frequency.y)) 
        if(is.numeric(res)){
          mat[j,k] <- res
        }else{
          mat[j,k] <- 1
        }
      }
    }
  }
  jensen_shannon_all[[i]] <- mat

# for(k in seq(1,length(immdata_vj),1)){
#     for(j in seq(k,length(immdata_vj),1)){
#       if(j != k){
#         k_num <- sum(immdata_vj[[k]]$templates)
#         j_num <- sum(immdata_vj[[j]]$templates)
#         if (k > j & j > 100){
#           temp_vj_k <- downsample(immdata_vj[[k]], j_num)
#           temp_vj_j <- immdata_vj[[j]]
#         }else if (k > 100){
#           temp_vj_k <- immdata_vj[[k]]
#           temp_vj_j <- downsample(immdata_vj[[j]], k_num)
#         }
#         immdata_comp_vj <- dplyr::full_join(temp_vj_k[[k]], temp_vj_j[[j]], by=c("v_resolved", "j_resolved"))%>%
#           replace_na(list(templates.x = 0, templates.y = 0, frequency.x = 0, frequency.y = 0))
#         mat_vj[j,k] <- jensen_shannon(immdata_comp_vj$frequency.x, immdata_comp_vj$frequency.y) 
#       }
#     }
#   }
#   jensen_shannon_all_vj[[i]] <- mat_vj
#   jensen_shannon_all_delta[[i]] <- sqrt(abs(mat-mat_vj))
}


for(i in seq(1,length(jensen_shannon_all),1)){
  sample_meta_data <- data_meta[[i]]
  j <- jensen_shannon_all[[i]]
  # j_vj <- jensen_shannon_all_vj[[i]]
  jensen_map_plot <- pheatmap(j, color = colorRampPalette(c("#E83F6F", "#FFFFFF", "#32936F"))(100), cluster_rows = FALSE, cluster_cols = FALSE, na_col = "white", border_color = NA, main = filenames[i], breaks = seq(0.1, 1.0, length.out = 100))#+
  # jensen_vj_map_plot <- pheatmap(j_vj, color = colorRampPalette(c("#E83F6F", "#FFFFFF", "#32936F"))(100), cluster_rows = FALSE, cluster_cols = FALSE, na_col = "white", border_color = NA, main = filenames[i], breaks = seq(0.1, 1.0, length.out = 100))
  # jensen_delta_map_plot <- pheatmap(sqrt(abs(j-j_vj)), color = colorRampPalette(c("#E83F6F", "#FFFFFF", "#32936F"))(100), cluster_rows = FALSE, cluster_cols = FALSE, na_col = "white", border_color = NA, main = filenames[i], breaks = seq(0.1, 1.0, length.out = 100))
  
  # ggsave(paste0('/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/Jensen/',sample_meta_data[1,1],'JensenShannonDivPlot.png'), plot = jensen_map_plot, width = 10, height = 5, dpi = 600)
}

```


```{r}
jensen_shannon_all_nonallo <- vector(mode = "list", length(data))
jensen_shannon_all_nonallo_vj <- vector(mode = "list", length(data))
jensen_shannon_all_nonallo_delta <- vector(mode = "list", length(data))

# Calculate Jensen Shannon Divergence between each timepoint in each patient. 0 is the same and 1 is different. 

for(i in seq(1,length(data),1)){ 
  immdata <- data_nonallo_w0[[i]]
  names(immdata) <- sapply(strsplit(names(immdata),"_"), function(x){x[[2]]})
  
  str_list = c("unstimulated","MLRCFSElo")
  for (idx in sort(na.omit(data.frame(strtoi(names(immdata))))[,1]))
  {
    str_list <-append(str_list, toString(idx))
  }
  
  immdata <- immdata[str_list]

  # immdata_unstim_top <- immdata[["unstimulated"]]%>%
  #   arrange(desc(frequency))%>%
  #   head(10000)%>%
  #   select(cdr3_amino_acid, v_resolved, j_resolved)
  
  immdata_top <- bind_rows(immdata)%>%
    group_by(cdr3_amino_acid, v_resolved, j_resolved, PatientID)%>%
    # filter(timepoint != "MLRCFSElo")%>%
    summarise(mean_freq = mean(frequency))%>%
    arrange(desc(mean_freq))%>%
    head(500)%>%
    select(cdr3_amino_acid, v_resolved, j_resolved)
  
  immdata <- lapply(immdata, function(x) {
    x %>%
      dplyr::group_by(cdr3_amino_acid,v_resolved, j_resolved) %>% #Combine any duplicates and drop any unnecessary rows
      # dplyr::inner_join(immdata_top, by = c("cdr3_amino_acid","v_resolved", "j_resolved"))%>%
      dplyr::summarise(templates = sum(templates), .groups = "drop") %>%
      mutate(frequency = templates / sum(templates)) %>%
      arrange(desc(frequency))
      })
  
  
  immdata_vj <- lapply(immdata, function(x) {
    x %>%
      dplyr::group_by(v_resolved, j_resolved) %>% #Combine any duplicates and drop any unnecessary rows 
      dplyr::inner_join(immdata_top, by = c("cdr3_amino_acid","v_resolved", "j_resolved"))%>%
      dplyr::summarise(templates = sum(templates), .groups = "drop") %>%
      mutate(frequency = templates / sum(templates)) %>%
      arrange(desc(frequency))  
      })
  
  mat <- matrix(nrow = length(immdata), ncol = length(immdata))
  colnames(mat) <- names(immdata)
  rownames(mat) <- names(immdata)
  
  mat_vj <- matrix(nrow = length(immdata_vj), ncol = length(immdata_vj))
  colnames(mat_vj) <- names(immdata_vj)
  rownames(mat_vj) <- names(immdata_vj)
  
  for(k in seq(1,length(immdata),1)){
    for(j in seq(k,length(immdata),1)){
      if(j != k){
        immdata_comp <- dplyr::full_join(immdata[[k]], immdata[[j]], by=unique_identifiers)#%>%
          # replace_na(list(frequency.x = 0, frequency.y = 0, templates.x = 0, templates.y = 0))
        res <- try(jensen_shannon(immdata_comp$templates.x, immdata_comp$templates.y))
        if(is.numeric(res)){
          mat[j,k] <- res
        }else{
          mat[j,k] <- 1
        }
      }
    }
  }
  jensen_shannon_all_nonallo[[i]] <- mat
  
for(k in seq(1,length(immdata),1)){
    for(j in seq(k,length(immdata),1)){
      if(j != k){
        immdata_comp <- dplyr::full_join(immdata_vj[[k]], immdata_vj[[j]], by=c("v_resolved", "j_resolved"))%>%
          replace_na(list(frequency.x = 1, frequency.y = 1, templates.x = 1, templates.y = 1))
        res <- try(jensen_shannon(immdata_comp$templates.x, immdata_comp$templates.y))
        if(is.numeric(res)){
          mat_vj[j,k] <- res
        }else{
          mat_vj[j,k] <- 1
        }
      }
    }
  }
  jensen_shannon_all_nonallo_vj[[i]] <- mat_vj
  jensen_shannon_all_nonallo_delta[[i]] <- abs(mat - mat_vj)
}

for(i in seq(1,length(jensen_shannon_all),1)){
  sample_meta_data <- data_meta[[i]]
  j <- jensen_shannon_all_nonallo[[i]]
  j_vj <- jensen_shannon_all_nonallo_vj[[i]]
  j_delta <- jensen_shannon_all_nonallo_delta[[i]]
  jensen_map_plot <- pheatmap(j, color = colorRampPalette(c("#E83F6F", "#FFFFFF", "#32936F"))(100), cluster_rows = FALSE, cluster_cols = FALSE, na_col = "white", border_color = NA, main = filenames[i], breaks = seq(0.1, 1.0, length.out = 100))#+
  # jensen_map_plot <- pheatmap(j_vj, color = colorRampPalette(c("#E83F6F", "#FFFFFF", "#32936F"))(100), cluster_rows = FALSE, cluster_cols = FALSE, na_col = "white", border_color = NA, main = filenames[i], breaks = seq(0.1, 1.0, length.out = 100))
  # jensen_map_plot <- pheatmap(j_delta, color = colorRampPalette(c("#E83F6F", "#FFFFFF", "#32936F"))(100), cluster_rows = FALSE, cluster_cols = FALSE, na_col = "white", border_color = NA, main = filenames[i], breaks = seq(0.1, 1.0, length.out = 100))
  # 
  # ggsave(paste0('/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/Jensen/',sample_meta_data[1,1],'JensenShannonDivPlot.png'), plot = jensen_map_plot, width = 10, height = 5, dpi = 600)
}

```

```{r}
jensen_shannon_all_allo <- vector(mode = "list", length(data))
jensen_shannon_all_allo_vj <- vector(mode = "list", length(data))
jensen_shannon_all_allo_delta <- vector(mode = "list", length(data))

# Calculate Jensen Shannon Divergence between each timepoint in each patient. 0 is the same and 1 is different. 

for(i in seq(1,length(data),1)){ 
  immdata <- data_allo_w0[[i]]
  
  names(immdata) <- sapply(strsplit(names(immdata),"_"), function(x){x[[2]]})
  
  str_list = c("unstimulated","MLRCFSElo")
  for (idx in sort(na.omit(data.frame(strtoi(names(immdata))))[,1])){
    str_list <-append(str_list, toString(idx))
  }
  
  immdata <- immdata[str_list]
  immdata_top <- bind_rows(immdata)%>%
    group_by(cdr3_amino_acid, v_resolved, j_resolved, PatientID)%>%
    # filter(timepoint != "unstimulated")%>%
    summarise(mean_freq = mean(frequency))%>%
    arrange(desc(mean_freq))%>%
    head(500)%>%
    select(cdr3_amino_acid, v_resolved, j_resolved)

  
  immdata_clone <- lapply(immdata, function(x) {
    x %>%
      dplyr::group_by(cdr3_amino_acid,v_resolved, j_resolved) %>% #Combine any duplicates and drop any unnecessary rows
      # filter(ifexpanded == T)%>%
      # dplyr::inner_join(immdata_top, by = c("cdr3_amino_acid", "v_resolved", "j_resolved"))%>%
      dplyr::summarise(templates = sum(templates), .groups = "drop") %>%
      mutate(frequency = templates / sum(templates)) %>%
      arrange(cdr3_amino_acid)
      })
  
  
  immdata_vj <- lapply(immdata, function(x) {
    x %>%
      dplyr::group_by(v_resolved, j_resolved) %>% #Combine any duplicates and drop any unnecessary rows 
      filter(ifexpanded == T)%>%
      dplyr::summarise(templates = sum(templates), .groups = "drop") %>%
      mutate(frequency = templates / sum(templates)) %>%
      arrange(v_resolved)  
      })
  
  mat <- matrix(nrow = length(immdata), ncol = length(immdata))
  colnames(mat) <- names(immdata)
  rownames(mat) <- names(immdata)
  
  mat_vj <- matrix(nrow = length(immdata_vj), ncol = length(immdata_vj))
  colnames(mat_vj) <- names(immdata_vj)
  rownames(mat_vj) <- names(immdata_vj)
  
  for(k in seq(1,length(immdata),1)){
    for(j in seq(k,length(immdata),1)){
      if(j != k){
        immdata_comp <- dplyr::full_join(immdata[[k]], immdata[[j]], by=unique_identifiers)#%>%
          # replace_na(list(frequency.x = 0, frequency.y = 0, templates.x = 0, templates.y = 0))
        res <- try(jensen_shannon(immdata_comp$templates.x, immdata_comp$templates.y))
        if(is.numeric(res)){
          mat[j,k] <- res
        }else{
          mat[j,k] <- 1
        }
      }
    }
  }
  jensen_shannon_all_allo[[i]] <- mat
  
  for(k in seq(1,length(immdata),1)){
    for(j in seq(k,length(immdata),1)){
      if(j != k){
        immdata_comp <- dplyr::full_join(immdata_vj[[k]], immdata_vj[[j]], by=c("v_resolved", "j_resolved"))%>%
          replace_na(list(frequency.x = 1, frequency.y = 1, templates.x = 1, templates.y = 1))
        res <- try(jensen_shannon(immdata_comp$templates.x, immdata_comp$templates.y))
        if(is.numeric(res)){
          mat_vj[j,k] <- res
        }else{
          mat_vj[j,k] <- 1
        }
      }
    }
  }
  jensen_shannon_all_allo_vj[[i]] <- mat_vj
  jensen_shannon_all_allo_delta[[i]] <- abs(mat-mat_vj)
}

for(i in seq(1,length(jensen_shannon_all_allo),1)){
  sample_meta_data <- data_meta[[i]]
  j <- jensen_shannon_all_allo[[i]]
  j_vj <- jensen_shannon_all_allo_vj[[i]]
  j_delta <- jensen_shannon_all_allo_delta[[i]]
  jensen_map_plot <- pheatmap(j, color = colorRampPalette(c("#E83F6F", "#FFFFFF", "#32936F"))(100), cluster_rows = FALSE, cluster_cols = FALSE, na_col = "white", border_color = NA, main = filenames[i], breaks = seq(0.1, 1.0, length.out = 100))#+
  # jensen_map_plot <- pheatmap(j_vj, color = colorRampPalette(c("#E83F6F", "#FFFFFF", "#32936F"))(100), cluster_rows = FALSE, cluster_cols = FALSE, na_col = "white", border_color = NA, main = filenames[i], breaks = seq(0.1, 1.0, length.out = 100))
  # jensen_map_plot <- pheatmap(j_delta, color = colorRampPalette(c("#E83F6F", "#FFFFFF", "#32936F"))(100), cluster_rows = FALSE, cluster_cols = FALSE, na_col = "white", border_color = NA, main = filenames[i], breaks = seq(0.1, 1.0, length.out = 100))
  # 
  # ggsave(paste0('/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/Jensen/',sample_meta_data[1,1],'JensenShannonDivPlot.png'), plot = jensen_map_plot, width = 10, height = 5, dpi = 600)
}

```

```{r, fig.width = 16, fig.height = 9, dpi = 600 }


type <- "GVHD_GRADE_1"

all_jsd <- data.frame()
allo_jsd <- data.frame()
nonallo_jsd <- data.frame()
for(i in seq(1, length(jensen_shannon_all), 1)){
  all <- data.frame(jensen_shannon_all[[i]])
  allo <- data.frame(jensen_shannon_all_allo[[i]])
  nonallo <- data.frame(jensen_shannon_all_nonallo[[i]])
  
  all <- all['unstimulated']
  allo <- allo['unstimulated']
  nonallo <- nonallo['unstimulated']
  
  all$timepoint <- rownames(all)
  allo$timepoint <- rownames(allo)
  nonallo$timepoint <- rownames(nonallo)
  
  all$PatientID <- sapply(str_split(filenames[i], "_"), function(x){paste0("R",x[[2]])})
  allo$PatientID <- sapply(str_split(filenames[i], "_"), function(x){paste0("R",x[[2]])})
  nonallo$PatientID <- sapply(str_split(filenames[i], "_"), function(x){paste0("R",x[[2]])})
  
  
  all[grepl("MLR", all$timepoint),]$timepoint <- "1"
  allo[grepl("MLR", allo$timepoint),]$timepoint <- "1"
  nonallo[grepl("MLR", nonallo$timepoint),]$timepoint <- "1"
  
  all[grepl("unstim", all$timepoint),]$timepoint <- "0"
  allo[grepl("unstim", allo$timepoint),]$timepoint <- "0"
  nonallo[grepl("unstim", nonallo$timepoint),]$timepoint <- "0"
  
  all <- all%>%
    mutate(timepoint = strtoi(timepoint))
  allo <- allo%>%
    mutate(timepoint = strtoi(timepoint))
  nonallo <- nonallo%>%
    mutate(timepoint = strtoi(timepoint))
  
  all_jsd <- bind_rows(all_jsd, all)
  allo_jsd <- bind_rows(allo_jsd, allo)
  nonallo_jsd <- bind_rows(nonallo_jsd, nonallo)
}

all_jsd$rep_div <- "All"
allo_jsd$rep_div <- "Alloreactive"
nonallo_jsd$rep_div <- "Non-Alloreactive"

all_jsd <- bind_rows(all_jsd, allo_jsd, nonallo_jsd)

all_jsd_unstim_mlr <- all_jsd[all_jsd$timepoint == 1,]

all_jsd_unstim_mlr_meta <- dplyr::inner_join(all_jsd_unstim_mlr, meta_data, by = "PatientID")

num_cuts <- 2 #Change this to change the binning of the timepoints

all_jsd <- all_jsd[all_jsd$timepoint <= 365 & all_jsd$timepoint != 0 & all_jsd$timepoint > 3,]%>% #get rid of timepoints before
  mutate( time_bin = cut(timepoint, breaks=quantile(timepoint, probs = 0:num_cuts/num_cuts) , include.lowest = TRUE))

all_jsd_meta <- dplyr::inner_join(all_jsd, meta_data, by = "PatientID") #Modify data for plotting
all_jsd_meta <- all_jsd_meta[all_jsd_meta$timepoint <=365,]
all_jsd_meta[all_jsd_meta$GVHD_GRADE_1 == "Unclassified/Severe",]$GVHD_GRADE_1 <- "Severe"
all_jsd_meta[all_jsd_meta$GVHD_GRADE_2 == "Unclassfied/Severe",]$GVHD_GRADE_2 <- "Severe"
all_jsd_meta <- all_jsd_meta[!is.na(all_jsd_meta$PTCy),]%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", "No and Mild"))
all_jsd_meta$GVHD_GRADE_1 <- factor(all_jsd_meta$GVHD_GRADE_1, levels = c("No and Mild", "Severe"))
all_jsd_meta$PTCy <- factor(all_jsd_meta$PTCy, levels = c("PTCy", "No PTCy"))


stat_test <- all_jsd_meta%>%
  group_by(time_bin, rep_div, PTCy)%>%
  pairwise_wilcox_test(as.formula("unstimulated ~ GVHD_GRADE_1"), p.adjust.method = "hochberg", exact = F)%>%
  add_xy_position(x = "time_bin")%>%
  mutate(p.adj_2 = p.adjust(p, method = "hochberg"))
  
tab20_colors <- c('#1f77b4','#aec7e8','#ff7f0e','#ffbb78','#2ca02c','#98df8a','#d62728','#ff9896','#9467bd','#c5b0d5','#8c564b','#c49c94','#e377c2','#f7b6d2','#7f7f7f','#c7c7c7','#bcbd22','#dbdb8d','#17becf','#9edae5')

jsd_box <- ggboxplot(all_jsd_meta, x = "time_bin", y = "unstimulated", fill= "GVHD_GRADE_1")+ #Boxplots
  geom_beeswarm(dodge.width  = 0.75, aes(group = GVHD_GRADE_1, color = PatientID), )+
  theme_minimal()+
  stat_pvalue_manual(stat_test, tip.length = 0)+
  scale_fill_manual(values=c("#09BB8C","#005A8F","#B85000", "grey"))+
  scale_color_manual(values=tab20_colors)+
  labs( y = "Jensen Shannon Divergence", x = "Days", title = "Jensen Shannon Divergence PTCy vs No PTCy", fill = "GVHD Grade")+
  facet_grid(rows = vars(PTCy), cols = vars(rep_div),space = "free", labeller = labeller(.col = c("All", "Alloreactive", "Non-Alloreactive")))+
  theme(legend.position = "left",legend.text = element_text(size=16), legend.key.size = unit(12, 'pt'), strip.text.x = element_text(size = 22), axis.title.y.left = element_text(size=18), axis.title.x.bottom = element_text(size=18), title = element_text(size = 20), axis.text.x.bottom = element_text(size = 16), panel.spacing = unit(2, "lines"), axis.text.y = element_text(size = 16),strip.text.y.right = element_text(size = 16), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black"), axis.ticks = element_line(color = "black"),  axis.ticks.length = unit(0.25, "cm"))


plot(jsd_box)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/ShannonDivergence.png'), plot = jsd_box, width = 16, height = 9, dpi = 600)

```


```{r, fig.width = 16, fig.height = 9, dpi = 600 }


type <- "GVHD_GRADE_1"

all_jsd <- data.frame()
allo_jsd <- data.frame()
nonallo_jsd <- data.frame()
for(i in seq(1, length(jensen_shannon_all), 1)){
  all <- data.frame(jensen_shannon_all[[i]])
  allo <- data.frame(jensen_shannon_all_allo[[i]])
  nonallo <- data.frame(jensen_shannon_all_nonallo[[i]])
  
  all <- all['unstimulated']
  allo <- allo['unstimulated']
  nonallo <- nonallo['unstimulated']
  
  all$timepoint <- rownames(all)
  allo$timepoint <- rownames(allo)
  nonallo$timepoint <- rownames(nonallo)
  
  all$PatientID <- sapply(str_split(filenames[i], "_"), function(x){paste0("R",x[[2]])})
  allo$PatientID <- sapply(str_split(filenames[i], "_"), function(x){paste0("R",x[[2]])})
  nonallo$PatientID <- sapply(str_split(filenames[i], "_"), function(x){paste0("R",x[[2]])})
  
  
  all[grepl("MLR", all$timepoint),]$timepoint <- "1"
  allo[grepl("MLR", allo$timepoint),]$timepoint <- "1"
  nonallo[grepl("MLR", nonallo$timepoint),]$timepoint <- "1"
  
  all[grepl("unstim", all$timepoint),]$timepoint <- "0"
  allo[grepl("unstim", allo$timepoint),]$timepoint <- "0"
  nonallo[grepl("unstim", nonallo$timepoint),]$timepoint <- "0"
  
  all <- all%>%
    mutate(timepoint = strtoi(timepoint))
  allo <- allo%>%
    mutate(timepoint = strtoi(timepoint))
  nonallo <- nonallo%>%
    mutate(timepoint = strtoi(timepoint))
  
  all_jsd <- bind_rows(all_jsd, all)
  allo_jsd <- bind_rows(allo_jsd, allo)
  nonallo_jsd <- bind_rows(nonallo_jsd, nonallo)
}

all_jsd$rep_div <- "All"
allo_jsd$rep_div <- "Alloreactive"
nonallo_jsd$rep_div <- "Non-Alloreactive"

all_jsd <- bind_rows(all_jsd, allo_jsd, nonallo_jsd)

all_jsd_unstim_mlr <- all_jsd[all_jsd$timepoint == 1,]

all_jsd_unstim_mlr_meta <- dplyr::inner_join(all_jsd_unstim_mlr, meta_data, by = "PatientID")

num_cuts <- 1 #Change this to change the binning of the timepoints

all_jsd_meta <- dplyr::inner_join(all_jsd, meta_data, by = "PatientID")%>%
  filter(rep_div != "Non-Alloreactive")#%>%
  #mutate(PTCy = ifelse(grepl("PTCy", PTCy), "PTCy", "No PTCy"))#Modify data for plotting

all_jsd_severe <- all_jsd_meta[all_jsd_meta$timepoint <= 365 & all_jsd_meta$timepoint != 0 & all_jsd_meta$timepoint > 3,]%>%
  filter(GVHD_GRADE_1 == "Severe", !is.na(unstimulated))#get rid of timepoints before

all_jsd_meta <- all_jsd_meta[all_jsd_meta$timepoint <= 365 & all_jsd_meta$timepoint != 0 & all_jsd_meta$timepoint > 3,]%>%
  mutate( time_bin = cut(timepoint, breaks=quantile(all_jsd_severe$timepoint, probs = 0:num_cuts/num_cuts) , include.lowest = TRUE))


all_jsd_meta <- all_jsd_meta[all_jsd_meta$timepoint <=365,]
all_jsd_meta <- all_jsd_meta[!is.na(all_jsd_meta$PTCy),]%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", "No and Mild"))
all_jsd_meta$GVHD_GRADE_1 <- factor(all_jsd_meta$GVHD_GRADE_1, levels = c("No and Mild", "Severe"))
all_jsd_meta$PTCy <- factor(all_jsd_meta$PTCy, levels = c("PTCy", "No PTCy"))
all_jsd_meta <- all_jsd_meta%>%
  filter(!is.na(unstimulated))
#   filter(PatientID != "R179")

stat_test <- all_jsd_meta%>%
  group_by(time_bin, rep_div, PTCy)%>%
  pairwise_wilcox_test(as.formula("unstimulated ~ GVHD_GRADE_1"), p.adjust.method = "hochberg", exact = F)%>%
  add_xy_position(x = "time_bin")%>%
  group_by(PTCy, time_bin, rep_div)%>%
  mutate(p.adj_2 = p.adjust(p, method = "hochberg"))%>%
  ungroup()
  


jsd_box <- ggboxplot(all_jsd_meta, x = "time_bin", y = "unstimulated", fill= "GVHD_GRADE_1", outlier.shape = NA)+ #Boxplots
  geom_beeswarm(dodge.width  = 0.75, aes(group = GVHD_GRADE_1, color = PatientID), )+
  theme_pubr()+
  stat_pvalue_manual(stat_test, tip.length = 0)+
  scale_fill_manual(values=c("#058b8e", "#B85000"))+
  # scale_color_manual(values=c("#058b8e", "#B85000"))+
  labs( y = "Jensen Shannon Divergence", x = "Days", title = "Jensen Shannon Divergence PTCy vs No PTCy", fill = "GVHD Grade")+
  facet_grid(rows = vars(PTCy),cols = vars(rep_div),space = "free", labeller = labeller(.col = c("All", "Alloreactive", "Non-Alloreactive")))+
  theme(legend.position = "left",legend.text = element_text(size=12), legend.key.size = unit(20, 'pt'), strip.text.x = element_text(size = 22), axis.title.y.left = element_text(size=18), axis.title.x.bottom = element_text(size=18), title = element_text(size = 20), axis.text.x.bottom = element_text(size = 16), panel.spacing = unit(2, "lines"), axis.text.y = element_text(size = 16),strip.text.y.right = element_text(size = 16), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black"), axis.ticks = element_line(color = "black"),  axis.ticks.length = unit(0.25, "cm"))



plot(jsd_box)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/ShannonDivergence.png'), plot = jsd_box, width = 16, height = 9, dpi = 600)


```

```{r}
for(patient in unique(all_jsd_meta$PatientID)){
#patient <- "R067"
  print(patient)
  curr_patient <- all_jsd_meta %>%
    filter(PatientID == patient) %>%
    mutate(dist_from_onset = timepoint-Onset)
  curr_patient_before <- curr_patient %>%
    filter(dist_from_onset < 0)
  curr_patient_before <- curr_patient_before[curr_patient_before$dist_from_onset == max(curr_patient_before$dist_from_onset),]
  curr_patient_after <- curr_patient %>%
    filter(dist_from_onset >= 0)
  curr_patient_after <- curr_patient_after[curr_patient_after$dist_from_onset == min(curr_patient_after$dist_from_onset),]
}


```

```{r}


data_unstimulated <- lapply(data, function(x){ #Collect all unstim timepoints from all patients
  x <- bind_rows(x[grepl('unstim', names(x))])
})

data_mlr <- lapply(data, function(x){ # collect all mlr timepoints from all patients
  x <- bind_rows(x[grepl('MLR', names(x))])
})


mat_unstim <- matrix(nrow = length(data_unstimulated), ncol = length(data_unstimulated))
colnames(mat_unstim) <- filenames
rownames(mat_unstim) <- filenames
mat_mlr <- mat_unstim
for(i in seq(1,length(data_unstimulated),1)){

  for(j in seq(1,length(data_unstimulated),1)){
    immdata_comp_unstim <- dplyr::full_join(data_unstimulated[[i]], data_unstimulated[[j]], by=unique_identifiers)%>%
      replace_na(list(templates.x = 0, templates.y = 0))
    mat_unstim[i,j] <- jensen_shannon(immdata_comp_unstim$templates.x, immdata_comp_unstim$templates.y)

    immdata_comp_mlr <- dplyr::full_join(data_mlr[[i]], data_mlr[[j]], by=unique_identifiers)%>%
      replace_na(list(templates.x = 0, templates.y = 0))
    mat_mlr[i,j] <- jensen_shannon(immdata_comp_mlr$templates.x, immdata_comp_mlr$templates.y)
  }
}
mat_unstim[mat_unstim == 0] <- NA
mat_mlr[mat_mlr == 0] <- NA
jensen_map_plot_unstim <- pheatmap(mat_unstim, color =  colorRampPalette(c("#E83F6F", "#FFFFFF", "#32936F"))(100), cluster_rows = TRUE, cluster_cols = TRUE, labels_row = filenames)
jensen_map_plot_mlr <- pheatmap(mat_mlr, cluster_rows = TRUE, cluster_cols = TRUE, labels_row = filenames)
# ggsave(paste0('/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/Jensen/','UnstimJensenShannonDivPlot.png'), plot = jensen_map_plot_unstim, width = 5, height = 5, dpi = 600)
# ggsave(paste0('/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/Jensen/','MLRJensenShannonDivPlot.png'), plot = jensen_map_plot_mlr, width = 5, height = 5, dpi = 600)

```

```{r, fig.width = 10, fig.height = 5}
all_pie_data <- data.frame()
data_allo_input <- data_allo_outer
all_allo_input <- lapply(all_allo, function(x){
  x%>%filter(tag == 'outer')
})
for(i in seq(1,length(data),1)){
  immdata_allo <- all_allo[[i]]
  immdata_nonallo <- data_nonallo[[i]]
  sample_meta_data <- data_meta[[i]]
  
  immdata_nonallo <- bind_rows(immdata_nonallo[[which(sample_meta_data[2,]== "unstimulated")[1]]], immdata_nonallo[[which(sample_meta_data[2,]== "MLRCFSElo")]])
  
  allo <- immdata_allo%>%
    mutate(uID = paste0(cdr3_amino_acid, v_resolved, j_resolved))
  
  nonallo <- immdata_nonallo%>%
    mutate(uID = paste0(cdr3_amino_acid, v_resolved, j_resolved))
  
  allo_outer <- allo %>% filter(tag == 'outer')
  allo_inner <- allo %>% filter(tag == 'inner')
  
  total_allo <- length(unique(allo$uID))
  total_nonallo <- length(unique(nonallo$uID))
  total_allo_inner <- length(unique(allo_inner$uID))
  total_allo_outer <- length(unique(allo_outer$uID))
  total_clones <- total_allo + total_nonallo
  
  pie_data <- data.frame(prop = c((total_nonallo)/(total_clones), (total_allo_inner+total_allo_outer)/(total_clones)), groups = c('non-alloreactive clones', 'alloreactive clones'), PatientID = c(sample_meta_data[1,1],sample_meta_data[1,1]))
  all_pie_data <-rbind(all_pie_data, pie_data)
}

all_pie_data$groups <- factor(all_pie_data$groups, levels = c('alloreactive clones','non-alloreactive clones' ))




inter_meta <- intersect(all_pie_data$PatientID, meta_data$PatientID)

meta_data_sel <- meta_data[meta_data$PatientID %in% inter_meta, ]
meta_data_sel['amount'] <- 1
all_pie_data <- dplyr::full_join(all_pie_data,meta_data_sel, by = "PatientID") #Join with GVHD Grade

all_pie_data <- all_pie_data%>%dplyr::arrange(desc(DonorType), .by_group = TRUE)
all_pie_data$PatientID <- factor(all_pie_data$PatientID, levels = unique(all_pie_data$PatientID))

# all_pie_data$GVHD_GRADE_1 <- factor(all_pie_data$GVHD_GRADE_1, levels = c( 'No', 'Mild', 'Severe'))
# all_pie_data <- all_pie_data%>%arrange(desc(GVHD_GRADE_1), .by_group = TRUE)
# all_pie_data$PatientID <- factor(all_pie_data$PatientID, levels = unique(all_pie_data$PatientID))


hlay <- rbind(c(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,NA,NA,NA), # For sizing of graph
              c(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,NA,NA,NA),
              c(NA,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2))
percent_bar <- ggplot(all_pie_data, aes(x = PatientID , y = prop, fill = groups))+ # generating percent bar graph
  geom_bar(stat = "identity", position = "stack")+
   scale_fill_brewer(palette = "Set2")+
  xlab('PatientID')+
  ylab('Proportion of unique clones')+
  theme_minimal()+
  coord_cartesian(ylim = c(.70, 1))

b <- ggplot(all_pie_data, aes(y = "Donor Type",x = amount, fill = DonorType, group = DonorType))+
  geom_bar(stat = 'identity', width = .2)+
  scale_fill_brewer(palette = "Set3")+
  theme_void()+
  guides()+
  theme(legend.key.height = unit(.2, "cm"), plot.margin = margin(0, -.75, 0, .4, "cm"), legend.margin = margin(0,1,0,-.75, "cm")) + labs(fill = 'Donor Type')

# b <- ggplot(all_pie_data, aes(y = "GVHD",x = amount, fill = GVHD_GRADE_1, group = GVHD_GRADE_1))+
#   geom_bar(stat = 'identity', width = .2)+
#   scale_fill_brewer(palette = "Set3")+
#   theme_void()+
#   guides()+
#   theme(legend.key.height = unit(.2, "cm")) + labs(y = 'GVHD Grade', fill = 'GVHD Grade')

unique_clone_plot <- grid.arrange(percent_bar, b,ncol = 1, heights = c(1,.2))

#plot(unique_clone_plot)
#ggsave(paste0('/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/','AllUniqueClone.png'), plot = unique_clone_plot, width = 10, height = 5, dpi = 600)
# ggsave(paste0('/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/','AllUniqueClone_DT.pdf'), plot = unique_clone_plot, width = 10, height = 5, dpi = 600)
# anova, if diff?
```

Patients with higher HLA mismatch are tend to have a higher proportion of alloreactive clones.

```{r, fig.height = 6, fig.width = 8}

all_pie_data_haplo <- all_pie_data %>%
  mutate(DonorTypeBin = ifelse(grepl('Haplo', DonorType), 'Haplo', 'Non-Haplo'))%>%
  subset(groups == 'alloreactive clones')
stat.test <- t.test(subset(all_pie_data_haplo, DonorTypeBin == 'Haplo')$prop, subset(all_pie_data_haplo, DonorTypeBin == 'Non-Haplo')$prop )
p <- ggplot(subset(all_pie_data_haplo, groups == 'alloreactive clones'), aes(x = DonorTypeBin, y = prop))+
  geom_boxplot()+
  geom_bracket(xmin = "Haplo", xmax = "Non-Haplo", y.position = 0.3, label = "p < 0.05", tip.length = c(0.0, 0.0))+
  geom_beeswarm(aes(x = DonorTypeBin, y = prop, fill = DonorType, color = DonorType),size = 2, cex = 2)+
  #geom_bracket(aes(xmin = "Haplo", xmax = "Non-Haplo", y.position = 0.3,label = signif(stat.test$p.value,2), tip.length = c(0.2, 0.4)))+
  # geom_bracket(xmin = c("Haplo"), xmax = c("Non-Haplo"), label = signif(stat.test$p.value,2), y.position = 35)+
  theme_minimal()
plot(p)


```

Matched unrelated donors have a higher proportion than Matched related and sibling donor.

```{r}
#Descriptive data table
descriptive_data <- data.frame()
All_unique_clones <- data.frame()
All_allo_unique_clones <- data.frame()
All_nonall_prop <- data.frame()
All_allo_prop <- data.frame()
for(i in seq(1, length(data),1)){
  immdata <- data[[i]]
  immdata_alloreactive <- data_allo[[i]]
  sample_meta_data <- data_meta[[i]]
  
  all_clones <- bind_rows(immdata, .id = 'SampleID')%>%
    group_by(SampleID)%>%
    dplyr::summarise(unique_clones = n())%>%
    mutate(SampleID = unlist(sample_meta_data[2,]))
  
  immdata_alloreactive <- lapply(immdata_alloreactive, function(x){
    if(length(x$timepoint) == 0){
      x <- rbind(x,data.frame(matrix(NA, ncol = ncol(x))))
    }else{
      x <- x
    }
  })
  
  allo_clones <- bind_rows(immdata_alloreactive, .id = 'SampleID')%>%
    group_by(SampleID)%>%
    dplyr::summarise(unique_allo_clones = n())%>%
    mutate(SampleID = unlist(sample_meta_data[2,]))
  
  clones <- dplyr::full_join(all_clones, allo_clones, by = 'SampleID')%>%
    mutate(allo_prop = unique_allo_clones/unique_clones, nonallo_prop = (unique_clones - unique_allo_clones)/unique_clones)
  
  clones$unique_allo_clones <- as.character(clones$unique_allo_clones)
  clones$unique_clones <- as.character(clones$unique_clones)
  clones$allo_prop <- as.character(clones$allo_prop)
  clones$nonallo_prop <- as.character(clones$nonallo_prop)

  clones <- clones %>%
    mutate(desc_data = paste(unique_clones, unique_allo_clones, allo_prop, nonallo_prop, sep = ", "), PatientID = sample_meta_data[1,1])

clones_wide <- pivot_wider(clones, id_cols = PatientID, names_from = SampleID, values_from = desc_data)
all_clones_wide <- pivot_wider(clones, id_cols = PatientID, names_from = SampleID, values_from = unique_clones)
allo_clones_wide <- pivot_wider(clones, id_cols = PatientID, names_from = SampleID, values_from = unique_allo_clones)
allo_prop_wide <- pivot_wider(clones, id_cols = PatientID, names_from = SampleID, values_from = allo_prop)
nonallo_prop_wide <- pivot_wider(clones, id_cols = PatientID, names_from = SampleID, values_from = nonallo_prop)

descriptive_data <- bind_rows(descriptive_data, clones_wide)
All_unique_clones <- bind_rows(All_unique_clones, all_clones_wide)
All_allo_unique_clones <- bind_rows(All_allo_unique_clones, allo_clones_wide)
All_allo_prop <- bind_rows(All_allo_prop, allo_prop_wide)
All_nonall_prop <- bind_rows(All_nonall_prop, nonallo_prop_wide)

}
 descriptive_data <- descriptive_data%>%
   select('PatientID',c(sort(colnames(descriptive_data)[!grepl("^\\d+$|MLR|unstim",colnames(descriptive_data))])), 'unstimulated', 'MLRCFSElo',
 c(as.character(sort(as.numeric(colnames(descriptive_data)[grepl("^\\d+$",colnames(descriptive_data))])))))
 
 #write.csv(descriptive_data,"/Users/pressm/Documents/AziziLab/GVHD/immunoseq_r/Descriptive_Data.csv")

```



```{r}
cum_freq <- function(allo_data, data){ 
  freqs <-as.data.frame(do.call(rbind, Map(function(x, y) {
    sum(x$frequency) / sum(y$frequency)
  }, x = allo_data, y = data)))
  return(freqs)
}
```

```{r}
avg_freq <- function(allo_data, data){ 
  freqs <-as.data.frame(do.call(rbind, Map(function(x, y) {
    mean(x$frequency) / sum(y$frequency)
  }, x = allo_data, y = data)))
  return(freqs)
}
```

```{r}

clone_frac <- function(allo_data, data){
  as.data.frame(do.call(rbind, Map(function(x, y) {
    nrow(x) / nrow(y)
  }, x = allo_data, y = data)))
}

```


```{r}

n_clone <- function(allo_data, data){
  n_clones <- as.data.frame(do.call(rbind, Map(function(x, y) {
    nrow(x)
  }, x = allo_data, y = data)))
}

```

```{r}

get_amino_arrangements <- function(immdata){#}, sample_meta_data){
  
  arrangements <- lapply(immdata, function(x){
    x %>%
      select("cdr3_amino_acid", "frequency") %>%
      mutate(cdr3_aminolength = nchar(cdr3_amino_acid))
  })
  
  arrangements <- bind_rows(arrangements, .id = "id")
  arrangements$timepoint <- str_split_fixed(arrangements$id,"_",3)[,2]
  arrangements$PatientID <- str_split_fixed(arrangements$id,"_",3)[,1]
  
  
  aa <- AAStringSet(arrangements$cdr3_amino_acid)
  freqs <- data.frame(alphabetFrequency(aa, as.prob = FALSE))
  
  arrangements <- bind_cols(arrangements, freqs)
  
  return(arrangements)
}

get_median_length <- function(arrangements, name){
  med_lengths <- arrangements %>%
    group_by(timepoint, PatientID)%>%
    dplyr::summarise(med_cdr3aminolength = mean(cdr3_aminolength), .groups = "drop")
  med_lengths[['Rep_div']] <- name
  
  return(med_lengths)
}

```


```{r}

div_fig <- function(patient_data, div, Patient){
  #selected <- subset(patient_data, patient_data$Rep_div %in% types)
  fig <- ggplot(patient_data, aes(SampleID,.data[[div]], group = Rep_div, color=Rep_div,shape = Rep_div)) +
    geom_point(size = 1) +
    geom_line() +
    scale_y_continuous(trans='log10', labels = scales::number_format(accuracy = 0.01))+
    labs(y = div, x = "Days Post Transplantation", title = Patient) +
    theme_minimal()+
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, size = 8))
  return(fig)
}

clone_fig <- function(immdata, type, title){
  fig <- ggplot(immdata, aes(timepoint, frequency, color = .data[[type]], group = tag)) +
    geom_point(size = 1) + 
    geom_line() +
    labs(title = title)+
    scale_y_continuous(trans = "log10", labels = scales::number_format(accuracy = 0.01)) +
    theme_minimal(base_size = 18)+
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, size = 8))
  return(fig)
}


add_tag <- function(immdata){
  immdata$tag <- paste(immdata$cdr3_amino_acid , immdata$v_resolved , immdata$j_resolved)
  return(immdata)
}

convert_id <- function(immdata){
  ids <- strsplit(immdata$id, "_")
  immdata$timepoints <- sapply(ids, function(x) x[[2]])
  immdata$PatientID <- sapply(ids, function(x) x[[1]])
  return(immdata)
}

```

```{r}
str_list = c("unstimulated","MLRCFSElo")
for (i in sort(na.omit(data.frame(strtoi(unique(data_CD48_all$timepoint))))[,1]))
{
  str_list <-append(str_list, toString(i))
}


data_CD48_all$timepoint <- factor(data_CD48_all$timepoint, level = str_list)
data_CD48_all <- data_CD48_all[!is.na(data_CD48_all$timepoint),]


```


```{r, fig.height = 10, fig.width=16}
data_allo_input <- data_allo
all_allo_div_input <- all_allo_div
full_allo_input <- all_full_allo
patient_data_all = data.frame()
all_amino = data.frame()
allo_amino = data.frame()

for(i in seq(1,length(data),1)){
  immdata2 <- data[[i]]
  immdata_alloreactives <- data_allo[[i]]
  immdata_nonalloreactives <- data_nonallo[[i]]
  allo <- data_allo_input[[i]]
  sample_meta_data <- data_meta[[i]]
  all_rep <- all_rep_div[[i]]
  allo_rep <- all_allo_div_input[[i]]
  nonallo_rep <- all_nonallo_div[[i]]
  full_allo <- full_allo_input[[i]]
  curr_file <- filenames[i]
  
  
  #add meta data
  all_rep$PatientID <- t(data.frame(strsplit(all_rep$ID, split = "_")))[,1]
  all_rep$SampleID <- t(data.frame(strsplit(all_rep$ID, split = "_")))[,2]
  all_rep$Rep_div <-"all"
  allo_rep$PatientID <- t(data.frame(strsplit(allo_rep$ID, split = "_")))[,1]
  allo_rep$SampleID <- t(data.frame(strsplit(allo_rep$ID, split = "_")))[,2]
  allo_rep$Rep_div <-"allo"
  nonallo_rep$PatientID <- t(data.frame(strsplit(nonallo_rep$ID, split = "_")))[,1]
  nonallo_rep$SampleID <- t(data.frame(strsplit(nonallo_rep$ID, split = "_")))[,2]
  nonallo_rep$Rep_div <-"nonallo"
  # Plotting cumulative frequency of all allo-reactive
  cum_freq_allo <- cum_freq(immdata_alloreactives, immdata2)
  avg_freq_allo <- avg_freq(immdata_alloreactives, immdata2)
  
  # Plotting clone fraction of all allo-reactive
  clone_fraction_allo <- clone_frac(immdata_alloreactives, immdata2)
  
    # Plotting clone fraction of all allo-reactive
  n_clone_allo <- n_clone(immdata_alloreactives, immdata2)
  
  #Cumulative frequencies of all non-alloreactives
  cum_freq_nonallo <- cum_freq(immdata_nonalloreactives, immdata2)
  
  avg_freq_nonallo <- avg_freq(immdata_nonalloreactives, immdata2)


  #Clone fraction of all non-alloreactives
  clone_fraction_nonallo <- clone_frac(immdata_nonalloreactives, immdata2)
  
  n_clone_nonallo <- n_clone(immdata_nonalloreactives, immdata2)
  
  #Cumulative frequences of all
  cum_freq_all <- cum_freq(immdata2, immdata2)
  avg_freq_all <- avg_freq(immdata2, immdata2)
  
  #Clone fraction of all 
  clone_fraction_all <- clone_frac(immdata2, immdata2)
  
  n_clone_all <- n_clone(immdata2, immdata2)

  
  allo_rep$clone_fraction <-clone_fraction_allo$V1
  allo_rep$cum_freq <-cum_freq_allo$V1
  allo_rep$avg_freq <-avg_freq_allo$V1
  allo_rep$n_clone <- n_clone_allo$V1
  nonallo_rep$clone_fraction <-clone_fraction_nonallo$V1
  nonallo_rep$cum_freq <-cum_freq_nonallo$V1
  nonallo_rep$avg_freq <-avg_freq_nonallo$V1
  nonallo_rep$n_clone <- n_clone_nonallo$V1
  all_rep$clone_fraction <-clone_fraction_all$V1
  all_rep$cum_freq <-cum_freq_all$V1
  all_rep$avg_freq <-avg_freq_all$V1
  all_rep$n_clone <- n_clone_allo$V1
  data_patient <- rbind(allo_rep, nonallo_rep, all_rep)
  allo_cd48 = full_allo
  sorted_CD48_allo <- allo_cd48[order(-allo_cd48$frequency.CFSELo),] 
  sorted_CD48_allo
  sorted_CD48_allo$index = as.numeric(rownames(sorted_CD48_allo))
  
  
  sample_list <-unique(data.frame(t(sample_meta_data))[,2])
  
  str_list = c("unstimulated","MLRCFSElo")
  for (j in sort(na.omit(data.frame(strtoi(sample_list)))[,1]))
  {
    str_list <-append(str_list, toString(j))
  }
  top_allo_48 <- allo_cd48
  
  
  
  fig_data <- bind_rows(immdata_alloreactives, .id = "id")
  fig_data <- add_tag(fig_data)
  fig_data <- convert_id(fig_data)
  fig_data$timepoints <- factor(fig_data$timepoints, levels=str_list)

  
  
  # plot figures
  allo_clone_fig <- ggplot(fig_data, aes(timepoints, frequency, group = tag, color = cell_type)) + #[fig_data$cell_type == "CD4",]
    geom_point(size = 1) + 
    geom_line() +
    scale_y_continuous(trans = "log10")+
    labs(x = curr_file, y = "Allo_Frequency") +
    theme_minimal()

  
  selected_data <- data_patient[which(data_patient['Rep_div'] =="allo"),]
  selected_data<- selected_data[selected_data$SampleID %in% str_list,]
  selected_data$SampleID <- factor(selected_data$SampleID, levels=str_list)
  selected_data_notnonallo <- data_patient#[which(data_patient['Rep_div'] !="all"),]
  selected_data_notnonallo<- selected_data_notnonallo[selected_data_notnonallo$SampleID %in% str_list,]
  selected_data_notnonallo$SampleID <- factor(selected_data_notnonallo$SampleID, levels=str_list)
  
  
    
  div_fig_1 <- div_fig(selected_data_notnonallo, 'clone_fraction', filenames[i])
  div_fig_2 <- div_fig(selected_data_notnonallo, 'gini.simpson', filenames[i])
  div_fig_3 <- div_fig(selected_data_notnonallo, 'shannons.entropy', filenames[i])
  div_fig_4 <- div_fig(selected_data_notnonallo, 'cum_freq', filenames[i])
  
  fig_data_top <- dplyr::inner_join(fig_data, data.frame(tag = data_CD48_all_expand$tag), by = c('tag'))
  
  fig_data$timepoints <- factor(fig_data$timepoints, levels=str_list)


  # plot figures
  top_clone_fig = ggplot(fig_data_top, aes(timepoints, frequency, group = tag, color = cell_type)) + #[fig_data_top$cell_type == "CD4",] 
    geom_point(size = 1) +
    geom_line() +
    scale_y_continuous(trans = "log10", ) +
    labs(x = curr_file, y = "Top_allo_Frequency") +
    theme_minimal()


  main_fig <-(grid.arrange(allo_clone_fig,top_clone_fig,div_fig_1,div_fig_2,div_fig_3,div_fig_4))#, length_over_time_plot,boxen_plot, boxen_plot_alp, ncol = 1))
  ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/MainFig_',curr_file,'.png'), plot = main_fig, width = 16, height = 10, dpi = 600)
  
  patient_data_all<- rbind(patient_data_all, data_patient)
}
```


```{r, fig.height = 5, fig.width = 7, dpi = 600 warning=FALSE}
data_squished_time <- data_squished %>% #Sum the frequency and the templates per timepoint+patient
  select(frequency, templates, timepoint, PatientID)%>%
  group_by(timepoint, PatientID)%>%
  dplyr::summarise(frequency = sum(frequency), templates = sum(templates), .groups = "drop")

data_squished_count <- data_squished %>% #get number of unique clones
  select(frequency, templates, timepoint, PatientID)%>%
  dplyr::group_by(timepoint, PatientID)%>%
  dplyr::count()
  # select(timepoint, PatientID)

data_squished_time$unique_clones <- data_squished_count$n

patient_data_all_2 <- patient_data_all #Add diversity data
colnames(patient_data_all_2)[grepl('SampleID',colnames(patient_data_all_2))] <- "timepoint"

data_squished_time <- dplyr::full_join(data_squished_time, patient_data_all_2,
               by = c("PatientID", "timepoint"))
data_squished_time <- filter(data_squished_time, Rep_div == "all")

data_squished_time <- dplyr::full_join(data_squished_time, meta_data_clin, by = "PatientID")%>%
  mutate(timepoint = strtoi(timepoint), GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1))%>%
  filter(!is.na(timepoint))


colnames(data_squished_time) <- sapply(colnames(data_squished_time), function(x){str_replace_all(x, "-", "")})
data_squished_time$amount <- 1
data_squished_time$GVHD_GRADE_1 <- factor(data_squished_time$GVHD_GRADE_1, levels = c("Severe", "Mild", "No"))


#   
data_squished_time$DonorType <- ifelse(data_squished_time$DonorType == "Matched Related Donor",
                                           "Matched Unrelated Donor",
                                           data_squished_time$DonorType)

data_squished_avg <- data_squished_time%>%
  group_by(PatientID)%>%
  summarise(avg_div = median(na.omit(shannons.diversity)))%>%
  ungroup()%>%
  arrange(by_group = avg_div)

data_squished_time <- data_squished_time %>%
  mutate(PatientID = factor(PatientID, levels = unique(data_squished_avg$PatientID)))%>%
  arrange(by_group = PatientID)

data_squished_meta <- data_squished_time%>%
  select(PatientID, PTCy, DonorType)%>%
  distinct()%>%
  arrange(by_group = PatientID)%>%
  mutate(amount = 1)

#this is for ptcy then grade
b2 <- ggplot(data_squished_meta, aes(y = amount,x = "PTCy", fill = PTCy, group = desc(PatientID)))+
  geom_bar(stat = 'identity', width = .1)+
  scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
  theme_void()+
  guides()+
  theme(legend.position=c(1.03,0.35), plot.margin=margin(-0.25,-2,0.75,-2, "cm"))+
  labs(fill = 'PTCy')

b3 <- ggplot(data_squished_meta, aes(y = amount,x = "Donor Type", fill = DonorType, group = desc(PatientID)))+
  geom_bar(stat = 'identity', width = .1)+
  scale_fill_brewer(palette = "Set3")+
  theme_void()+
  guides()+
  theme(legend.position=c(1.29,0.1), plot.margin=margin(-0.25,-2,0.75,-2, "cm"))+
  labs(fill = 'DonorType')


# data_squished_time$PatientID <- factor(data_squished_time$PatientID, levels = unique(data_squished_time$PatientID))

dot_plots_div <- ggplot(data_squished_time, aes(x = shannons.entropy, y = PatientID, color = GVHD_GRADE_1, alpha = timepoint)) +
  geom_point(fill = "white", color = "black", size = 4, pch = 21, alpha = 1) +
  geom_point(aes(fill = GVHD_GRADE_1, alpha = timepoint), color = "black", size = 4, pch = 21) +
  # scale_x_continuous(trans = "log10") +
  scale_fill_manual(values = c("Severe" = "#B85000", "Mild" = "#005A8F", "No" = "#09BB8C")) +
  scale_alpha_continuous(trans = "log2", name = "Days")+
  theme_minimal() +
  labs(title = "All", x = "Shannon's Diversity Index", fill = "GVHD Grade") +
  theme(legend.position = c(1.28, 0.74), legend.margin = margin(-.1,-.1,-.1,-.1, "cm"), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black"), axis.ticks = element_line(color = "black"),  axis.ticks.length = unit(0.25, "cm"))

dot_plots <- grid.arrange(dot_plots_div, b2 , b3, nrow = 1, widths = c(2,0.05,0.05,1))

plot(dot_plots)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/All_ShannonsEntropyDotPlots.png'), plot = dot_plots, width = 7, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/All_ShannonsEntropyDotPlots.pdf'), plot = dot_plots, width = 7, height = 5, dpi = 600)
```
```{r, fig.height = 5, fig.width = 7, dpi = 600 warning=FALSE}
data_squished_time <- data_squished %>% #Sum the frequency and the templates per timepoint+patient
  select(frequency, templates, timepoint, PatientID)%>%
  group_by(timepoint, PatientID)%>%
  dplyr::summarise(frequency = sum(frequency), templates = sum(templates), .groups = "drop")

data_squished_count <- data_squished %>% #get number of unique clones
  select(frequency, templates, timepoint, PatientID)%>%
  dplyr::group_by(timepoint, PatientID)%>%
  dplyr::count()
  # select(timepoint, PatientID)

data_squished_time$unique_clones <- data_squished_count$n

patient_data_all_2 <- patient_data_all #Add diversity data
colnames(patient_data_all_2)[grepl('SampleID',colnames(patient_data_all_2))] <- "timepoint"

data_squished_time <- dplyr::full_join(data_squished_time, patient_data_all_2,
               by = c("PatientID", "timepoint"))
data_squished_time <- filter(data_squished_time, Rep_div == "allo")

data_squished_time <- dplyr::full_join(data_squished_time, meta_data_clin, by = "PatientID")%>%
  mutate(timepoint = strtoi(timepoint), GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1))%>%
  filter(!is.na(timepoint))


colnames(data_squished_time) <- sapply(colnames(data_squished_time), function(x){str_replace_all(x, "-", "")})
data_squished_time$amount <- 1
data_squished_time$GVHD_GRADE_1 <- factor(data_squished_time$GVHD_GRADE_1, levels = c("Severe", "Mild", "No"))


#   
data_squished_time$DonorType <- ifelse(data_squished_time$DonorType == "Matched Related Donor",
                                           "Matched Unrelated Donor",
                                           data_squished_time$DonorType)

data_squished_avg <- data_squished_time%>%
  group_by(PatientID)%>%
  summarise(avg_div = median(na.omit(shannons.diversity)))%>%
  ungroup()%>%
  arrange(by_group = avg_div)

data_squished_time <- data_squished_time %>%
  mutate(PatientID = factor(PatientID, levels = unique(data_squished_avg$PatientID)))%>%
  arrange(by_group = PatientID)

data_squished_meta <- data_squished_time%>%
  select(PatientID, PTCy, DonorType)%>%
  distinct()%>%
  arrange(by_group = PatientID)%>%
  mutate(amount = 1)

#this is for ptcy then grade
b2 <- ggplot(data_squished_meta, aes(y = amount,x = "PTCy", fill = PTCy, group = desc(PatientID)))+
  geom_bar(stat = 'identity', width = .1)+
  scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
  theme_void()+
  guides()+
  theme(legend.position=c(1.03,0.35), plot.margin=margin(-0.25,-2,0.75,-2, "cm"))+
  labs(fill = 'PTCy')

b3 <- ggplot(data_squished_meta, aes(y = amount,x = "Donor Type", fill = DonorType, group = desc(PatientID)))+
  geom_bar(stat = 'identity', width = .1)+
  scale_fill_brewer(palette = "Set3")+
  theme_void()+
  guides()+
  theme(legend.position=c(1.29,0.1), plot.margin=margin(-0.25,-2,0.75,-2, "cm"))+
  labs(fill = 'DonorType')


# data_squished_time$PatientID <- factor(data_squished_time$PatientID, levels = unique(data_squished_time$PatientID))

dot_plots_div <- ggplot(data_squished_time, aes(x = shannons.entropy, y = PatientID, color = GVHD_GRADE_1, alpha = timepoint)) +
  geom_point(fill = "white", color = "black", size = 4, pch = 21, alpha = 1) +
  geom_point(aes(fill = GVHD_GRADE_1, alpha = timepoint), color = "black", size = 4, pch = 21) +
  # scale_x_continuous(trans = "log10") +
  scale_fill_manual(values = c("Severe" = "#B85000", "Mild" = "#005A8F", "No" = "#09BB8C")) +
  scale_alpha_continuous(trans = "log2", name = "Days")+
  theme_minimal() +
  labs(title = "Alloreactive", x = "Shannon's Diversity Index", fill = "GVHD Grade") +
  theme(legend.position = c(1.28, 0.74), legend.margin = margin(-.1,-.1,-.1,-.1, "cm"), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black"), axis.ticks = element_line(color = "black"),  axis.ticks.length = unit(0.25, "cm"))


dot_plots <- grid.arrange(dot_plots_div, b2 , b3, nrow = 1, widths = c(2,0.05,0.05,1))

plot(dot_plots)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/Allo_ShannonsEntropyDotPlots.png'), plot = dot_plots, width = 7, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/Allo_ShannonsEntropyDotPlots.png'), plot = dot_plots, width = 7, height = 5, dpi = 600)

```
```{r, fig.height = 5, fig.width = 7, dpi = 600 warning=FALSE}
data_squished_time <- data_squished %>% #Sum the frequency and the templates per timepoint+patient
  select(frequency, templates, timepoint, PatientID)%>%
  group_by(timepoint, PatientID)%>%
  dplyr::summarise(frequency = sum(frequency), templates = sum(templates), .groups = "drop")

data_squished_count <- data_squished %>% #get number of unique clones
  select(frequency, templates, timepoint, PatientID)%>%
  dplyr::group_by(timepoint, PatientID)%>%
  dplyr::count()
  # select(timepoint, PatientID)

data_squished_time$unique_clones <- data_squished_count$n

patient_data_all_2 <- patient_data_all #Add diversity data
colnames(patient_data_all_2)[grepl('SampleID',colnames(patient_data_all_2))] <- "timepoint"

data_squished_time <- dplyr::full_join(data_squished_time, patient_data_all_2,
               by = c("PatientID", "timepoint"))
data_squished_time <- filter(data_squished_time, Rep_div == "nonallo")

data_squished_time <- dplyr::full_join(data_squished_time, meta_data_clin, by = "PatientID")%>%
  mutate(timepoint = strtoi(timepoint), GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1))%>%
  filter(!is.na(timepoint))


colnames(data_squished_time) <- sapply(colnames(data_squished_time), function(x){str_replace_all(x, "-", "")})
data_squished_time$amount <- 1
data_squished_time$GVHD_GRADE_1 <- factor(data_squished_time$GVHD_GRADE_1, levels = c("Severe", "Mild", "No"))


#   
data_squished_time$DonorType <- ifelse(data_squished_time$DonorType == "Matched Related Donor",
                                           "Matched Unrelated Donor",
                                           data_squished_time$DonorType)

data_squished_avg <- data_squished_time%>%
  group_by(PatientID)%>%
  summarise(avg_div = median(na.omit(shannons.diversity)))%>%
  ungroup()%>%
  arrange(by_group = avg_div)

data_squished_time <- data_squished_time %>%
  mutate(PatientID = factor(PatientID, levels = unique(data_squished_avg$PatientID)))%>%
  arrange(by_group = PatientID)

data_squished_meta <- data_squished_time%>%
  select(PatientID, PTCy, DonorType)%>%
  distinct()%>%
  arrange(by_group = PatientID)%>%
  mutate(amount = 1)

#this is for ptcy then grade
b2 <- ggplot(data_squished_meta, aes(y = amount,x = "PTCy", fill = PTCy, group = desc(PatientID)))+
  geom_bar(stat = 'identity', width = .1)+
  scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
  theme_void()+
  guides()+
  theme(legend.position=c(1.03,0.35), plot.margin=margin(-0.25,-2,0.75,-2, "cm"))+
  labs(fill = 'PTCy')

b3 <- ggplot(data_squished_meta, aes(y = amount,x = "Donor Type", fill = DonorType, group = desc(PatientID)))+
  geom_bar(stat = 'identity', width = .1)+
  scale_fill_brewer(palette = "Set3")+
  theme_void()+
  guides()+
  theme(legend.position=c(1.29,0.1), plot.margin=margin(-0.25,-2,0.75,-2, "cm"))+
  labs(fill = 'DonorType')


dot_plots_div <- ggplot(data_squished_time, aes(x = shannons.entropy, y = PatientID, color = GVHD_GRADE_1, alpha = timepoint)) +
  geom_point(fill = "white", color = "black", size = 4, pch = 21, alpha = 1) +
  geom_point(aes(fill = GVHD_GRADE_1, alpha = timepoint), color = "black", size = 4, pch = 21) +
  scale_fill_manual(values = c("Severe" = "#B85000", "Mild" = "#005A8F", "No" = "#09BB8C")) +
  scale_alpha_continuous(trans = "log2", name = "Days")+
  theme_minimal() +
  labs(title = "Non-Alloreactive",x = "Shannon's Diversity Index", fill = "GVHD Grade") +
  theme(legend.position = c(1.28, 0.74), legend.margin = margin(-.1,-.1,-.1,-.1, "cm"), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black"), axis.ticks = element_line(color = "black"),  axis.ticks.length = unit(0.25, "cm"))

dot_plots <- grid.arrange(dot_plots_div, b2 , b3, nrow = 1, widths = c(2,0.05,0.05,1))

plot(dot_plots)

ggsave(paste0('//Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig//Figure1ex/Nonallo_ShannonsEntropyDotPlots.png'), plot = dot_plots, width = 7, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/Nonallo_ShannonsEntropyDotPlots.png'), plot = dot_plots, width = 7, height = 5, dpi = 600)

```

```{r, fig.height = 5, fig.width = 7, dpi = 600, warning=FALSE}
data_squished_time <- data_squished %>% #Sum the frequency and the templates per timepoint+patient
  select(frequency, templates, timepoint, PatientID)%>%
  group_by(timepoint, PatientID)%>%
  dplyr::summarise(frequency = sum(frequency), templates = sum(templates), .groups = "drop")

data_squished_count <- data_squished %>% #get number of unique clones
  select(frequency, templates, timepoint, PatientID)%>%
  dplyr::group_by(timepoint, PatientID)%>%
  dplyr::count()
  # select(timepoint, PatientID)

data_squished_time$unique_clones <- data_squished_count$n

patient_data_all_2 <- patient_data_all #Add diversity data
colnames(patient_data_all_2)[grepl('SampleID',colnames(patient_data_all_2))] <- "timepoint"

data_squished_time <- dplyr::full_join(data_squished_time, patient_data_all_2,
               by = c("PatientID", "timepoint"))
data_squished_time <- filter(data_squished_time, Rep_div == "allo")

data_squished_time <- dplyr::full_join(data_squished_time, meta_data_clin, by = "PatientID")%>%
  mutate(timepoint = strtoi(timepoint), GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1))%>%
  filter(!is.na(timepoint))


colnames(data_squished_time) <- sapply(colnames(data_squished_time), function(x){str_replace_all(x, "-", "")})
data_squished_time$amount <- 1
data_squished_time$GVHD_GRADE_1 <- factor(data_squished_time$GVHD_GRADE_1, levels = c("Severe", "Mild", "No"))


#   
data_squished_time$DonorType <- ifelse(data_squished_time$DonorType == "Matched Related Donor",
                                           "Matched Unrelated Donor",
                                           data_squished_time$DonorType)

data_squished_avg <- data_squished_time%>%
  group_by(PatientID)%>%
  summarise(avg_div = median(na.omit(cum_freq)))%>%
  ungroup()%>%
  arrange(by_group = avg_div)

data_squished_time <- data_squished_time %>%
  mutate(PatientID = factor(PatientID, levels = unique(data_squished_avg$PatientID)), timepoint = ifelse(timepoint > 365, 365, timepoint),
         Chronic = ifelse(Chronic, "Yes", "No"))%>%
  arrange(by_group = PatientID)

data_squished_meta <- data_squished_time%>%
  select(PatientID, PTCy, DonorType, Chronic)%>%
  distinct()%>%
  arrange(by_group = PatientID)%>%
  mutate(amount = 1)

#this is for ptcy then grade
b2 <- ggplot(data_squished_meta, aes(y = amount,x = "PTCy", fill = PTCy, group = desc(PatientID)))+
  geom_bar(stat = 'identity', width = .1)+
  scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
  theme_void()+
  guides()+
  theme(legend.position=c(1.03,0.35), plot.margin=margin(0.35,-2,0.8,-2, "cm"))+
  labs(fill = 'PTCy')

b3 <- ggplot(data_squished_meta, aes(y = amount,x = "Donor Type", fill = DonorType, group = desc(PatientID)))+
  geom_bar(stat = 'identity', width = .1)+
  scale_fill_brewer(palette = "Set3")+
  theme_void()+
  guides()+
  theme(legend.position=c(1.29,0.1), plot.margin=margin(0.35,-2,0.8,-2, "cm"))+
  labs(fill = 'DonorType')

b4 <- ggplot(data_squished_meta, aes(y = amount,x = "Chronic ", fill = Chronic, group = desc(PatientID)))+
  geom_bar(stat = 'identity', width = .1)+
  scale_fill_manual(values = c("#48A4E3", "#BF6196"))+
  theme_void()+
  guides()+
  theme(legend.position=c(1.45,0.35), plot.margin=margin(0.35,-2,0.8,-2, "cm"))+
  labs(fill = 'Chronic')

dot_plots_div <- ggplot(data_squished_time, aes(x = cum_freq, y = PatientID, color = GVHD_GRADE_1, alpha = timepoint)) +
  geom_point(fill = "white", color = "black", size = 4, pch = 21, alpha = 1) +
  geom_point(aes(fill = GVHD_GRADE_1, alpha = timepoint), color = "black", size = 4, pch = 21) +
  scale_color_gradient(low = "white", high = "black") +
  scale_x_continuous(trans = "log10", labels = scales::number_format(accuracy = 0.00001)) +
  scale_fill_manual(values = c("Severe" = "#B85000", "Mild" = "#005A8F", "No" = "#09BB8C")) +
  # scale_alpha_continuous(trans = "log10")+
  guides(alpha = FALSE) +
  theme_minimal() +
  labs(title = "Alloreactive T Cell Clones", x = "Cumulative Frequency", fill = "GVHD Grade", alpha = "Days") +
  theme(legend.position = c(1.28, 0.59), legend.margin = margin(-.1,-.1,-.1,-.1, "cm"), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black"), axis.ticks = element_line(color = "black"),  axis.ticks.length = unit(0.25, "cm"))

dummy_plot <- ggplot(data_squished_time, aes(x = 1, y = timepoint, fill = timepoint)) +
  geom_tile() +
  scale_fill_gradient( low = "white", high = "black",
                      breaks = c(100,200, 300, 365), # Specify your desired breaks
                         labels = c("100","200","300", "365+")) +
  theme_void()+
  labs(fill = "Days")
legend_timeplots <- plot_grid(plotlist = list(NULL, cowplot::get_legend(dummy_plot), NULL, NULL), nrow = 1, rel_widths = c(0.0001, 1), rel_heights = c(.1,1))

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/legendtimeplot.png'), plot = legend_timeplots, width = 7, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/legendtimeplot.pdf'), plot = legend_timeplots, width = 7, height = 5, dpi = 600)

dot_plots <- grid.arrange(dot_plots_div, b2 , b3, b4, nrow = 1, widths = c(2,0.05,0.05,0.05, 0.05,1))

plot(legend_timeplots)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/AlloFrequnceyDotPlot.png'), plot = dot_plots, width = 7, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/AlloFrequnceyDotPlot.pdf'), plot = dot_plots, width = 7, height = 5, dpi = 600)
```
```{r}
##### Analysis for patients with all frequency below 0.001

low_freq_patients <- data_squished_time %>%
  dplyr::group_by(PatientID) %>%
  summarize(all_low_freq = all(cum_freq < 0.001)) 
  #filter(all_low_freq == TRUE) %>%
  #pull(PatientID)

patient_grades <- data_squished_time %>%
  select(PatientID, GVHD_GRADE_1) %>%
  distinct()

# Now join this with our low_freq_patients table
low_freq_patients_with_grade <- low_freq_patients %>%
  left_join(patient_grades, by = "PatientID")


# Create a contingency table
contingency_table <- table(low_freq_patients_with_grade$all_low_freq, 
                           low_freq_patients_with_grade$GVHD_GRADE_1 == "No")

# Run Fisher's exact test
fisher_test <- fisher.test(contingency_table)
print(fisher_test)
```

```{r, fig.height = 5, fig.width = 7, dpi = 600 warning=FALSE}
data_squished_time <- data_squished %>% #Sum the frequency and the templates per timepoint+patient
  select(frequency, templates, timepoint, PatientID)%>%
  group_by(timepoint, PatientID)%>%
  dplyr::summarise(frequency = sum(frequency), templates = sum(templates), .groups = "drop")

data_squished_count <- data_squished %>% #get number of unique clones
  select(frequency, templates, timepoint, PatientID)%>%
  dplyr::group_by(timepoint, PatientID)%>%
  dplyr::count()
  # select(timepoint, PatientID)

data_squished_time$unique_clones <- data_squished_count$n

patient_data_all_2 <- patient_data_all #Add diversity data
colnames(patient_data_all_2)[grepl('SampleID',colnames(patient_data_all_2))] <- "timepoint"

data_squished_time <- dplyr::full_join(data_squished_time, patient_data_all_2,
               by = c("PatientID", "timepoint"))
data_squished_time <- filter(data_squished_time, Rep_div == "nonallo")

data_squished_time <- dplyr::full_join(data_squished_time, meta_data_clin, by = "PatientID")%>%
  mutate(timepoint = strtoi(timepoint), GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1))%>%
  filter(!is.na(timepoint))


colnames(data_squished_time) <- sapply(colnames(data_squished_time), function(x){str_replace_all(x, "-", "")})
data_squished_time$amount <- 1
data_squished_time$GVHD_GRADE_1 <- factor(data_squished_time$GVHD_GRADE_1, levels = c("Severe", "Mild", "No"))



data_squished_avg <- data_squished_time%>%
  group_by(PatientID)%>%
  summarise(avg_div = median(na.omit(cum_freq)))%>%
  ungroup()%>%
  arrange(by_group = avg_div)

data_squished_time <- data_squished_time %>%
  mutate(PatientID = factor(PatientID, levels = unique(data_squished_avg$PatientID)))%>%
  arrange(by_group = PatientID)

data_squished_meta <- data_squished_time%>%
  select(PatientID, PTCy, DonorType, Chronic)%>%
  distinct()%>%
  arrange(by_group = PatientID)%>%
  mutate(amount = 1, Chronic = ifelse(Chronic, "Yes", "No"))

#this is for ptcy then grade
b2 <- ggplot(data_squished_meta, aes(y = amount,x = "PTCy", fill = PTCy, group = desc(PatientID)))+
  geom_bar(stat = 'identity', width = .1)+
  scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
  theme_void()+
  guides()+
  theme(legend.position=c(1.03,0.35), plot.margin=margin(-0.25,-2,0.75,-2, "cm"))+
  labs(fill = 'PTCy')

b3 <- ggplot(data_squished_meta, aes(y = amount,x = "Donor Type", fill = DonorType, group = desc(PatientID)))+
  geom_bar(stat = 'identity', width = .1)+
  scale_fill_brewer(palette = "Set3")+
  theme_void()+
  guides()+
  theme(legend.position=c(1.29,0.1), plot.margin=margin(-0.25,-2,0.75,-2, "cm"))+
  labs(fill = 'DonorType')

b4 <- ggplot(data_squished_meta, aes(y = amount,x = "Chronic GVHD", fill = Chronic, group = desc(PatientID)))+
  geom_bar(stat = 'identity', width = .1)+
  scale_fill_manual(values = c("#48A4E3", "#BF6196"))+
  theme_void()+
  guides()+
  theme(legend.position=c(1.59,0.3), plot.margin=margin(-0.25,-2,0.75,-2, "cm"))+
  labs(fill = 'Chronic GVHD')


# data_squished_time$PatientID <- factor(data_squished_time$PatientID, levels = unique(data_squished_time$PatientID))

dot_plots_div <- ggplot(data_squished_time, aes(x = cum_freq, y = PatientID, color = GVHD_GRADE_1, alpha = timepoint)) +
  geom_point(fill = "white", color = "black", size = 4, pch = 21, alpha = 1) +
  geom_point(aes(fill = GVHD_GRADE_1, alpha = timepoint), color = "black", size = 4, pch = 21) +
  scale_x_continuous(trans = "log10") +
  scale_fill_manual(values = c("Severe" = "#B85000", "Mild" = "#005A8F", "No" = "#09BB8C")) +
  scale_alpha_continuous(trans = "log2", name = "Days")+
  theme_minimal() +
  labs(x = "Cumulative Frequency", fill = "GVHD Grade") +
  theme(legend.position = c(1.28, 0.74), legend.margin = margin(-.1,-.1,-.1,-.1, "cm"), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black"), axis.ticks = element_line(color = "black"),  axis.ticks.length = unit(0.25, "cm"))

# dummy_plot <- ggplot(data_squished_time, aes(x = 1, y = timepoint, fill = timepoint)) +
#   geom_tile() +
#   scale_fill_gradient(low = "white", high = "black") +
#   theme_void() +
#   theme(legend.position = c(5.5, 0.25))+
#   labs(fill = "Timepoints")
# legend_timeplots <- cowplot::get_legend(dummy_plot)

dot_plots <- grid.arrange(dot_plots_div, b2 , b3, b4, nrow = 1, widths = c(2,0.05,0.05,0.05,1))


ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/NonalloFrequnceyDotPlot.png'), plot = dot_plots, width = 7, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/NonalloFrequnceyDotPlot.pdf'), plot = dot_plots, width = 7, height = 5, dpi = 600)
```


```{r}
#https://stackoverflow.com/questions/47651868/split-violin-plot-with-ggplot2-with-quantiles
GeomSplitViolin <- ggproto("GeomSplitViolin", GeomViolin,
  draw_group = function(self, data, ..., draw_quantiles = NULL) {
    # Original function by Jan Gleixner (@jan-glx)
    # Adjustments by Wouter van der Bijl (@Axeman)
    data <- transform(data, xminv = x - violinwidth * (x - xmin), xmaxv = x + violinwidth * (xmax - x))
    grp <- data[1, "group"]
    newdata <- plyr::arrange(transform(data, x = if (grp %% 2 == 1) xminv else xmaxv), if (grp %% 2 == 1) y else -y)
    newdata <- rbind(newdata[1, ], newdata, newdata[nrow(newdata), ], newdata[1, ])
    newdata[c(1, nrow(newdata) - 1, nrow(newdata)), "x"] <- round(newdata[1, "x"])
    if (length(draw_quantiles) > 0 & !scales::zero_range(range(data$y))) {
      stopifnot(all(draw_quantiles >= 0), all(draw_quantiles <= 1))
      quantiles <- create_quantile_segment_frame(data, draw_quantiles, split = TRUE, grp = grp)
      aesthetics <- data[rep(1, nrow(quantiles)), setdiff(names(data), c("x", "y")), drop = FALSE]
      aesthetics$alpha <- rep(1, nrow(quantiles))
      both <- cbind(quantiles, aesthetics)
      quantile_grob <- GeomPath$draw_panel(both, ...)
      ggplot2:::ggname("geom_split_violin", grid::grobTree(GeomPolygon$draw_panel(newdata, ...), quantile_grob))
    }
    else {
      ggplot2:::ggname("geom_split_violin", GeomPolygon$draw_panel(newdata, ...))
    }
  }
)

create_quantile_segment_frame <- function(data, draw_quantiles, split = FALSE, grp = NULL) {
  dens <- cumsum(data$density) / sum(data$density)
  ecdf <- stats::approxfun(dens, data$y)
  ys <- ecdf(draw_quantiles)
  violin.xminvs <- (stats::approxfun(data$y, data$xminv))(ys)
  violin.xmaxvs <- (stats::approxfun(data$y, data$xmaxv))(ys)
  violin.xs <- (stats::approxfun(data$y, data$x))(ys)
  if (grp %% 2 == 0) {
    data.frame(
      x = ggplot2:::interleave(violin.xs, violin.xmaxvs),
      y = rep(ys, each = 2), group = rep(ys, each = 2)
    )
  } else {
    data.frame(
      x = ggplot2:::interleave(violin.xminvs, violin.xs),
      y = rep(ys, each = 2), group = rep(ys, each = 2)
    )
  }
}

geom_split_violin <- function(mapping = NULL, data = NULL, stat = "ydensity", position = "identity", ..., 
                              draw_quantiles = NULL, trim = TRUE, scale = "area", na.rm = FALSE, 
                              show.legend = NA, inherit.aes = TRUE) {
  layer(data = data, mapping = mapping, stat = stat, geom = GeomSplitViolin, position = position, 
        show.legend = show.legend, inherit.aes = inherit.aes, 
        params = list(trim = trim, scale = scale, draw_quantiles = draw_quantiles, na.rm = na.rm, ...))
}

```


```{r}


num_clone_fig <- function(immdata, type, title = "", num_cuts = 5, groupings = NA, colors = c("#E6BE60", "#A483AF"), method = "hochberg", legend_title = "", ylab = ""){
  #' Visualize the number of clones over time
  #'
  #' @description This function calculates and visualizes the number of clones over time, grouping by specified categories and applying statistical tests to compare groups.
  #'
  #' @param immdata A dataframe containing immunosequencing data with `timepoint`, `PatientID`, and specified `type`.
  #' @param type A character string specifying the grouping variable for the analysis.
  #' @param title A character string for the plot title.
  #' @param num_cuts An integer indicating the number of bins to divide the timepoints into (default is 5).
  #' @param groupings Optional grouping variable for pairwise comparisons (default is NA).
  #' @param colors A character vector specifying colors for the plot (default is c("#E6BE60", "#A483AF")).
  #' @param method A character string specifying the p-value adjustment method (default is "hochberg").
  #' @param legend_title A character string for the legend title.
  #' @param ylab A character string for the y-axis label in the plot.
  #' @usage num_clone_fig(immdata, type, title, num_cuts, groupings, colors, method, legend_title, ylab)
  #' @return A list containing the plots, statistical test results, and processed data.
  
  immdata$timepoint <- as.character(immdata$timepoint) #Ensure timepoints are characters
  
  immdata <- immdata%>%
    mutate(timepoint = strtoi(timepoint)) #Convert back to integers
  immdata <- immdata[!is.na(immdata$timepoint),] #remove any NA timepoints
  immdata <- immdata[immdata$timepoint <= 365,]
  immdata_3 <- immdata[immdata$timepoint <= 3,]%>%
    mutate(timepoint = 3, bins = "3")%>%
    dplyr::filter(templates != 0)
  immdata <- immdata[immdata$timepoint > 3,]
  immdata_all_timepoint <- immdata %>% #This is to keep track of timepoints that do not have any expanded clones
    select(all_of(c("timepoint","PatientID", type)))%>%
    distinct()
  
  immdata <- immdata[immdata$templates !=0, ]
  
  breaks <- unique(quantile(immdata_all_timepoint$timepoint, probs = 0:num_cuts/num_cuts, na.rm = T)) #Breaks based of all timepoints recorded
  print(breaks)
  
  immdata <- immdata%>%
    mutate(bins = cut(immdata$timepoint, breaks = breaks, include.lowest = TRUE))
  
  
  immdata_all_timepoint <- immdata_all_timepoint %>%
    mutate(bins = cut(immdata_all_timepoint$timepoint, breaks = breaks, include.lowest = TRUE))
  
  if(length(immdata_3$timepoint) != 0){
    immdata <- bind_rows(immdata_3, immdata)%>%
      mutate(bins = factor(bins, level = c("3", levels(immdata_all_timepoint$bins))))
  }else{
    immdata <- immdata%>%
      mutate(bins = factor(bins, level = levels(immdata_all_timepoint$bins)))
  }
  
  immdata_fig <- immdata%>% 
    group_by(across(all_of(c("bins","timepoint","PatientID", type))))%>%
    tally()%>%
    ungroup()%>%
    dplyr::full_join(immdata_all_timepoint, by =c("PatientID", "bins","timepoint", type), multiple = "first")%>%
    replace_na(list(n = 0))%>%
    group_by(across(all_of(c("bins", type))))
  
  immdata_sum <- immdata_fig%>% ungroup()%>%
    group_by(across(all_of(c("bins", type))))%>%
    dplyr::summarise(m = mean(n), s = sqrt(var(n)))
  

  stat_test <- immdata_fig %>%
    group_by(bins) %>%
    mutate(n_in = n())%>%
    filter(n_in >= 3, sum(n != 0) > 0)%>% #dont do statistical test on groups that dont have enough samples
    pairwise_wilcox_test(as.formula(paste0(c("n ~ ",type))), p.adjust.method = "hochberg", exact = F, detailed = T)%>%
    add_xy_position(x = "bins")%>%
    ungroup()%>%
    mutate(
      y.position = log10(y.position+1),
      p.adj = p.adjust(p, method = "hochberg"), 
      p.adj.signif = symnum(p.adj,cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1), 
                            symbols = c("****", "***", "**", "*", "ns"))
      )
  
  
  fig <- ggplot(immdata_sum, aes(x = bins, y = m, fill = .data[[type]], group = .data[[type]], color = .data[[type]])) +
    geom_line()+
    geom_point()+
    geom_ribbon(aes(ymin = m - s, ymax = m + s, linetype = NA), alpha = 0.3)+
    geom_errorbar(aes(ymin = m - s, ymax = m + s), width = 0.1)+
    scale_x_discrete(na.translate = F)+
    theme_minimal()+
    theme(axis.text = element_text(size = 18),
          axis.text.x = element_text(angle = 75, vjust = 0.5), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank(), 
          axis.line = element_line(color = "black"), 
          axis.ticks = element_line(color = "black"),  
          axis.ticks.length = unit(0.25, "cm"))+
    labs(title = title, y = ylab, x = "Days")+
    scale_fill_manual(values=colors)+
    scale_color_manual(values=colors)
  
  fig_vio <- ggboxplot(immdata_fig, x = "bins", y = "n", fill = type, outlier.shape = NA)+
    geom_beeswarm(data = immdata_fig[immdata_fig$n != 0,], dodge.width = 0.75, cex = 0.1, corral.width = 0.2, method = "compactswarm", aes(x= bins, y = n, group = .data[[type]]))+
    stat_pvalue_manual(stat_test, label = "p.adj.signif", tip.length = 0.01, step.increase = 0.00, size = 8, bracket.nudge.y = -0.15)+
    scale_x_discrete(na.translate = F)+
    scale_y_continuous(trans = "log10", labels = scales::number_format(accuracy = 1))+
    theme_pubr()+
    theme(text = element_text(size = 20),
          axis.text.x = element_text(angle = 45, vjust = 0.5), 
          panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank(), 
          axis.line = element_line(color = "black"), 
          axis.ticks = element_line(color = "black"),  
          axis.ticks.length = unit(0.25, "cm"), 
          legend.text = element_markdown())+
    labs(title = title, y = ylab, x = "Days Post Transplant")+
    scale_fill_manual(values=colors)+
    scale_color_manual(values=colors)
  
  return(list(fig, fig_vio, stat_test, immdata_fig, immdata_sum))
}

```


```{r, fig.height=4,fig.width = 10, warning = FALSE}
remove_pat_regx <- ""#"067|070"

data_CD48_all_expand_adj <- dplyr::full_join(data_CD48_all, clinical_data, by = 'PatientID')
data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!is.na(data_CD48_all_expand_adj$cdr3_amino_acid),]
data_CD48_all_expand_adj <- data_CD48_all_expand_adj%>%
  mutate(PTCy = factor(PTCy, c("No PTCy", "PTCy")))%>%
  filter(PatientID != "R070")

if(remove_pat_regx != ""){
  data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!grepl(remove_pat_regx,data_CD48_all_expand_adj$PatientID),]
}

data_CD48_all_expand_noPTCy <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PTCy == "No PTCy", ]
data_CD48_all_expand_PTCy <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PTCy == "PTCy",]

Both_results <- num_clone_fig(data_CD48_all_expand_adj, "PTCy", "PTCy vs No PTCy", num_cuts = 4, legend_title = "Treatment", ylab = "# of Clones")
freq_Both <- Both_results[[1]]
freq_Both_vio <- Both_results[[2]]
Both_results_stat <- Both_results[[3]]


data_CD48_all_expand_CD4 <- subset(data_CD48_all_expand_adj, grepl("4", data_CD48_all_expand_adj$cell_type) & !(grepl("4", data_CD48_all_expand_adj$cell_type) & grepl("8", data_CD48_all_expand_adj$cell_type)))
CD4_results <- num_clone_fig(data_CD48_all_expand_CD4, "PTCy", "CD4+", num_cuts = 4, legend_title = "Treatment", ylab = "# of Clones")
freq_CD4 <- CD4_results[[1]]
freq_CD4_vio <- CD4_results[[2]]
CD4_results_stat <- CD4_results[[3]]


data_CD48_all_expand_CD8 <- subset(data_CD48_all_expand_adj, grepl("8", data_CD48_all_expand_adj$cell_type) & !(grepl("4", data_CD48_all_expand_adj$cell_type) & grepl("8", data_CD48_all_expand_adj$cell_type)))
CD8_results <- num_clone_fig(data_CD48_all_expand_CD8, "PTCy", "CD8+", num_cuts = 4, legend_title = "Treatment", ylab = "# of Clones")
freq_CD8 <- CD8_results[[1]]
freq_CD8_vio <- CD8_results[[2]]
CD8_results_stat <- CD8_results[[3]]

plot(freq_Both)
plot(freq_Both_vio)
plot(freq_CD4)
plot(freq_CD4_vio)
plot(freq_CD8)
plot(freq_CD8_vio)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_Both.png'), freq_Both, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD4.png'), freq_CD4, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD8.png'), freq_CD8, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_Both_vio.png'), freq_Both_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD4_vio.png'), freq_CD4_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD8_vio.png'), freq_CD8_vio, width = 8.5, height = 4, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_Both.pdf'), freq_Both, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD4.pdf'), freq_CD4, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_Both_vio.pdf'), freq_Both_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD4_vio.pdf'), freq_CD4_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD8_vio.pdf'), freq_CD8_vio, width = 8.5, height = 4, dpi = 600)


```

```{r, fig.height=4,fig.width = 10, warning = FALSE, dpi=600}

data_CD48_all_expand_adj <- dplyr::full_join(data_CD48_all_expand, clinical_data, by = 'PatientID')
data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!is.na(data_CD48_all_expand_adj$cdr3_amino_acid),]
data_CD48_all_expand_adj <- data_CD48_all_expand_adj%>%
  mutate(PTCy = factor(PTCy, c("No PTCy", "PTCy")))%>%
  filter(PatientID != "R070")


if(remove_pat_regx != ""){
  data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!grepl(remove_pat_regx,data_CD48_all_expand_adj$PatientID),]
}

data_CD48_all_expand_noPTCy <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PTCy == "No PTCy",]
data_CD48_all_expand_PTCy <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PTCy == "PTCy",]



Both_results <- num_clone_fig(data_CD48_all_expand_adj, "PTCy", "PTCy vs No PTCy", num_cuts = 4, colors = c("#E6BE60", "#A483AF") ,ylab = "# of Expanded \n Alloreactive Clones" , legend_title = "Treatment")
freq_Both <- Both_results[[1]]
freq_Both_vio <- Both_results[[2]]+
  scale_fill_manual(values = c("#E6BE60", "#A483AF"),
                    labels = c(
      `No PTCy` = "<span style='color:#E6BE60;'>No PTCy</span>",
      PTCy = "<span style='color:#A483AF;'>PTCy</span>"))
Both_results_p <- Both_results[[3]]



data_CD48_all_expand_CD4 <- subset(data_CD48_all_expand_adj, grepl("4", data_CD48_all_expand_adj$cell_type) & !(grepl("4", data_CD48_all_expand_adj$cell_type) & grepl("8", data_CD48_all_expand_adj$cell_type)))
CD4_results <- num_clone_fig(data_CD48_all_expand_CD4, "PTCy", "Expanded CD4+", num_cuts = 3, colors = c("#E6BE60", "#A483AF"), legend_title = "Treatment")
freq_CD4 <- CD4_results[[1]]
freq_CD4_vio <- CD4_results[[2]]+
  scale_fill_manual(values = c("#E6BE60", "#A483AF"),
                    labels = c(
      `No PTCy` = "<span style='color:#E6BE60;'>No PTCy</span>",
      PTCy = "<span style='color:#A483AF;'>PTCy</span>"))
CD4_results_p <- CD4_results[[3]]

data_CD48_all_expand_CD8 <- subset(data_CD48_all_expand_adj, grepl("8", data_CD48_all_expand_adj$cell_type) & !(grepl("4", data_CD48_all_expand_adj$cell_type) & grepl("8", data_CD48_all_expand_adj$cell_type)))
CD8_results <- num_clone_fig(data_CD48_all_expand_CD8, "PTCy", "Expanded CD8+", num_cuts = 3, colors = c("#E6BE60", "#A483AF"), legend_title = "Treatment")
freq_CD8 <- CD8_results[[1]]
freq_CD8_vio <- CD8_results[[2]]+
  scale_fill_manual(values = c("#E6BE60", "#A483AF"),
                    labels = c(
      `No PTCy` = "<span style='color:#E6BE60;'>No PTCy</span>",
      PTCy = "<span style='color:#A483AF;'>PTCy</span>"))
CD8_results_p <- CD8_results[[3]]

plot(freq_Both)
plot(freq_Both_vio)
plot(freq_CD4)
plot(freq_CD4_vio)
plot(freq_CD8)
plot(freq_CD8_vio)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/freq_Both_ex.png'), freq_Both, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD4_ex.png'), freq_CD4, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD8_ex.png'), freq_CD8, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_Both_vio_ex.png'), freq_Both_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD4_vio_ex.png'), freq_CD4_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD8_vio_ex.png'), freq_CD8_vio, width = 8.5, height = 4, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_Both_ex.pdf'), freq_Both, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD4_ex.pdf'), freq_CD4, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD8_ex.pdf'), freq_CD8, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/freq_Both_vio_ex.pdf'), freq_Both_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD4_vio_ex.pdf'), freq_CD4_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/freq_CD8_vio_ex.pdf'), freq_CD8_vio, width = 8.5, height = 4, dpi = 600)
```


```{r, fig.height=5,fig.width = 9.5, warning = FALSE}
#Look at this again and add 0 when expanded are not there.
data_CD48_all_expand_adj <- dplyr::full_join(data_CD48_all_expand, clinical_data, by = 'PatientID')
data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!is.na(data_CD48_all_expand_adj$cdr3_amino_acid),]

type <- "GVHD_GRADE_1"

if(remove_pat_regx != ""){
  data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!grepl(remove_pat_regx,data_CD48_all_expand_adj$PatientID),]
}

data_CD48_all_expand_adj <- data_CD48_all_expand_adj%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", "No and Mild"))%>%
  filter(timepoint != "3")
data_CD48_all_expand_adj$GVHD_GRADE_1 <- factor(data_CD48_all_expand_adj$GVHD_GRADE_1, c("No and Mild", "Severe") )

data_CD48_all_expand_noPTCy <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PTCy == "No PTCy",]
data_CD48_all_expand_PTCy <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PTCy == "PTCy",]

Both_results <- num_clone_fig(data_CD48_all_expand_adj, type, "", num_cuts = 4,
                              colors = c("#058b8e", "#B85000", "grey"),
                              ylab = "# of Alloreactive \n Expanded Clones") #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_Both <- Both_results[[1]]
freq_Both_vio_ex <- Both_results[[2]]+
  labs(fill = "GVHD Grade")+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))

data_CD48_all_expand_adj_CD4 <- data_CD48_all_expand_adj%>%
  filter(grepl("CD4",cell_type) & !grepl("CD8", cell_type))

data_CD48_all_expand_adj_CD8 <- data_CD48_all_expand_adj%>%
  filter(!grepl("CD4",cell_type) & grepl("CD8", cell_type))

CD4_results <- num_clone_fig(data_CD48_all_expand_adj_CD4, 
                         type, "GVHD Grade CD4+", num_cuts = 2,
                         colors = c("#058b8e", "#B85000", "grey"),
                         ylab = "# of Alloreactive \n Expanded Clones")

freq_cd4_vio_ex <- CD4_results[[2]]+
  labs(fill = "GVHD Grade")+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))

CD8_results <- num_clone_fig(data_CD48_all_expand_adj_CD8, 
                         type, "GVHD Grade CD8+", num_cuts = 2,
                         colors = c("#058b8e", "#B85000", "grey"),
                         ylab = "# of Alloreactive \n Expanded Clones")

freq_cd8_vio_ex <- CD8_results[[2]]+
  labs(fill = "GVHD Grade")+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))


Both_results_p <- Both_results[[3]]

PTCy_Both_results <- num_clone_fig(data_CD48_all_expand_PTCy, type, "Expanded PTCy", num_cuts = 4, colors = c("#058b8e", "#B85000", "grey"), ylab = "# of Alloreactive Expanded Clones") #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_PTCy_Both <- PTCy_Both_results[[1]]
freq_PTCy_Both_vio <- PTCy_Both_results[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
PTCy_Both_results_p <- PTCy_Both_results[[3]]

noPTCy_Both_results <- num_clone_fig(data_CD48_all_expand_noPTCy, type, "Expanded No PTCy", num_cuts = 4, colors = c("#058b8e", "#B85000", "grey"), ylab = "# of Alloreactive Expanded Clones") #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_noPTCy_Both <- noPTCy_Both_results[[1]]
freq_noPTCy_Both_vio <- noPTCy_Both_results[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
noPTCy_Both_results_p <- noPTCy_Both_results[[3]]

data_CD48_all_expand_PTCy_CD4 <- subset(data_CD48_all_expand_PTCy, grepl("4", data_CD48_all_expand_PTCy$cell_type) & !(grepl("4", data_CD48_all_expand_PTCy$cell_type) & grepl("8", data_CD48_all_expand_PTCy$cell_type)))

PTCy_CD4_results <- num_clone_fig(data_CD48_all_expand_PTCy_CD4, type, "Expanded PTCy CD4+", num_cuts = 3, colors = c("#058b8e", "#B85000", "grey")) #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_PTCy_CD4 <- PTCy_CD4_results[[1]]
freq_PTCy_CD4_vio <- PTCy_CD4_results[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
PTCy_CD4_results_p <- PTCy_CD4_results[[3]]

data_CD48_all_expand_PTCy_CD8 <- subset(data_CD48_all_expand_PTCy, grepl("8", data_CD48_all_expand_PTCy$cell_type) & !(grepl("4", data_CD48_all_expand_PTCy$cell_type) & grepl("8", data_CD48_all_expand_PTCy$cell_type)))

PTCy_CD8_results <- num_clone_fig(data_CD48_all_expand_PTCy_CD8, type, "Expanded PTCy CD8+", num_cuts = 3, colors = c("#058b8e", "#B85000", "grey")) #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_PTCy_CD8 <- PTCy_CD8_results[[1]]
freq_PTCy_CD8_vio <- PTCy_CD8_results[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
PTCy_CD8_results_p <- PTCy_CD8_results[[3]]

data_CD48_all_expand_noPTCy_CD4 <- subset(data_CD48_all_expand_noPTCy, grepl("4", data_CD48_all_expand_noPTCy$cell_type) & !(grepl("4", data_CD48_all_expand_noPTCy$cell_type) & grepl("8", data_CD48_all_expand_noPTCy$cell_type)))

NoPTCy_CD4_results <- num_clone_fig(data_CD48_all_expand_noPTCy_CD4, type, "Expanded No PTCy CD4+", num_cuts = 3, colors = c("#058b8e", "#B85000", "grey")) #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_noPTCy_CD4 <- NoPTCy_CD4_results[[1]]
freq_noPTCy_CD4_vio <- NoPTCy_CD4_results[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
NoPTCy_CD4_results_p <- NoPTCy_CD4_results[[3]]


data_CD48_all_expand_noPTCy_CD8 <- subset(data_CD48_all_expand_noPTCy, grepl("8", data_CD48_all_expand_noPTCy$cell_type) & !(grepl("4", data_CD48_all_expand_noPTCy$cell_type) & grepl("8", data_CD48_all_expand_noPTCy$cell_type)))

NoPTCy_CD8_results <- num_clone_fig(data_CD48_all_expand_noPTCy_CD8, type, "Expanded No PTCy CD8+", num_cuts = 3, colors = c("#058b8e", "#B85000", "grey")) #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_noPTCy_CD8 <- NoPTCy_CD8_results[[1]]
freq_noPTCy_CD8_vio <- NoPTCy_CD8_results[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
NoPTCy_CD8_results_p <- NoPTCy_CD8_results[[3]]

plot(freq_Both)
plot(freq_PTCy_Both)
plot(freq_noPTCy_Both)
plot(freq_PTCy_CD4)
plot(freq_PTCy_CD8)
plot(freq_noPTCy_CD4)
plot(freq_noPTCy_CD8)

plot(freq_Both_vio_ex)
plot(freq_PTCy_Both_vio)
plot(freq_noPTCy_Both_vio)
plot(freq_PTCy_CD4_vio)
plot(freq_PTCy_CD8_vio)
plot(freq_noPTCy_CD4_vio)
plot(freq_noPTCy_CD8_vio)

plot(freq_cd4_vio_ex)
plot(freq_cd8_vio_ex)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_Both',type,'_ex.png'), freq_Both, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_Both',type,'_ex.png'), freq_PTCy_Both, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_Both',type,'_ex.png'), freq_noPTCy_Both, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_CD4',type,'_ex.png'), freq_PTCy_CD4, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_CD8',type,'_ex.png'), freq_PTCy_CD8, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_CD4',type,'_ex.png'), freq_noPTCy_CD4, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_CD8',type,'_ex.png'), freq_noPTCy_CD8, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_cd4_vio',type,'_ex.png'), freq_cd4_vio_ex, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_cd8_vio',type,'_ex.png'), freq_cd8_vio_ex, width = 9.5, height = 5, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/freq_Both_vio',type,'_ex.svg'), freq_Both_vio_ex, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_Both_vio',type,'_ex.svg'), freq_PTCy_Both_vio, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_Both_vio',type,'_ex.svg'), freq_noPTCy_Both_vio, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_CD4_vio',type,'_ex.svg'), freq_PTCy_CD4_vio, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_CD8_vio',type,'_ex.svg'), freq_PTCy_CD8_vio, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_CD4_vio',type,'_ex.svg'), freq_noPTCy_CD4_vio, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_CD8_vio',type,'_ex.svg'), freq_noPTCy_CD8_vio, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_cd4_vio',type,'_ex.svg'), freq_cd4_vio_ex, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_cd8_vio',type,'_ex.svg'), freq_cd8_vio_ex, width = 9.5, height = 5, dpi = 600)


ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/freq_Both_vio',type,'_ex.pdf'), freq_Both_vio_ex, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_Both_vio',type,'_ex.pdf'), freq_PTCy_Both_vio, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_Both_vio',type,'_ex.pdf'), freq_noPTCy_Both_vio, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_CD4_vio',type,'_ex.pdf'), freq_PTCy_CD4_vio, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_CD8_vio',type,'_ex.pdf'), freq_PTCy_CD8_vio, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_CD4_vio',type,'_ex.pdf'), freq_noPTCy_CD4_vio, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_CD8_vio',type,'_ex.pdf'), freq_noPTCy_CD8_vio, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_cd4_vio',type,'_ex.pdf'), freq_cd4_vio_ex, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_cd8_vio',type,'_ex.pdf'), freq_cd8_vio_ex, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_cd4_vio',type,'_ex.png'), freq_cd4_vio_ex, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_cd8_vio',type,'_ex.png'), freq_cd8_vio_ex, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_cd4_vio',type,'_ex.svg'), freq_cd4_vio_ex, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_cd8_vio',type,'_ex.svg'), freq_cd8_vio_ex, width = 9.5, height = 5, dpi = 600)

```

```{r, fig.height=5,fig.width = 9.5, warning = FALSE}

data_CD48_all_expand_adj <- dplyr::full_join(data_CD48_all, clinical_data, by = 'PatientID')
data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!is.na(data_CD48_all_expand_adj$cdr3_amino_acid),]%>%
  filter(strtoi(timepoint) > 3)

type <- "GVHD_GRADE_1"

if(remove_pat_regx != ""){
  data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!grepl(remove_pat_regx,data_CD48_all_expand_adj$PatientID),]
}

data_CD48_all_expand_adj <- data_CD48_all_expand_adj%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("No|Mild", GVHD_GRADE_1), "No and Mild", "Severe"))
data_CD48_all_expand_adj$GVHD_GRADE_1 <- factor(data_CD48_all_expand_adj$GVHD_GRADE_1, c("No and Mild", "Severe") )
# data_CD

data_CD48_all_expand_adj_CD4 <- data_CD48_all_expand_adj%>%
  filter(grepl("CD4",cell_type) & !grepl("CD8", cell_type))

data_CD48_all_expand_adj_CD8 <- data_CD48_all_expand_adj%>%
  filter(!grepl("CD4",cell_type) & grepl("CD8", cell_type))


data_CD48_all_expand_noPTCy <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PTCy == "PTCy", ]
data_CD48_all_expand_PTCy <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PTCy == "No PTCy", ]


Both_results <- num_clone_fig(data_CD48_all_expand_adj, type, "GVHD Grade", num_cuts = 4, colors = c("#058b8e", "#B85000", "grey")) #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_Both <- Both_results[[1]]
freq_Both_vio <- Both_results[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(y = "# of Alloreactive Clones", fill = "GVHD Grade")
Both_results_p <- Both_results[[3]]

Both_results_CD4 <- num_clone_fig(data_CD48_all_expand_adj_CD4, type, "GVHD Grade CD4+", num_cuts = 2, colors = c("#058b8e", "#B85000", "grey")) #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_Both_CD4 <- Both_results_CD4[[1]]
freq_Both_CD4_vio <- Both_results_CD4[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(y = "# of Alloreactive Clones", fill = "GVHD Grade")
Both_results_CD4_p <- Both_results_CD4[[3]]


Both_results_CD8 <- num_clone_fig(data_CD48_all_expand_adj_CD8, type, "GVHD Grade CD8+", num_cuts = 2, colors = c("#058b8e", "#B85000", "grey")) #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_Both_CD8 <- Both_results_CD8[[1]]
freq_Both_CD8_vio <- Both_results_CD8[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(y = "# of Alloreactive Clones", fill = "GVHD Grade")
Both_results_CD8_p <- Both_results_CD8[[3]]

PTCy_Both_results <- num_clone_fig(data_CD48_all_expand_PTCy, type, "PTCy", num_cuts = 4, colors = c("#058b8e", "#B85000", "grey")) #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_PTCy_Both <- PTCy_Both_results[[1]]
freq_PTCy_Both_vio <- PTCy_Both_results[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(y = "# of Alloreactive Clones", fill = "GVHD Grade")
PTCy_Both_results_p <- PTCy_Both_results[[3]]

noPTCy_Both_results <- num_clone_fig(data_CD48_all_expand_noPTCy, type, "No PTCy", num_cuts = 4,  colors = c("#058b8e", "#B85000", "grey")) #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_noPTCy_Both <- noPTCy_Both_results[[1]]
freq_noPTCy_Both_vio <- noPTCy_Both_results[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(y = "# of Alloreactive Clones", fill = "GVHD Grade")
noPTCy_Both_results_p <- noPTCy_Both_results[[3]]

data_CD48_all_expand_PTCy_CD4 <- subset(data_CD48_all_expand_PTCy, grepl("4", data_CD48_all_expand_PTCy$cell_type) & !(grepl("4", data_CD48_all_expand_PTCy$cell_type) & grepl("8", data_CD48_all_expand_PTCy$cell_type)))

PTCy_CD4_results <- num_clone_fig(data_CD48_all_expand_PTCy_CD4, type, "PTCy CD4+", num_cuts = 3,  colors = c("#058b8e", "#B85000", "grey")) #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_PTCy_CD4 <- PTCy_CD4_results[[1]]
freq_PTCy_CD4_vio <- PTCy_CD4_results[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(y = "# of Alloreactive Clones", fill = "GVHD Grade")
PTCy_CD4_results_p <- PTCy_CD4_results[[3]]

data_CD48_all_expand_PTCy_CD8 <- subset(data_CD48_all_expand_PTCy, grepl("8", data_CD48_all_expand_PTCy$cell_type) & !(grepl("4", data_CD48_all_expand_PTCy$cell_type) & grepl("8", data_CD48_all_expand_PTCy$cell_type)))

PTCy_CD8_results <- num_clone_fig(data_CD48_all_expand_PTCy_CD8, type, "PTCy CD8+", num_cuts = 3, colors = c("#058b8e", "#B85000", "grey")) #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_PTCy_CD8 <- PTCy_CD8_results[[1]]
freq_PTCy_CD8_vio <- PTCy_CD8_results[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(y = "# of Alloreactive Clones", fill = "GVHD Grade")
PTCy_CD8_results_p <- PTCy_CD8_results[[3]]

data_CD48_all_expand_noPTCy_CD4 <- subset(data_CD48_all_expand_noPTCy, grepl("4", data_CD48_all_expand_noPTCy$cell_type) & !(grepl("4", data_CD48_all_expand_noPTCy$cell_type) & grepl("8", data_CD48_all_expand_noPTCy$cell_type)))

NoPTCy_CD4_results <- num_clone_fig(data_CD48_all_expand_noPTCy_CD4, type, "No PTCy CD4+", num_cuts = 2, colors = c("#058b8e", "#B85000", "grey")) #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_noPTCy_CD4 <- NoPTCy_CD4_results[[1]]
freq_noPTCy_CD4_vio <- NoPTCy_CD4_results[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(y = "# of Alloreactive Clones", fill = "GVHD Grade")
NoPTCy_CD4_results_p <- NoPTCy_CD4_results[[3]]


data_CD48_all_expand_noPTCy_CD8 <- subset(data_CD48_all_expand_noPTCy, grepl("8", data_CD48_all_expand_noPTCy$cell_type) & !(grepl("4", data_CD48_all_expand_noPTCy$cell_type) & grepl("8", data_CD48_all_expand_noPTCy$cell_type)))

NoPTCy_CD8_results <- num_clone_fig(data_CD48_all_expand_noPTCy_CD8, type, "No PTCy CD8+", num_cuts = 3, colors = c("#058b8e", "#B85000", "grey")) #c("#09BB8C", "#005A8F", "#B85000", "grey"))
freq_noPTCy_CD8 <- NoPTCy_CD8_results[[1]]
freq_noPTCy_CD8_vio <- NoPTCy_CD8_results[[2]]+
  scale_fill_manual(values = c("#058b8e", "#B85000", "grey"), labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(y = "# of Alloreactive Clones", fill = "GVHD Grade")
NoPTCy_CD8_results_p <- NoPTCy_CD8_results[[3]]

plot(freq_Both)
plot(freq_PTCy_Both)
plot(freq_noPTCy_Both)
plot(freq_PTCy_CD4)
plot(freq_PTCy_CD8)
plot(freq_noPTCy_CD4)
plot(freq_noPTCy_CD8)

plot(freq_Both_vio)
plot(freq_Both_CD4_vio)
plot(freq_Both_CD8_vio)
plot(freq_PTCy_Both_vio)
plot(freq_noPTCy_Both_vio)
plot(freq_PTCy_CD4_vio)
plot(freq_PTCy_CD8_vio)
plot(freq_noPTCy_CD4_vio)
plot(freq_noPTCy_CD8_vio)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_Both',type,'.png'), freq_Both_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_Both_cd4',type,'.png'), freq_Both_CD4_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_Both_cd8',type,'.png'), freq_Both_CD8_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_Both',type,'.png'), freq_PTCy_Both_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_Both',type,'.png'), freq_noPTCy_Both_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_CD4',type,'.png'), freq_PTCy_CD4_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_CD8',type,'.png'), freq_PTCy_CD8_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_CD4',type,'.png'), freq_noPTCy_CD4_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_CD8',type,'.png'), freq_noPTCy_CD8_vio, width = 8.5, height = 4, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_Both',type,'.pdf'), freq_Both_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_Both_cd4',type,'.pdf'), freq_Both_CD4_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_Both_cd8',type,'.pdf'), freq_Both_CD8_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_Both',type,'.pdf'), freq_PTCy_Both_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_Both',type,'.pdf'), freq_noPTCy_Both_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_CD4',type,'.pdf'), freq_PTCy_CD4_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_CD8',type,'.pdf'), freq_PTCy_CD8_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_CD4',type,'.pdf'), freq_noPTCy_CD4_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_CD8',type,'.pdf'), freq_noPTCy_CD8_vio, width = 8.5, height = 4, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_Both',type,'.svg'), freq_Both_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_Both_cd4',type,'.svg'), freq_Both_CD4_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_Both_cd8',type,'.svg'), freq_Both_CD8_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_Both',type,'.svg'), freq_PTCy_Both_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_Both',type,'.svg'), freq_noPTCy_Both_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_CD4',type,'.svg'), freq_PTCy_CD4_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_PTCy_CD8',type,'.svg'), freq_PTCy_CD8_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_CD4',type,'.svg'), freq_noPTCy_CD4_vio, width = 8.5, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/freq_noPTCy_CD8',type,'.svg'), freq_noPTCy_CD8_vio, width = 8.5, height = 4, dpi = 600)


```

```{r, fig.width = 21, fig.height = 5}
patient_data_allo <- dplyr::full_join(patient_data_all_2, meta_data,  by = "PatientID")
colnames(patient_data_allo) <- sapply(colnames(patient_data_allo), function(x){str_replace_all(x, "-", "")})
patient_data_allo <- patient_data_allo[patient_data_allo$Rep_div == "allo",]
patient_data_allo <- patient_data_allo[which(!is.na(patient_data_allo$timepoint)),]
# patient_data_allo[patient_data_allo$timepoint == "MLRCFSElo",]$timepoint  <- "1"
# patient_data_allo[patient_data_allo$timepoint == "unstimulated",]$timepoint  <- "0"
patient_data_allo_copy <- patient_data_allo%>%
  mutate(timepoint = strtoi(timepoint))
patient_data_allo_copy <- patient_data_allo_copy[which(!is.na(patient_data_allo_copy$timepoint)),]
patient_data_allo_copy <- patient_data_allo_copy[which(!is.na(patient_data_allo_copy$shannons.diversity)),]
patient_data_allo_copy <- patient_data_allo_copy%>%
  mutate(GVHD_grade_14 = ifelse(GVHD_grade_14 == "Yes", 1, 0), GVHD_grade_24 = ifelse(GVHD_grade_24 == "Yes", 1, 0), GVHD_grade_34 = ifelse(GVHD_grade_34 == "Yes", 1, 0))%>%
  mutate(GVHD_GRADE = GVHD_grade_14 + GVHD_grade_24 + GVHD_grade_34)

data_CD48_all_expand_adj <- dplyr::full_join(data_CD48_all, clinical_data, by = 'PatientID')
data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!is.na(data_CD48_all_expand_adj$cdr3_amino_acid),]
data_CD48_all_expand_adj <- data_CD48_all_expand_adj


data_CD48_all_expand_noPTCy <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PTCy == "No PTCy", ]
data_CD48_all_expand_PTCy <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PTCy == "PTCy",]
# data_CD48_all_expand_noPTCy <- data_CD48_all_expand_noPTCy[!grepl("70",data_CD48_all_expand_noPTCy$PatientID),]
Both_results <- num_clone_fig(data_CD48_all, "PTCy", "PTCy vs No PTCy", num_cuts = 7, legend_title = "Treatment")
immdata_meta <- dplyr::full_join(Both_results[[4]], meta_data_clin, by = "PatientID")%>%
  mutate(PTCy = PTCy.x)%>%
  filter(!is.na(PTCy))


patient_data_allo_copy <- patient_data_allo_copy %>%
  filter(timepoint != 3)

if(remove_pat_regx != ""){
  data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!grepl(remove_pat_regx,data_CD48_all_expand_adj$PatientID),]
  patient_data_allo_copy <- patient_data_allo_copy[!grepl(remove_pat_regx,patient_data_allo_copy$PatientID),]
}

cum_freq_plot <- ggplot(patient_data_allo_copy, aes(x = PTCy, y = cum_freq, group = PTCy, fill = PTCy))+ #grepl("PTCy", PTCy) grepl("Haplo",DonorType)
  geom_boxplot()+
  stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test", label = "p.signif", hide.ns = TRUE)+
  geom_beeswarm()+
  scale_y_continuous(trans = "log10", labels = scales::number_format(accuracy = 0.0001))+
  labs(y = "Alloreactive\n Cumulative Frequency")+
  theme_pubr()+
  theme(text = element_text(size = 24), legend.text = element_markdown(), legend.position = "none")+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"),
                    labels = c(
      `No PTCy` = "<span style='color:#E6BE60;'>No PTCy</span>",
      PTCy = "<span style='color:#A483AF;'>PTCy</span>"))


shannons_plot <- ggplot(patient_data_allo_copy, aes(x = PTCy, y = shannons.entropy, group = PTCy, fill = PTCy))+ 
  geom_boxplot()+
  stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test", label = "p.signif", hide.ns = TRUE)+
  geom_beeswarm()+
  labs(y = "Shannon's Diversity Index")+
  scale_y_continuous(labels = scales::number_format(accuracy = 1))+
  theme_pubr()+
  theme(text = element_text(size = 24), legend.text = element_markdown())+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"),
                    labels = c(
      `No PTCy` = "<span style='color:#E6BE60;'>No PTCy</span>",
      PTCy = "<span style='color:#A483AF;'>PTCy</span>"))

unique_clone_ptcy_plot <- ggplot(immdata_meta, aes(x = PTCy, y = n+1, group = PTCy, fill = PTCy))+ 
  geom_boxplot()+
  stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test", label = "p.signif", hide.ns = TRUE)+
  geom_beeswarm()+
  scale_y_continuous(trans = "log10", labels = scales::number_format(accuracy = 1))+
  scale_x_discrete(na.translate = F)+
  theme_pubr()+
  theme(text = element_text(size = 24), legend.text = element_markdown(), legend.position = "none")+
  labs(y = "# of Clones")+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"),
                    labels = c(
      `No PTCy` = "<span style='color:#E6BE60;'>No PTCy</span>",
      PTCy = "<span style='color:#A483AF;'>PTCy</span>"))


print(wilcox.test(patient_data_allo_copy[grepl("Haplo",patient_data_allo_copy$DonorType),]$cum_freq, patient_data_allo_copy[!grepl("Haplo",patient_data_allo_copy$DonorType),]$cum_freq))

print(wilcox.test(patient_data_allo_copy[grepl("Haplo",patient_data_allo_copy$DonorType),]$shannons.entropy, patient_data_allo_copy[!grepl("Haplo",patient_data_allo_copy$DonorType),]$shannons.entropy))

plot(cum_freq_plot)
plot(shannons_plot)
plot(unique_clone_ptcy_plot)

combined_plot <- plot_grid(cum_freq_plot, shannons_plot, unique_clone_ptcy_plot, nrow = 1, align = "hv")

plot(combined_plot)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/CombinedPTCy.png'), combined_plot, width = 21, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/CombinedPTCy.pdf'), combined_plot, width = 21, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/CombinedPTCy.svg'), combined_plot, width = 21, height = 5, dpi = 600)



ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/PTCyDiffCumulativeFrequuency.png'), cum_freq_plot, width = 7, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/PTCyDiffShannons.png'), shannons_plot, width = 7, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/PTCyDiffUniqueClones.png'), unique_clone_ptcy_plot, width = 7, height = 5, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/PTCyDiffCumulativeFrequuency.pdf'), cum_freq_plot, width = 7, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/PTCyDiffShannons.pdf'), shannons_plot, width = 7, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/PTCyDiffUniqueClones.pdf'), unique_clone_ptcy_plot, width = 7, height = 5, dpi = 600)


cum_freq_pval <- wilcox.test(
  patient_data_allo_copy[patient_data_allo_copy$PTCy == "PTCy",]$cum_freq,
  patient_data_allo_copy[patient_data_allo_copy$PTCy == "No PTCy",]$cum_freq
)$p.value

# 2. For Shannon's entropy comparison (PTCy vs No PTCy)
shannon_pval <- wilcox.test(
  patient_data_allo_copy[patient_data_allo_copy$PTCy == "PTCy",]$shannons.entropy,
  patient_data_allo_copy[patient_data_allo_copy$PTCy == "No PTCy",]$shannons.entropy
)$p.value

# 3. For unique clone count comparison (PTCy vs No PTCy)
unique_clone_pval <- wilcox.test(
  immdata_meta[immdata_meta$PTCy == "PTCy",]$n,
  immdata_meta[immdata_meta$PTCy == "No PTCy",]$n
)$p.value

```

```{r, fig.width = 21, fig.height = 5}
patient_data_allo <- dplyr::full_join(patient_data_all_2, meta_data,  by = "PatientID")
colnames(patient_data_allo) <- sapply(colnames(patient_data_allo), function(x){str_replace_all(x, "-", "")})
patient_data_allo <- patient_data_allo[patient_data_allo$Rep_div == "allo",]
patient_data_allo <- patient_data_allo[which(!is.na(patient_data_allo$timepoint)),]
# patient_data_allo[patient_data_allo$timepoint == "MLRCFSElo",]$timepoint  <- "1"
# patient_data_allo[patient_data_allo$timepoint == "unstimulated",]$timepoint  <- "0"
patient_data_allo_copy <- patient_data_allo%>%
  mutate(timepoint = strtoi(timepoint))
patient_data_allo_copy <- patient_data_allo_copy[which(!is.na(patient_data_allo_copy$timepoint)),]
patient_data_allo_copy <- patient_data_allo_copy[which(!is.na(patient_data_allo_copy$shannons.diversity)),]

patient_data_allo_copy <- patient_data_allo_copy%>%
  mutate(GVHD_grade_14 = ifelse(GVHD_grade_14 == "Yes", 1, 0), GVHD_grade_24 = ifelse(GVHD_grade_24 == "Yes", 1, 0), GVHD_grade_34 = ifelse(GVHD_grade_34 == "Yes", 1, 0))%>%
  mutate(GVHD_GRADE = GVHD_grade_14 + GVHD_grade_24 + GVHD_grade_34)

# data_CD48_all_expand_adj <- dplyr::full_join(data_CD48_all, clinical_data, by = 'PatientID')
# data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!is.na(data_CD48_all_expand_adj$cdr3_amino_acid),]
# data_CD48_all_expand_adj <- data_CD48_all_expand_adj%>%mutate(PTCy = factor(PTCy, c("No PTCy", "PTCy")))

#AVERAGE CLONE SIZE

# data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!grepl("67",data_CD48_all_expand_adj$PatientID),]
# data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!grepl("70",data_CD48_all_expand_adj$PatientID),]
# 
# data_CD48_all_expand_noPTCy <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PTCy == "No PTCy", ]
# data_CD48_all_expand_PTCy <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PTCy == "PTCy",]
Both_results <- num_clone_fig(data_CD48_all, "PTCy", "PTCy vs No PTCy", num_cuts = 7, legend_title = "Treatment")
immdata_meta <- dplyr::full_join(Both_results[[4]], meta_data_clin, by = "PatientID")%>%
  mutate(PTCy = PTCy.x)%>%
  mutate(GVHD_GRADE_1 = factor(GVHD_GRADE_1, levels = c("No", "Mild", "Severe")))%>%
  filter(!is.na(timepoint))


patient_data_allo_copy <- patient_data_allo_copy%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1))#%>%
  #mutate(PTCy = ifelse(grepl("PTCy", PTCy), "PTCy", "No PTCy"))

patient_data_allo_copy <- patient_data_allo_copy %>%
  filter(timepoint != 3)%>%
  mutate(GVHD_GRADE_1 = factor(GVHD_GRADE_1, levels = c("No", "Mild", "Severe")))

if(remove_pat_regx != ""){
  data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!grepl(remove_pat_regx,data_CD48_all_expand_adj$PatientID),]
  patient_data_allo_copy <- patient_data_allo_copy[!grepl(remove_pat_regx,patient_data_allo_copy$PatientID),]
  immdata_meta <- immdata_meta[!grepl(remove_pat_regx,immdata_meta$PatientID),]
}


cum_freq_plot <- ggplot(patient_data_allo_copy, aes(x = GVHD_GRADE_1, y = cum_freq, group = GVHD_GRADE_1, fill = GVHD_GRADE_1))+ 
  geom_boxplot()+
  geom_beeswarm()+
  stat_compare_means(data = patient_data_allo_copy, comparisons =  list(c("No", "Mild"),c("No", "Severe"), c("Mild", "Severe")), method = "wilcox.test", label = "p.signif", hide.ns = TRUE, p.adjust.method = "hochberg", inherit.aes = FALSE, exact = F, aes(x = GVHD_GRADE_1, y = cum_freq, group = GVHD_GRADE_1))+
  scale_y_continuous(trans = "log10", labels = scales::number_format(accuracy = 0.0001))+
  scale_x_discrete(na.translate = F)+
  theme_pubr()+
  theme(text = element_text(size = 18), legend.position = "right")+
  scale_fill_manual(values = c("#09BB8C", "#005A8F", "#B85000", "grey"))+
  labs(title = "Alloreactive", x = "GVHD Grade", y = "Cumulative Frequency", fill = "GVHD Grade")
  # scale_color_manual(values=c("#E6BE60", "#A483AF"))

shannons_plot <- ggplot(patient_data_allo_copy, aes(x = GVHD_GRADE_1, y = shannons.entropy, group = GVHD_GRADE_1, fill = GVHD_GRADE_1))+ 
  geom_boxplot()+
  stat_compare_means(comparisons = list(c("No", "Mild"),c("No", "Severe"), c("Mild", "Severe")), exact = F, method = "wilcox.test", label = "p.signif", hide.ns = TRUE, p.adjust.method = "hochberg")+
  geom_beeswarm()+
  scale_y_continuous(labels = scales::number_format(accuracy = 1))+
  theme_pubr()+
  theme(text = element_text(size = 18), legend.position = "right")+
  scale_fill_manual(values=c("#09BB8C", "#005A8F", "#B85000", "grey"))+
  labs(title = "Alloreactive", x = "GVHD Grade", y = "Shannon's Diversity Index", fill = "GVHD Grade")
  # scale_color_manual(values=c("#E6BE60", "#A483AF"))

unique_clone_ptcy_plot <- ggplot(immdata_meta, aes(x = GVHD_GRADE_1, y = n+1, group = GVHD_GRADE_1, fill = GVHD_GRADE_1))+ 
  geom_boxplot()+
  stat_compare_means(comparisons = list(c("No", "Mild"),c("No", "Severe"), c("Mild", "Severe")), method = "wilcox.test", label = "p.signif", hide.ns = TRUE, exact = F, p.adjust.method = "hochberg")+
  geom_beeswarm()+
  scale_y_continuous(trans = "log10", labels = scales::number_format(accuracy = 1))+
  scale_x_discrete(na.translate = F)+
  theme_pubr()+
  theme(text = element_text(size = 18), legend.position = "right")+
  scale_fill_manual(values=c("#09BB8C", "#005A8F", "#B85000", "grey"))+
  labs(title = "Alloreactive", x = "GVHD Grade", y = "# of Clones", fill = "GVHD Grade")
  # scale_color_manual(values=c("#E6BE60", "#A483AF"))


cum_freq_plot_NoPTCy <- ggplot(patient_data_allo_copy[patient_data_allo_copy$PTCy == "No PTCy",], aes(x = GVHD_GRADE_1, y = cum_freq, group = GVHD_GRADE_1, fill = GVHD_GRADE_1))+ #grepl("PTCy", PTCy) grepl("Haplo",DonorType)
  geom_boxplot()+
  stat_compare_means(comparisons = list(c("No", "Mild"),c("No", "Severe"), c("Mild", "Severe")), method = "wilcox.test", label = "p.signif", hide.ns = TRUE, exact = F, p.adjust.method = "hochberg")+
  geom_beeswarm(aes(color = DonorType))+
  scale_y_continuous(trans = "log10", labels = scales::number_format(accuracy = 0.01))+
  scale_x_discrete(na.translate = F)+
  theme_pubr()+
  theme(text = element_text(size = 18))+
  scale_fill_manual(values = c("#09BB8C", "#005A8F", "#B85000", "grey"))+
  labs(x = "GVHD Grade", y = "Alloreactive Frequency", title = "No PTCy", fill = "GVHD Grade")

shannons_plot_NoPTCy <- ggplot(patient_data_allo_copy[patient_data_allo_copy$PTCy == "No PTCy",], aes(x = GVHD_GRADE_1, y = shannons.entropy, group = GVHD_GRADE_1, fill = GVHD_GRADE_1))+ 
  geom_boxplot()+
  stat_compare_means(comparisons = list(c("No", "Mild"),c("No", "Severe"), c("Mild", "Severe")), method = "wilcox.test", label = "p.signif", hide.ns = TRUE, exact = F, p.adjust.method = "hochberg")+
  geom_beeswarm(aes(color = DonorType))+
  scale_y_continuous(labels = scales::number_format(accuracy = 0.01))+
  scale_x_discrete(na.translate = F)+
  theme_pubr()+
  theme(text = element_text(size = 18))+
  scale_fill_manual(name = "GVHD Grade", values=c("#09BB8C", "#005A8F", "#B85000", "grey"))+
  labs(x = "GVHD Grade", y = "Shannon's Diversity Index", title = "No PTCy", fill = "GVHD Grade")

unique_clone_plot_NoPTCy <- ggplot(immdata_meta[immdata_meta$PTCy == "No PTCy",], aes(x = GVHD_GRADE_1, y = n+1, group = GVHD_GRADE_1, fill = GVHD_GRADE_1))+ 
  geom_boxplot()+
  stat_compare_means(comparisons = list(c("No", "Mild"),c("No", "Severe"), c("Mild", "Severe")), method = "wilcox.test", label = "p.signif", hide.ns = TRUE, exact = F, p.adjust.method = "hochberg")+
  geom_beeswarm(aes(color = DonorType))+
  scale_y_continuous(trans = "log10", labels = scales::number_format(accuracy = 0.01))+
  scale_x_discrete(na.translate = F)+
  theme_pubr()+
  theme(text = element_text(size = 18))+
  scale_fill_manual(values=c("#09BB8C", "#005A8F", "#B85000", "grey"))
  labs(x = "GVHD Grade", y = "# of Clones", title = "No PTCy", fill = "GVHD Grade")
  

cum_freq_plot_PTCy <- ggplot(patient_data_allo_copy[patient_data_allo_copy$PTCy == "PTCy",], aes(x = GVHD_GRADE_1, y = cum_freq, group = GVHD_GRADE_1, fill = GVHD_GRADE_1))+ #grepl("PTCy", PTCy) grepl("Haplo",DonorType)
  geom_boxplot()+
  stat_compare_means(comparisons = list(c("No", "Mild"),c("No", "Severe"), c("Mild", "Severe")), method = "wilcox.test", label = "p.signif", hide.ns = TRUE, exact = F, p.adjust.method = "hochberg")+
  geom_beeswarm(aes(color = PatientID))+
  scale_y_continuous(trans = "log10", labels = scales::number_format(accuracy = 0.01))+
  scale_x_discrete(na.translate = F)+
  labs(x = "GVHD Grade", y = "Alloreactive Frequency", title = "PTCy")+
  theme_pubr()+
  theme(text = element_text(size = 18))+
  scale_fill_manual(name = "GVHD Grade", values = c("#09BB8C", "#005A8F", "#B85000", "grey"))

shannons_plot_PTCy <- ggplot(patient_data_allo_copy[patient_data_allo_copy$PTCy == "PTCy",], aes(x = GVHD_GRADE_1, y = shannons.entropy, group = GVHD_GRADE_1, fill = GVHD_GRADE_1))+ 
  geom_boxplot()+
  stat_compare_means(comparisons = list(c("No", "Mild"),c("No", "Severe"), c("Mild", "Severe")), method = "wilcox.test", label = "p.signif", hide.ns = TRUE, exact = F, p.adjust.method = "hochberg")+
  geom_beeswarm(aes(color = PatientID))+
  labs(x = "GVHD Grade", y = "Shannon's Diversity Index", title = "PTCy")+
  scale_y_continuous(labels = scales::number_format(accuracy = 0.01))+
  scale_x_discrete(na.translate = F)+
  theme_minimal()+
  theme(text = element_text(size = 18))+
  scale_fill_manual(name = "GVHD Grade", values=c("#09BB8C", "#005A8F", "#B85000", "grey"))

unique_clone_plot_PTCy <- ggplot(immdata_meta[immdata_meta$PTCy == "PTCy",], aes(x = GVHD_GRADE_1, y = n+1, group = GVHD_GRADE_1, fill = GVHD_GRADE_1))+ 
  geom_boxplot()+
  stat_compare_means(comparisons = list(c("No", "Mild"),c("No", "Severe"), c("Mild", "Severe")), method = "wilcox.test", label = "p.signif", hide.ns = TRUE, exact = F, p.adjust.method = "hochberg")+
  geom_beeswarm(aes(color = PatientID))+
  scale_y_continuous(trans = "log10", labels = scales::number_format(accuracy = 0.01))+
  scale_x_discrete(na.translate = F)+
  theme_pubr()+
  theme(text = element_text(size = 18))+
  labs(x = "GVHD Grade", y = "# of Clones", title = "PTCy")+
  scale_fill_manual(name = "GVHD Grade", values=c("#09BB8C", "#005A8F", "#B85000", "grey"))


print(wilcox.test(patient_data_allo_copy[grepl("Haplo",patient_data_allo_copy$DonorType),]$cum_freq, patient_data_allo_copy[!grepl("Haplo",patient_data_allo_copy$DonorType),]$cum_freq))

print(wilcox.test(patient_data_allo_copy[grepl("Haplo",patient_data_allo_copy$DonorType),]$shannons.entropy, patient_data_allo_copy[!grepl("Haplo",patient_data_allo_copy$DonorType),]$shannons.entropy))

data_CD48_all_expand_not_PTCy <- data_CD48_all_expand_adj[!grepl("PTCy", data_CD48_all_expand_adj$PTCy), ]

# print(wilcox.test(immdata_meta[grepl("PTCy", immdata_meta$PTCy,) ]$n, immdata_meta[!grepl("PTCy", immdata_meta$PTCy,)]$n))


plot(cum_freq_plot)
plot(shannons_plot)
plot(unique_clone_ptcy_plot)
plot(cum_freq_plot_NoPTCy)
plot(shannons_plot_NoPTCy)
plot(unique_clone_plot_NoPTCy)
plot(cum_freq_plot_PTCy)
plot(shannons_plot_PTCy)
plot(unique_clone_plot_PTCy)


cum_freq_plot_c1 <- cum_freq_plot+
  theme(legend.position = "none")
shannons_plot_c2 <- shannons_plot+
  theme(legend.position = "none")


ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/C-1.png'), cum_freq_plot_c1, width = 8, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/C-2.png'), shannons_plot_c2, width = 8, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/C-3.png'), unique_clone_ptcy_plot, width = 8, height = 5, dpi = 600)

combined_plot <- plot_grid(cum_freq_plot_c1, shannons_plot_c2, unique_clone_ptcy_plot, nrow = 1, align = "h", rel_widths = c(1,1,1.3))

plot(combined_plot)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/C.png'), combined_plot, width = 21, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/C.pdf'), combined_plot, width = 21, height = 5, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GVHDDiffCumulativeFrequencyNoPTCy.png'), cum_freq_plot_NoPTCy, width = 6, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GVHDDiffShannonsNoPTCy.png'), shannons_plot_NoPTCy, width = 6, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GVHDDiffUniqueClonesNoPTCy.png'), unique_clone_plot_NoPTCy, width = 6, height = 5, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GVHDDiffCumulativeFrequencyPTCy.png'), cum_freq_plot_PTCy, width = 6, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GVHDDiffShannonsPTCy.png'), shannons_plot_PTCy, width = 6, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GVHDDiffUniqueClonesPTCy.png'), unique_clone_plot_PTCy, width = 6, height = 5, dpi = 600)


ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/C-1.pdf'), cum_freq_plot_c1, width = 8, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/C-2.pdf'), shannons_plot_c2, width = 8, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/C-3.pdf'), unique_clone_ptcy_plot, width = 8, height = 5, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GVHDDiffCumulativeFrequencyNoPTCy.png'), cum_freq_plot, width = 6, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GVHDDiffShannonsNoPTCy.png'), shannons_plot, width = 6, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GVHDDiffUniqueClonesNoPTCy.png'), unique_clone_ptcy_plot, width = 6, height = 5, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GVHDDiffCumulativeFrequencyPTCy.png'), cum_freq_plot, width = 6, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GVHDDiffShannonsPTCy.png'), shannons_plot, width = 6, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GVHDDiffUniqueClonesPTCy.png'), unique_clone_ptcy_plot, width = 6, height = 5, dpi = 600)


```

```{r, fig.height = 8, fig.width=16, dpi = 600}


get_timepoint <- function(df) {
  #' Get earliest post-stimulation timepoints for each patient
  #'
  #' @description This function selects all timepoints less than or equal to 3 and the earliest timepoint greater than 3 for each patient.
  #'
  #' @param df A dataframe containing `timepoint` and `PatientID` columns.
  #' @return A dataframe with the earliest timepoints for each patient.
  
  df <- df %>%
    mutate(timepoint = strtoi(timepoint)) %>%
    filter(!is.na(timepoint))
  
  # Select all timepoints less than or equal to 3
  df_res <- df %>%
    filter(timepoint <= 3)
  
  # For each patient, select the earliest timepoint greater than 3
  df_patients <- df %>%
    group_by(PatientID) %>%
    filter(timepoint > 3) %>%
    filter(timepoint == min(timepoint)) %>%
    ungroup()
  
  # Combine the selected timepoints
  df_res <- bind_rows(df_res, df_patients)
  
  return(df_res)
}


get_early_timepoint <- function(df) {
  #' Get earliest and specific early timepoints for each patient
  #'
  #' @description This function selects the earliest timepoint greater than 3 and also specific early timepoints between 2 and 3 for each patient.
  #'
  #' @param df A dataframe containing `timepoint` and `PatientID` columns.
  #' @return A dataframe with the earliest timepoints and specific early timepoints for each patient.
  
  df_res <- data.frame()
  
  for (patient in unique(df$PatientID)) {
    # Select earliest timepoint greater than 3 for each patient
    df_pat <- df %>%
      filter(PatientID == patient, timepoint > 3) %>%
      filter(timepoint == min(timepoint))
    
    # Select timepoints between 2 and 3 for each patient
    df_pat_3 <- df %>%
      filter(PatientID == patient, timepoint >= 2, timepoint <= 3)
    
    # Combine the selected timepoints
    df_res <- bind_rows(df_res, df_pat, df_pat_3)
  }
  
  return(df_res)
}



unstim_plot_freqs <- vector("list", 3)
unstim_3_plot_freqs <- vector("list", 3)
early_plot_freqs <- vector("list", 3)
earliest_plot_freqs <- vector("list", 3)
unstim_plot_shannons <- vector("list", 3)
unstim_3_plot_shannons <- vector("list", 3)
early_plot_shannons <- vector("list", 3)
earliest_plot_shannons <- vector("list", 3)
early_plot_n_clones <- vector("list", 3)
early_plot_avg_freqs <- vector("list", 3)

for(rep_div in c("allo", "nonallo", "all")){
  if (rep_div == "allo"){
    i = 1
    rep_title = "Alloreactive"
  }else if(rep_div == "nonallo"){ 
    i = 2
    rep_title = "Non-Alloreactive"
  }else{
    i = 3
    rep_title = "All"
  }
  patient_data_allo <- dplyr::full_join(patient_data_all, meta_data,  by = "PatientID")%>%
    mutate(timepoint = SampleID)%>%
    select(-c(SampleID))
  colnames(patient_data_allo) <- sapply(colnames(patient_data_allo), function(x){str_replace_all(x, "-", "")})
  patient_data_allo <- patient_data_allo[patient_data_allo$Rep_div == rep_div,]
  patient_data_allo <- patient_data_allo[which(!is.na(patient_data_allo$timepoint)),]
  patient_data_allo[patient_data_allo$timepoint == "MLRCFSElo",]$timepoint  <- "1"
  patient_data_allo[patient_data_allo$timepoint == "unstimulated",]$timepoint  <- "0"
  patient_data_allo_copy <- patient_data_allo%>%
    mutate(timepoint = strtoi(timepoint))
  patient_data_allo_copy <- patient_data_allo_copy[which(!is.na(patient_data_allo_copy$timepoint)),]
  
  
  patient_data_allo_copy <- patient_data_allo_copy%>%
    mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1))%>%
    mutate(GVHD_GRADE_1 = factor(GVHD_GRADE_1, levels = c("No", "Mild", "Severe")))%>%
    replace_na(list(shannons.diversity = 0, shannons.entropy = 0))#%>%
    # mutate(PTCy = ifelse(grepl("PTCy", PTCy), "PTCy", "No PTCy"))
  
  df_unstimulated <- patient_data_allo_copy[patient_data_allo_copy$timepoint == 0,]%>%
    mutate(timepoint = "Donor")%>%
    filter(PTCy == "PTCy")
  
  df_unstimulated_all <- patient_data_allo_copy[patient_data_allo_copy$timepoint == 0,]%>%
    mutate(timepoint = "Donor")
  
  
  patient_data_allo_copy <- patient_data_allo_copy[patient_data_allo_copy$timepoint > 1,]

  df_early_timepoint_all <- get_early_timepoint(patient_data_allo_copy)%>%
    mutate(timepoint = ifelse(timepoint <= 3, "<= 3", "Earliest Timepoint"))%>%
    mutate(timepoint = factor(timepoint, levels = c("<= 3", "Earliest Timepoint")))%>%
    group_by(PatientID,GVHD_GRADE_1, timepoint)
  
  df_early_timepoint <- df_early_timepoint_all %>%
    filter(PTCy == "PTCy")
  
  df_early_timepoint <- df_early_timepoint[df_early_timepoint$PatientID != "R070",]
  
  data_CD48_all_expand_adj <- dplyr::full_join(data_CD48_all_expand, clinical_data, by = 'PatientID')
  data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!is.na(data_CD48_all_expand_adj$cdr3_amino_acid),]
  data_CD48_all_expand_adj <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PatientID != "R070",]
  
  df_timepoint <- get_timepoint(patient_data_allo_copy)%>%
    mutate(timepoint = ifelse(timepoint <= 3 & timepoint >=2, "<= 3", "Earliest Timepoint"))%>%
    mutate(timepoint = factor(timepoint, levels = c("Donor", "<= 3", "Earliest Timepoint")))%>%
    filter(PTCy == "PTCy")
  
  df_timepoint <- df_timepoint[df_timepoint$PatientID != 70,]
  

  df_unstimulated_3 <- bind_rows(df_unstimulated%>%filter(PatientID %in% unique(df_timepoint$PatientID)), df_early_timepoint)%>%
    mutate(timepoint = factor(timepoint, levels = c("Donor", "<= 3", "Earliest Timepoint")))%>%
    # filter(GVHD_GRADE_1 != "No")%>%
    group_by(PatientID)%>%
    mutate(n_in = n())%>%
    filter(n_in >= 3)%>%
    ungroup()%>%
    replace_na(list(avg_freq = 0))
  

  unstim_plot_freq <- ggplot(df_unstimulated_all, aes(x = PTCy, y = cum_freq, fill = PTCy))+
    geom_boxplot()+
    geom_beeswarm(aes(color = GVHD_GRADE_1))+
    scale_color_manual(values = c("#09BB8C", "#005A8F", "#B85000", "grey"))+
    scale_y_continuous(labels = scales::number_format(accuracy = 0.0001))+
    stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test", label = "p.signif", hide.ns = FALSE, label.y = quantile(df_unstimulated_all$cum_freq, probs = c(0.9)), size = 12)+
    labs(x = "Timepoint", y = "Frequency", title = paste(rep_title, "Unstimulated"), color = "GVHD Grade")+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
    # expand_limits(y = max(df_unstimulated_all$cum_freq)+.05)+
    theme_pubr()+
    theme(legend.position = "none")


  unstim_3_plot_freq <- ggplot(df_unstimulated_3, aes(x = timepoint, y = cum_freq, shape = PatientID))+
    geom_point(aes(color = GVHD_GRADE_1), size = 6)+
    geom_line(aes(group = PatientID, color = GVHD_GRADE_1), linewidth = 1.5)+
    scale_color_manual(values = c( "#005A8F", "#B85000", "grey"))+
    scale_y_continuous(labels = scales::number_format(accuracy = 0.0001))+
    stat_compare_means(comparisons = list(c("Donor", "<= 3")), method = "wilcox.test", paired = TRUE, label = "p.signif", hide.ns = FALSE, group.by = "PatientID", size = 12, label.y = quantile(df_unstimulated_3$cum_freq, probs = c(0.9)))+
    labs(x = "Timepoint", y = "Cumulative Frequency", title = paste(rep_title, "Unstimulated"), color = "GVHD Grade")+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
    # expand_limits(y = max(df_unstimulated_3$cum_freq)+.05)+
    theme_pubr()+
    theme(legend.position = "none")
  
  earliest_plot_freq <- ggplot(df_early_timepoint_all[df_early_timepoint_all$timepoint == "Earliest Timepoint",], aes(x = PTCy, y = cum_freq, fill = PTCy))+
    geom_boxplot()+
    geom_beeswarm(aes(color = GVHD_GRADE_1))+
    scale_y_continuous(labels = scales::number_format(accuracy = 0.0001))+
    scale_color_manual(values = c("#09BB8C", "#005A8F", "#B85000", "grey"))+
    stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test",label = "p.signif", hide.ns = FALSE, label.y = quantile(df_early_timepoint_all$cum_freq, probs = c(0.9)), size = 6)+
    labs(x = "Timepoint", y = "Cumulative Frequency", title = paste(rep_title, "Earliest Timepoint"), color = "GVHD Grade")+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
    # expand_limits(y = max(df_early_timepoint_all$cum_freq)+.05)+
    theme_pubr()+
    theme(legend.position = "none")

  test_res <- wilcox.test(cum_freq ~ timepoint, data = df_unstimulated_3[!grepl("Donor",df_unstimulated_3$timepoint),], paired = TRUE, alternative = "greater")
  print(test_res)
 
  early_plot_freq <- ggplot(df_unstimulated_3[df_unstimulated_3$timepoint != "Donor",], aes(x = timepoint, y = cum_freq, shape = PatientID))+
    geom_line(aes(group = PatientID, color = GVHD_GRADE_1), linewidth = 1.5)+
    geom_point(aes(color = GVHD_GRADE_1), size = 6)+
    scale_y_continuous(labels = scales::number_format(accuracy = 0.0001))+
    stat_compare_means(comparisons = list(c("<= 3", "Earliest Timepoint")), method = "wilcox.test", paired = TRUE, label = "p.signif", hide.ns = FALSE, group.by = "PatientID", method.args = list(alternative = "greater"), size = 12, label.y = quantile(df_unstimulated_3$cum_freq, probs = c(0.9)))+
    scale_color_manual(values = c("#005A8F", "#B85000", "grey"))+
    # expand_limits(y = max(df_timepoint$cum_freq)+.05)+
    labs(x = "Timepoint", y = "Cumulative Frequency", title = rep_title, color = "GVHD Grade")+
    # coord_cartesian(ylim = c(0,0.4))+
    theme_minimal()+
    theme_pubr()+
    theme(legend.position = "none")
  
   
  unstim_plot_shannon <- ggplot(df_unstimulated_all, aes(x = PTCy, y = shannons.diversity, fill = PTCy))+
    geom_boxplot()+
    geom_beeswarm(aes(color = GVHD_GRADE_1))+
    scale_color_manual(values = c("#09BB8C", "#005A8F", "#B85000", "grey"))+
    stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test",label = "p.signif", hide.ns = FALSE, label.y = quantile(df_unstimulated_all$shannons.diversity, probs = c(0.9)), size = 6)+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
    # expand_limits(y = max(df_unstimulated_all$shannons.diversity)+3)+
    labs(x = "Timepoint", y = "Shannon's Diversity", title = paste(rep_title, "Unstimulated"), color = "GVHD Grade")+
    theme_minimal()+
    theme(plot.title = element_text(size = 18), axis.title.x = element_text(size = 20), axis.title.y = element_text(size = 14), axis.text = element_text(size = 12), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black"), axis.ticks = element_line(color = "black"),  axis.ticks.length = unit(0.25, "cm"))+
    theme(legend.position = "none")
  # 
  unstim_3_plot_shannon <- ggplot(df_unstimulated_3, aes(x = timepoint, y = shannons.diversity, shape = PatientID))+
    geom_beeswarm(aes(color = GVHD_GRADE_1), size = 2)+
    geom_line(aes(group = PatientID, color = GVHD_GRADE_1), linewidth = 1.5)+
    scale_color_manual(values = c("#005A8F", "#B85000", "grey"))+
    scale_y_continuous(labels = scales::number_format(accuracy = 1))+
    stat_compare_means(comparisons = list(c("unstimulated", "<= 3")), method = "wilcox.test", paired = TRUE, label = "p.signif", hide.ns = FALSE, group.by = "PatientID", size = 12, label.y = quantile(df_unstimulated_3$shannons.diversity, probs = c(0.9)))+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
    # expand_limits(y = max(df_unstimulated_3$shannons.diversity)+3)+
    labs(x = "Timepoint", y = "Shannon's Diversity", title = paste(rep_title, "Unstimulated"), color = "GVHD Grade")+
    theme_minimal()+
    theme(plot.title = element_text(size = 18), axis.title.x = element_text(size = 20), axis.title.y = element_text(size = 14), axis.text = element_text(size = 12), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.line = element_line(color = "black"), axis.ticks = element_line(color = "black"),  axis.ticks.length = unit(0.25, "cm"))+
    theme(legend.position = "none")

  earliest_plot_shannon <- ggplot(df_early_timepoint_all[df_early_timepoint_all$timepoint == "Earliest Timepoint",], aes(x = PTCy, y = shannons.diversity, fill = PTCy))+
    geom_boxplot()+
    geom_beeswarm(aes(color = GVHD_GRADE_1))+
    scale_color_manual(values = c("#09BB8C", "#005A8F", "#B85000", "grey"))+
    scale_y_continuous(labels = scales::number_format(accuracy = 1))+
    stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test",label = "p.format", hide.ns = FALSE, label.y = quantile(df_early_timepoint_all$shannons.diversity, probs = c(0.90)), size = 12)+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
    # expand_limits(y = max(df_early_timepoint_all$shannons.diversity)+3)+
    labs(x = "Timepoint", y = "Shannon's Diversity", title = paste(rep_title, "Earliest Timepoint"), color = "GVHD Grade")+
    theme_pubr()+
    theme(legend.position = "none")

  # df_early_timepoint <- df_early_timepoint%>%
  #   filter(PatientID %in% unique(df_timepoint$PatientID))
  df_timepoint_ptcy <- df_timepoint%>%filter(PatientID %in% unique(df_timepoint$PatientID))
  
  early_plot_shannon <- ggplot(df_unstimulated_3[df_unstimulated_3$timepoint != "Donor",], aes(x = timepoint, y = shannons.diversity, shape = PatientID, group = PatientID))+
    geom_line(aes(group = PatientID, color = GVHD_GRADE_1), linewidth = 1.5)+
    geom_point(aes(color = GVHD_GRADE_1), size = 6)+
    scale_y_continuous(labels = scales::number_format(accuracy = 1))+
    stat_compare_means(comparisons = list(c("<= 3", "Earliest Timepoint")), paired = T, label = "p.signif", method.args = list(alternative = "greater"), size = 6, label.y = quantile(df_unstimulated_3[df_unstimulated_3$timepoint != "Donor",]$shannons.diversity, probs = c(0.9)))+
    scale_color_manual(values = c( "#005A8F", "#B85000", "grey"))+
    labs(x = "Timepoint", y = "Shannon's Diversity", title = rep_title, color = "GVHD Grade")+
    theme_pubr()+
    theme(legend.position = "right")#, axis.text.x.bottom = element_text(angle = 45, vjust = 0))
  
  
  early_plot_n_clone <- ggplot(df_unstimulated_3[df_unstimulated_3$timepoint != "Donor",], aes(x = timepoint, y = n_clone, shape = PatientID))+
    geom_line(aes(group = PatientID, color = GVHD_GRADE_1), linewidth = 1.5)+
    geom_point(aes(color = GVHD_GRADE_1), size = 6)+
    scale_y_continuous(labels = scales::number_format(accuracy = 1))+
    stat_compare_means(comparisons = list(c("<= 3", "Earliest Timepoint")), method = "wilcox.test", paired = TRUE, label = "p.signif", hide.ns = FALSE, group.by = "PatientID", method.args = list(alternative = "greater"), size = 6, label.y = quantile(df_unstimulated_3[df_unstimulated_3$timepoint != "Donor",]$n_clone, probs = c(0.90)))+
    scale_color_manual(values = c("#005A8F", "#B85000", "grey"))+
    # expand_limits(y = max(df_timepoint$cum_freq)+.05)+
    labs(x = "Timepoint", y = "# of Clones", title = rep_title, color = "GVHD Grade")+
    # coord_cartesian(ylim = c(0,0.4))+
    theme_minimal()+
    theme_pubr()+
    theme(legend.position = "none")
  
  early_plot_avg_freq <- ggplot(df_unstimulated_3[df_unstimulated_3$timepoint != "Donor",], aes(x = timepoint, y = avg_freq, shape = PatientID))+
    geom_line(aes(group = PatientID, color = GVHD_GRADE_1), linewidth = 1.5)+
    geom_point(aes(color = GVHD_GRADE_1), size = 6)+
    scale_y_continuous(labels = scales::number_format(accuracy = 0.0001))+
    stat_compare_means(comparisons = list(c("<= 3", "Earliest Timepoint")), method = "wilcox.test", paired = TRUE, label = "p.signif", hide.ns = FALSE, group.by = "PatientID", method.args = list(alternative = "greater"), size = 6, label.y = quantile(df_unstimulated_3[df_unstimulated_3$timepoint != "Donor",]$avg_freq, probs = c(0.90)))+
    scale_color_manual(values = c("#005A8F", "#B85000", "grey"))+
    # expand_limits(y = max(df_timepoint$cum_freq)+.05)+
    labs(x = "Timepoint", y = "Average Clone Frequency", title = rep_title, color = "GVHD Grade")+
    # coord_cartesian(ylim = c(0,0.4))+
    theme_minimal()+
    theme_pubr()+
    theme(legend.position = "none")
  



  # group.by = "PatientID"method = "kruskal.test"
unstim_plot_freqs[[i]] <- unstim_plot_freq
unstim_3_plot_freqs[[i]] <- unstim_3_plot_freq
early_plot_freqs[[i]] <- early_plot_freq
earliest_plot_freqs[[i]] <- earliest_plot_freq
unstim_plot_shannons[[i]] <- unstim_plot_shannon
unstim_3_plot_shannons[[i]] <- unstim_3_plot_shannon
early_plot_shannons[[i]] <- early_plot_shannon
earliest_plot_shannons[[i]] <- earliest_plot_shannon
early_plot_n_clones[[i]] <- early_plot_n_clone
early_plot_avg_freqs[[i]] <- early_plot_avg_freq
}


# 
early_plot_freqs[[1]] <-  early_plot_freqs[[1]]+theme(legend.position = "none")
early_plot_freqs[[2]]  <- early_plot_freqs[[2]]+theme(legend.position = "none")
early_plot_shannons[[1]] <- early_plot_shannons[[1]]+theme(legend.position = "none")
legend_for_plot <- cowplot::get_legend(early_plot_shannons[[2]])
early_plot_shannons[[2]] <- early_plot_shannons[[2]]+theme(legend.position = "none")
unstim_plot_freq_m <- grid.arrange(unstim_plot_freqs[[1]], unstim_plot_freqs[[2]], unstim_plot_freqs[[3]], nrow = 1)
unstim_3_plot_freqs_m <- grid.arrange(unstim_3_plot_freqs[[1]], unstim_3_plot_freqs[[2]], unstim_3_plot_freqs[[3]], nrow = 1)
early_plot_top <- plot_grid(early_plot_freqs[[1]], early_plot_n_clones[[1]], early_plot_avg_freqs[[1]], early_plot_shannons[[1]], nrow = 1, rel_widths = c(1,1,1,1), align = "h")
early_plot_bot <- plot_grid( early_plot_freqs[[2]], early_plot_n_clones[[2]], early_plot_avg_freqs[[2]], early_plot_shannons[[2]], nrow = 1, rel_widths = c(1,1,1,1), align = "h")
early_plot_w0_leg <- plot_grid(early_plot_top, early_plot_bot, nrow = 2)
early_plot <- plot_grid(early_plot_w0_leg, legend_for_plot, nrow = 1, rel_widths = c(1, .1))
plot(early_plot)
# 
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/PTCyD3VEarliestTP.png'), early_plot,  height = 8, width=16, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/PTCyD3VEarliestTP.pdf'), early_plot, height = 8, width = 16, dpi = 600)


```
```{r}
for(rep_div in c("allo")){#}, "nonallo", "all")){
  if (rep_div == "allo"){
    i = 1
    rep_title = "Alloreactive"
  }else if(rep_div == "nonallo"){ 
    i = 2
    rep_title = "Non-Alloreactive"
  }else{
    i = 3
    rep_title = "All"
  }
  patient_data_allo <- dplyr::full_join(patient_data_all, meta_data,  by = "PatientID")%>%
    mutate(timepoint = SampleID)%>%
    select(-c(SampleID))
  colnames(patient_data_allo) <- sapply(colnames(patient_data_allo), function(x){str_replace_all(x, "-", "")})
  patient_data_allo <- patient_data_allo[patient_data_allo$Rep_div == rep_div,]
  patient_data_allo <- patient_data_allo[which(!is.na(patient_data_allo$timepoint)),]
  patient_data_allo[patient_data_allo$timepoint == "MLRCFSElo",]$timepoint  <- "1"
  patient_data_allo[patient_data_allo$timepoint == "unstimulated",]$timepoint  <- "0"
  patient_data_allo_copy <- patient_data_allo%>%
    mutate(timepoint = strtoi(timepoint))
  patient_data_allo_copy <- patient_data_allo_copy[which(!is.na(patient_data_allo_copy$timepoint)),]
  
  
  patient_data_allo_copy <- patient_data_allo_copy%>%
    mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1))%>%
    mutate(GVHD_GRADE_1 = factor(GVHD_GRADE_1, levels = c("No", "Mild", "Severe")))%>%
    replace_na(list(shannons.diversity = 0, shannons.entropy = 0))#%>%
    # mutate(PTCy = ifelse(grepl("PTCy", PTCy), "PTCy", "No PTCy"))
  
  df_unstimulated <- patient_data_allo_copy[patient_data_allo_copy$timepoint == 0,]%>%
    mutate(timepoint = "Donor")%>%
    filter(PTCy == "PTCy")
  
  df_unstimulated_all <- patient_data_allo_copy[patient_data_allo_copy$timepoint == 0,]%>%
    mutate(timepoint = "Donor")
  
  
  patient_data_allo_copy <- patient_data_allo_copy[patient_data_allo_copy$timepoint >= 2,]
  # patient_data_allo <- patient_data_allo[!is.na(patient_data_allo$ID),]
  
  df_early_timepoint_all <- get_early_timepoint(patient_data_allo_copy)%>%
    mutate(timepoint = ifelse(timepoint <= 3, "<= 3", "Earliest Timepoint"))%>%
    mutate(timepoint = factor(timepoint, levels = c("<= 3", "Earliest Timepoint")))%>%
    group_by(PatientID,GVHD_GRADE_1, timepoint)
  
  df_early_timepoint <- df_early_timepoint_all %>%
    filter(PTCy == "PTCy")
  
  df_early_timepoint <- df_early_timepoint[df_early_timepoint$PatientID != "R070",]
  
  data_CD48_all_expand_adj <- dplyr::full_join(data_CD48_all_expand, clinical_data, by = 'PatientID')
  data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!is.na(data_CD48_all_expand_adj$cdr3_amino_acid),]
  data_CD48_all_expand_adj <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$PatientID != "R070",]
  
  df_timepoint <- get_timepoint(patient_data_allo_copy)%>%
    mutate(timepoint = ifelse(timepoint <= 3 & timepoint >=2, "<= 3", "Earliest Timepoint"))%>%
    # mutate(GVHD_GRADE_1 = ifelse(grepl("Severe",GVHD_GRADE_1), "Severe", GVHD_GRADE_1), GVHD_GRADE_2 = ifelse(grepl("Severe",GVHD_GRADE_2), "Severe", GVHD_GRADE_2))%>%
    mutate(timepoint = factor(timepoint, levels = c("Donor", "<= 3", "Earliest Timepoint")))%>%
    filter(PTCy == "PTCy")#%>%
    # group_by(PatientID,GVHD_GRADE_1, timepoint)
  df_timepoint <- df_timepoint[df_timepoint$PatientID != 70,]
  

  df_unstimulated_3 <- bind_rows(df_unstimulated%>%filter(PatientID %in% unique(df_timepoint$PatientID)), df_early_timepoint)%>%
    mutate(timepoint = factor(timepoint, levels = c("Donor", "<= 3", "Earliest Timepoint")))%>%
    filter(GVHD_GRADE_1 != "No")%>%
    group_by(PatientID)%>%
    mutate(n_in = n())%>%
    filter(n_in >= 3)%>%
    ungroup()%>%
    replace_na(list(avg_freq = 0))
  
  df_unstimulated_filt <- df_unstimulated_3%>%filter(timepoint != "Donor")%>%
    select(PatientID, GVHD_GRADE_1, shannons.diversity, timepoint)
  View(df_unstimulated_filt)
  
  results <- t.test(as.formula("shannons.diversity ~ timepoint"), data = df_unstimulated_filt, exact = F, paired = T)
  print(results)
}
```

```{r}
all_data_squished <- bind_rows(lapply(data, bind_rows))
all_pat <- all_data_squished%>%
    filter(timepoint == "3" | timepoint == "2")
patients <- unique(all_pat$PatientID)
pat_idx <- which(unique(data_squished$PatientID) %in% patients)
data_allo3 <- vector("list", length(patients))
for(i in seq(1,length(pat_idx),1)){
  immdata2 <- data[[pat_idx[i]]]
  immdata2_squished <- bind_rows(data[[pat_idx[i]]])
  
  # allo_cd48 <- all_pat%>%
  #   filter(PatientID == patients[i])%>%
  #   subset(select = c(unique_identifiers, "cell_type", "allo"))
  immdata_3 <- immdata2_squished%>%
    filter(timepoint == "3" | timepoint == "2")
  
  immdata_unstim <- immdata2_squished%>%
    filter(timepoint == "unstimulated")
  
  data_3_allo <- full_join(immdata_unstim, immdata_3, by = c(unique_identifiers, "PatientID"),
                           suffix = c(".unstim", ".3"))%>%
    replace_na(list(templates.unstim = 0, templates.3 = 0, frequency.unstim = 0, templates.3 = 0))%>%
    filter(frequency.3 >= 2*frequency.unstim & frequency.3 >= 1e-5)%>%
    mutate(cell_type = cell_type.unstim)%>%
    replace_na(list(cell_type = "CD4CD8"))%>%
    select(-c("cell_type.unstim", "cell_type.3"))
  
  data_3_allo_clone <- data_3_allo %>%
    select(c(unique_identifiers))
  data_3 <- lapply(immdata2, function(x) {
      x%>%
      dplyr::right_join(data_3_allo_clone, by =  c(unique_identifiers))%>%
      replace_na(list(frequency = 0, templates = 0))#%>%
      #mutate(cell_type = cell_type.y, timepoint = na.omit(timepoint)[1], PatientID = na.omit(PatientID)[1], tag = paste(cdr3_amino_acid, v_resolved, j_resolved))%>%
      #select(-c(cell_type.x, cell_type.y))
      
    }
  )
  
  data_allo3[[i]] <- data_3
    
}

```

```{r}
res_allo <- data_allo_squished_w0 %>%
  group_by(PatientID, timepoint,cell_type)%>%
  summarise(cum_freq = sum(frequency))%>%
  filter(cell_type != "CD4CD8")

res_nonallo <- data_nonallo_squished_w0 %>%
  group_by(PatientID, timepoint, cell_type)%>%
  summarise(cum_freq = sum(frequency))%>%
  filter(cell_type != "CD4CD8")
```

```{r, fig.height = 3.2, fig.width=5, dpi = 600}
nonallo_all_CD4 <- bind_rows(lapply(all_nonallo_div_CD4, bind_rows))%>%
  filter(!is.na(shannons.diversity))%>%
  mutate(ids = str_split(ID, "_"), rep_div = "nonallo_CD4")%>%
  mutate(PatientID = sapply(ids, function(x){x[[1]]}), timepoint = sapply(ids, function(x){x[[2]]}))%>%
  full_join(res_nonallo[res_nonallo$cell_type == "CD4",], by = c("PatientID", "timepoint"))

nonallo_all_CD8 <- bind_rows(lapply(all_nonallo_div_CD8, bind_rows))%>%
  filter(!is.na(shannons.diversity))%>%
  mutate(ids = str_split(ID, "_"), rep_div = "nonallo_CD8")%>%
  mutate(PatientID = sapply(ids, function(x){x[[1]]}), timepoint = sapply(ids, function(x){x[[2]]}))%>%
  full_join(res_nonallo[res_nonallo$cell_type == "CD8",], by = c("PatientID", "timepoint"))

allo_all_CD4 <- bind_rows(lapply(all_allo_div_CD4, bind_rows))%>%
  filter(!is.na(shannons.diversity))%>%
  mutate(ids = str_split(ID, "_"), rep_div = "allo_CD4")%>%
  mutate(PatientID = sapply(ids, function(x){x[[1]]}), timepoint = sapply(ids, function(x){x[[2]]}))%>%
  full_join(res_allo[res_allo$cell_type == "CD4",], by = c("PatientID", "timepoint"))

allo_all_CD8 <- bind_rows(lapply(all_allo_div_CD8, bind_rows))%>%
  filter(!is.na(shannons.diversity))%>%
  mutate(ids = str_split(ID, "_"), rep_div = "allo_CD8")%>%
  mutate(PatientID = sapply(ids, function(x){x[[1]]}), timepoint = sapply(ids, function(x){x[[2]]}))%>%
  full_join(res_allo[res_allo$cell_type == "CD8",], by = c("PatientID", "timepoint"))

all_CD48_div <- bind_rows(nonallo_all_CD4, nonallo_all_CD8, allo_all_CD4, allo_all_CD8)%>%
  full_join(meta_data_clin, by = c("PatientID"))%>%
  filter(!is.na(ID))

all_CD48_div[all_CD48_div$timepoint == "MLRCFSElo",]$timepoint  <- "1"
all_CD48_div[all_CD48_div$timepoint == "unstimulated",]$timepoint  <- "0"
all_CD48_div <- all_CD48_div%>%
    mutate(timepoint = strtoi(timepoint), GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1))

for(r_div in c("allo_CD4", "allo_CD8", "nonallo_CD4", "nonallo_CD8")){
  if (r_div == "allo_CD4"){
    rep_title = "CD4+ Alloreactive"
  }else if (r_div == "allo_CD8"){
    rep_title = "CD8+ Alloreactive"
  }else if (r_div == "nonallo_CD4"){
    rep_title = "CD4+ Non-Alloreactive"
  } else{
    rep_title = "CD8+ Non-Alloreactive"
  }
  
  rep_div_CD48_div <- all_CD48_div%>%
    filter(rep_div == r_div)
  
  df_unstimulated <- rep_div_CD48_div[rep_div_CD48_div$timepoint == 0,]%>%
    mutate(timepoint = "unstimulated")%>%
    filter(PTCy == "PTCy")
  
  df_unstimulated_all <- rep_div_CD48_div[rep_div_CD48_div$timepoint == 0,]%>%
    mutate(timepoint = "unstimulated")%>%
    filter(!is.na(ID))
  
  
  patient_data_allo_copy <- rep_div_CD48_div[rep_div_CD48_div$timepoint >= 2,]
  # patient_data_allo <- patient_data_allo[!is.na(patient_data_allo$ID),]
  
  df_early_timepoint_all <- get_early_timepoint(patient_data_allo_copy)%>%
    mutate(timepoint = ifelse(timepoint <= 3, "<= 3", "Earliest Timepoint"))%>%
    mutate(timepoint = factor(timepoint, levels = c("<= 3", "Earliest Timepoint")), GVHD_GRADE_1 = factor(GVHD_GRADE_1, levels = c("No", "Mild", "Severe")))%>%
    group_by(PatientID,GVHD_GRADE_1, timepoint)
  
  df_early_timepoint <- df_early_timepoint_all %>%
    filter(PTCy == "PTCy")
  
  df_early_timepoint <- df_early_timepoint[df_early_timepoint$PatientID != "R070",]

  
  df_timepoint <- get_timepoint(patient_data_allo_copy)%>%
    mutate(timepoint = ifelse(timepoint <= 3 & timepoint >=2, "<= 3", "Earliest Timepoint"))%>%
    mutate(GVHD_GRADE_1 = ifelse(grepl("Severe",GVHD_GRADE_1), "Severe", GVHD_GRADE_1), GVHD_GRADE_2 = ifelse(grepl("Severe",GVHD_GRADE_2), "Severe", GVHD_GRADE_2))%>%
    mutate(timepoint = factor(timepoint, levels = c("<= 3", "Earliest Timepoint")), GVHD_GRADE_1 = factor(GVHD_GRADE_1, levels = c("No", "Mild", "Severe")))%>%
    filter(PTCy == "PTCy")%>%
    group_by(PatientID,GVHD_GRADE_1, timepoint)
  df_timepoint <- df_timepoint[df_timepoint$PatientID != 70,]
  

  df_unstimulated_3 <- bind_rows(df_unstimulated%>%filter(PatientID %in% unique(df_timepoint$PatientID)), df_early_timepoint[df_early_timepoint$timepoint == "<= 3",])%>%
    mutate(timepoint = factor(timepoint, levels = c("unstimulated", "<= 3")))
  
  
   unstim_plot_freq <- ggplot(df_unstimulated_all, aes(x = PTCy, y = cum_freq, fill = PTCy))+
    geom_boxplot()+
    geom_beeswarm(aes(color = GVHD_GRADE_1))+
    scale_color_manual(values = c("#09BB8C", "#005A8F", "#B85000", "grey"))+
    stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test", label = "p.format", hide.ns = FALSE, label.y = max(df_unstimulated_all$cum_freq)-.01)+
    labs(y = "Frequency", title = paste(rep_title, "Unstimulated"), color = "GVHD Grade")+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
    expand_limits(y = max(df_unstimulated_all$cum_freq)+.05)+
    theme_minimal()+
    theme(plot.title = element_text(size = 18), axis.title.x = element_text(size = 14), axis.title.y = element_text(size = 14), axis.text = element_text(size = 12))
  
  
  unstim_3_plot_freq <- ggplot(df_unstimulated_3, aes(x = timepoint, y = cum_freq, shape = PatientID))+
    geom_point(aes(color = GVHD_GRADE_1), size = 2)+
    geom_line(aes(group = PatientID, color = GVHD_GRADE_1), linewidth = 1.5)+
    scale_color_manual(values = c( "#005A8F", "#B85000", "grey"))+
    stat_compare_means(comparisons = list(c("unstimulated", "<= 3")), method = "wilcox.test", paired = TRUE, label = "p.format", hide.ns = FALSE, group.by = "PatientID", label.y = max(df_unstimulated_3$cum_freq)-.01)+
    labs(x = "Timepoint", y = "Frequency", title = paste(rep_title, "Unstimulated"), color = "GVHD Grade")+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
    expand_limits(y = max(df_unstimulated_3$cum_freq)+.05)+
    theme_minimal()+
    theme(plot.title = element_text(size = 18), axis.title.x = element_text(size = 14), axis.title.y = element_text(size = 14), axis.text = element_text(size = 12))
  
  earliest_plot_freq <- ggplot(df_early_timepoint_all[df_early_timepoint_all$timepoint == "Earliest Timepoint",], aes(x = PTCy, y = cum_freq, fill = PTCy))+
    geom_boxplot()+
    geom_beeswarm(aes(color = GVHD_GRADE_1))+
    scale_color_manual(values = c("#09BB8C", "#005A8F", "#B85000", "grey"))+
    stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test",label = "p.format", hide.ns = FALSE, label.y = max(df_early_timepoint_all$cum_freq)-.01)+
    labs(y = "Frequency", title = paste(rep_title, "Earliest Timepoint"), color = "GVHD Grade")+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
    expand_limits(y = max(df_early_timepoint_all$cum_freq)+.05)+
    theme_minimal()+
    theme(plot.title = element_text(size = 18), axis.title.x = element_text(size = 14), axis.title.y = element_text(size = 14), axis.text = element_text(size = 12))
  
  early_plot_freq <- ggplot(df_timepoint, aes(x = timepoint, y = cum_freq, shape = PatientID))+
    geom_line(aes(group = PatientID, color = GVHD_GRADE_1), linewidth = 1.5)+
    geom_point(aes(color = GVHD_GRADE_1), size = 2)+
    stat_compare_means(comparisons = list(c("<= 3", "Earliest Timepoint")), method = "wilcox.test", paired = TRUE, label = "p.format", hide.ns = FALSE, group.by = "PatientID", label.y = max(df_timepoint$cum_freq)-.01)+
    scale_color_manual(values = c("#005A8F", "#B85000", "grey"))+
    expand_limits(y = max(df_timepoint$cum_freq)+.05)+
    labs(x = "Timepoint", y = "Frequency", title = rep_title, color = "GVHD Grade")+
    theme_minimal()+
    theme(plot.title = element_text(size = 18), axis.title.x = element_text(size = 14), axis.title.y = element_text(size = 14), axis.text = element_text(size = 12))
  
  
  unstim_plot_shannon <- ggplot(df_unstimulated_all, aes(x = PTCy, y = shannons.diversity, fill = PTCy))+
    geom_boxplot()+
    geom_beeswarm(aes(color = GVHD_GRADE_1))+
    scale_color_manual(values = c("#09BB8C", "#005A8F", "#B85000", "grey"))+
    stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test",label = "p.format", hide.ns = FALSE, label.y = max(df_unstimulated_all$shannons.diversity)+2)+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
    expand_limits(y = max(df_unstimulated_all$shannons.diversity)+3)+
    labs(y = "Shannon's Diversity", title = paste(rep_title, "Unstimulated"), color = "GVHD Grade")+
    theme_minimal()+
    theme(plot.title = element_text(size = 18), axis.title.x = element_text(size = 14), axis.title.y = element_text(size = 14), axis.text = element_text(size = 12))
  
  unstim_3_plot_shannon <- ggplot(df_unstimulated_3, aes(x = timepoint, y = shannons.diversity, shape = PatientID))+
    geom_beeswarm(aes(color = GVHD_GRADE_1), size = 2)+
    geom_line(aes(group = PatientID, color = GVHD_GRADE_1), linewidth = 1.5)+
    scale_color_manual(values = c("#005A8F", "#B85000", "grey"))+
    stat_compare_means(comparisons = list(c("unstimulated", "<= 3")), method = "wilcox.test", paired = TRUE, label = "p.format", hide.ns = FALSE, group.by = "PatientID", label.y = max(df_unstimulated_3$shannons.diversity)+2)+    
    scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
    expand_limits(y = max(df_unstimulated_3$shannons.diversity)+3)+
    labs(x = "Timepoint", y = "Shannon's Diversity", title = paste(rep_title, "Unstimulated"), color = "GVHD Grade")+
    theme_minimal()+
    theme(plot.title = element_text(size = 18), axis.title.x = element_text(size = 14), axis.title.y = element_text(size = 14), axis.text = element_text(size = 12))
  
  earliest_plot_shannon <- ggplot(df_early_timepoint_all[df_early_timepoint_all$timepoint == "Earliest Timepoint",], aes(x = PTCy, y = shannons.diversity, fill = PTCy))+
    geom_boxplot()+
    geom_beeswarm(aes(color = GVHD_GRADE_1))+
    scale_color_manual(values = c("#09BB8C", "#005A8F", "#B85000", "grey"))+
    stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test",label = "p.format", hide.ns = FALSE, label.y = max(df_early_timepoint_all$shannons.diversity)+2)+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"))+
    expand_limits(y = max(df_early_timepoint_all$shannons.diversity)+3)+
    labs(y = "Shannon's Diversity", title = paste(rep_title, "Earliest Timepoint"), color = "GVHD Grade")+
    theme_minimal()+
    theme(plot.title = element_text(size = 18), axis.title.x = element_text(size = 14), axis.title.y = element_text(size = 14), axis.text = element_text(size = 12))
      
  # df_early_timepoint <- df_early_timepoint%>%
  #   filter(PatientID %in% unique(df_timepoint$PatientID))
  df_timepoint_ptcy <- df_timepoint%>%filter(PatientID %in% unique(df_timepoint$PatientID))
  
  early_plot_shannon <- ggplot(df_timepoint_ptcy, aes(x = timepoint, y = shannons.diversity, shape = PatientID))+
    geom_line(aes(group = PatientID, color = GVHD_GRADE_1), linewidth = 1.5)+
    geom_beeswarm(aes(color = GVHD_GRADE_1), size = 2)+
    stat_compare_means(comparisons = list(c("<= 3", "Earliest Timepoint")), method = "wilcox.test", paired = TRUE, label = "p.format", hide.ns = FALSE, group.by = "PatientID", label.y = max(df_timepoint_ptcy$shannons.diversity)+2)+
    scale_color_manual(values = c( "#005A8F", "#B85000", "grey"))+
    labs(x = "Timepoint", y = "Shannon's Diversity", title = rep_title, color = "GVHD Grade")+
    expand_limits(y = max(df_timepoint_ptcy$shannons.diversity)+3)+
    theme_minimal()+
    theme( plot.title = element_text(size = 18, margin = margin(b = 20, unit = "pt")), axis.title.x = element_text(size = 14), axis.title.y = element_text(size = 14), axis.text = element_text(size = 12))
    
  plot(unstim_plot_freq)
  plot(unstim_3_plot_freq)
  plot(early_plot_freq)
  plot(earliest_plot_freq)
  plot(unstim_plot_shannon)
  plot(unstim_3_plot_shannon)
  plot(early_plot_shannon)
  plot(earliest_plot_shannon)
  
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/',r_div,"unstim_plot_freq.png"), unstim_plot_freq, width = 5, height = 3.2, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/',r_div,"unstim_3_plot_freq.png"), unstim_3_plot_freq, width = 5, height = 3.2, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/',r_div,"early_plot_freq.png"), early_plot_freq, width = 5, height = 3.2, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/',r_div,"earliest_plot_freq.png"), earliest_plot_freq, width = 5, height = 3.2, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/',r_div,"unstim_plot_shannon.png"), unstim_plot_shannon, width = 5, height = 3.2, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/',r_div,"unstim_3_plot_shannon.png"), unstim_3_plot_shannon, width = 5, height = 3.2, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/',r_div,"early_plot_shannon.png"), early_plot_shannon, width = 5, height = 3.2, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/',r_div,"earliest_plot_shannon.png"), earliest_plot_shannon, width = 5, height = 3.2, dpi = 600)
}
```



```{r}
clone_fig <- function(immdata, type, title, alpha = 0.3) {
  #' Visualize clone frequencies over time
  #'
  #' @description This function visualizes the frequency of clones over time for different groups using scatter and line plots.
  #'
  #' @param immdata A dataframe containing immunosequencing data with `timepoint`, `frequency`, `tag`, and specified `type`.
  #' @param type A character string specifying the grouping variable for the analysis.
  #' @param title A character string for the plot title.
  #' @param alpha A numeric value indicating the transparency level of points and lines in the plot (default is 0.3).
  #' @usage clone_fig(immdata, type, title, alpha)
  #' @return A ggplot object showing the clone frequency over time.
  
  # Convert timepoints to integer and remove NA values
  immdata <- immdata %>%
    mutate(timepoint = strtoi(timepoint)) %>%
    filter(!is.na(timepoint))
  
  # Calculate average and standard deviation of frequency per group and timepoint
  immdata_stat <- immdata %>%
    group_by(across(all_of(c(type, "timepoint")))) %>%
    summarise(
      avg_freq = mean(frequency, na.rm = TRUE),
      std_freq = sd(frequency, na.rm = TRUE),
      .groups = 'drop'
    )
  
  # Generate the plot
  fig <- ggplot(immdata, aes(x = timepoint, y = frequency, color = .data[[type]], group = tag)) +
    geom_point(size = 1, alpha = alpha) + 
    geom_line(alpha = alpha) +
    labs(title = title, x = "Timepoint", y = "Frequency") +
    scale_y_continuous(trans = "log10", labels = scales::number_format(accuracy = 0.00001)) +
    scale_colour_discrete(na.translate = FALSE) +
    scale_shape_discrete(na.translate = FALSE) +
    guides(color = guide_legend(override.aes = list(alpha = 1))) +
    theme_pubr() +
    theme(
      axis.text.x = element_text(angle = 60, vjust = 0.5),
      legend.text = element_markdown(size = 20)
    )
  
  return(fig)
}

```


```{r, warning=FALSE, fig.width=8, fig.height = 4}


data_CD48_all_expand_adj <- data_CD48_all_expand
data_CD48_all_expand_adj$cell_type <- gsub(" ", "",data_CD48_all_expand_adj$cell_type )
data_CD48_all_expand_adj <- dplyr::full_join(data_CD48_all_expand_adj, clinical_data, by = "PatientID")%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1))%>%
  mutate(GVHD_GRADE_1 = factor(GVHD_GRADE_1, levels = c("No", "Mild", "Severe")))


All_by_grade <- clone_fig(data_CD48_all_expand_adj, 'GVHD_GRADE_1', "", alpha = 0.1)
All_by_grade_fig_ex <- All_by_grade + theme(text = element_text(size = 20))+
  scale_color_manual(values = c("#09BB8C", "#005A8F", "#B85000", "grey"), labels = c(
      No = "<span style='color:#09BB8C;'>No</span>",
      Mild = "<span style='color:#005A8F;'>Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  coord_cartesian(xlim = c(0,365))+
  labs(y = "Expanded Alloreactive\n
       Clone Frequency", x = "Days", color = "GVHD Grade")+
  theme(axis.title.y = element_text(lineheight = 0.5))


All_by_ptcy <- clone_fig(data_CD48_all_expand_adj, 'PTCy', "Expanded Alloreactive Clones", alpha = 0.3)
All_by_ptcy_fig_ex <- All_by_ptcy + 
  theme(text = element_text(size = 20)) + 
  scale_color_manual(values = c("#E6BE60", "#A483AF"), 
                    labels = c(
                      `No PTCy` = "<span style='color:#E6BE60;'>No PTCy</span>",
                      PTCy = "<span style='color:#A483AF;'>PTCy</span>")) + 
  coord_cartesian(xlim = c(0,365)) + 
  labs(y = "Expanded Alloreactive\n 
       Clone Frequency", x = "Days", color = "PTCy")+
  theme(axis.title.y = element_text(lineheight = 0.5))


plot(All_by_grade_fig_ex)
plot(All_by_ptcy_fig_ex)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/All_by_grade_fig.png'), All_by_grade_fig_ex, width = 8, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/All_by_ptcy_fig.png'), All_by_ptcy_fig_ex, width = 8, height = 5, dpi = 600)


ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/All_by_grade_fig.pdf'), All_by_grade_fig_ex, width = 8, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/All_by_ptcy_fig.pdf'), All_by_ptcy_fig_ex, width = 8, height = 5, dpi = 600)

```

```{r}

avg_clone_fig <- function(immdata, type, title = "", num_cuts = 5, groupings = NA, colors = c("#E6BE60", "#A483AF"), method = "hochberg", legend_title = "", ylab = "") {
  #' Visualize average clone frequency over time
  #'
  #' @description This function processes immunosequencing data to compute average clone frequencies over time, divides them into bins, performs statistical tests, and visualizes the results.
  #'
  #' @param immdata A dataframe containing immunosequencing data with `timepoint`, `PatientID`, and specified `type`.
  #' @param type A character string specifying the grouping variable for the analysis.
  #' @param title A character string for the plot title.
  #' @param num_cuts An integer indicating the number of bins to divide the timepoints into (default is 5).
  #' @param groupings Optional grouping variable for further subsetting (default is NA).
  #' @param colors A character vector specifying colors for the plot (default is c("#E6BE60", "#A483AF")).
  #' @param method A character string specifying the p-value adjustment method (default is "hochberg").
  #' @param legend_title A character string for the legend title.
  #' @param ylab A character string for the y-axis label in the plot.
  #' @usage avg_clone_fig(immdata, type, title, num_cuts, groupings, colors, method, legend_title, ylab)
  #' @return A list containing the plot, statistical test results, processed data, and summary statistics.

  # Convert timepoints to integer and filter data
  immdata <- immdata %>%
    mutate(timepoint = strtoi(timepoint)) %>%
    filter(!is.na(timepoint), timepoint <= 365)

  # Separate early timepoints (<=3 days) and remaining timepoints
  immdata_3 <- immdata %>%
    filter(timepoint <= 3, templates != 0) %>%
    mutate(timepoint = 3, bins = "3")
  
  immdata <- immdata %>%
    filter(timepoint > 3, templates != 0)

  # Create bins for timepoints
  breaks <- unique(quantile(immdata$timepoint, probs = 0:num_cuts/num_cuts, na.rm = TRUE))
  immdata <- immdata %>%
    mutate(bins = cut(timepoint, breaks = breaks, include.lowest = TRUE))
  
  # Create an all-timepoint reference dataset
  immdata_all_timepoint <- immdata %>%
    select(all_of(c("timepoint", "PatientID", type))) %>%
    distinct() %>%
    mutate(bins = cut(timepoint, breaks = breaks, include.lowest = TRUE))

  # Combine early and binned timepoints, setting factor levels
  if (nrow(immdata_3) > 0) {
    immdata <- bind_rows(immdata_3, immdata) %>%
      mutate(bins = factor(bins, levels = c("3", levels(immdata_all_timepoint$bins))))
  } else {
    immdata <- immdata %>%
      mutate(bins = factor(bins, levels = levels(immdata_all_timepoint$bins)))
  }

  # Aggregate data for plotting
  immdata_fig <- immdata %>%
    group_by(across(all_of(c("bins", "timepoint", "PatientID", type)))) %>%
    full_join(immdata_all_timepoint, by = c("PatientID", "bins", "timepoint", type)) %>%
    replace_na(list(frequency = 0)) %>%
    group_by(across(all_of(c("bins", "PatientID", type)))) %>%
    summarise(frequency = mean(frequency), .groups = 'drop')

  # Summarize data for error bars
  immdata_sum <- immdata_fig %>%
    group_by(across(all_of(c("bins", type)))) %>%
    summarise(m = mean(frequency), s = sd(frequency), .groups = 'drop')

  # Perform pairwise Wilcoxon tests
  stat_test <- immdata_fig %>%
    group_by(bins) %>%
    filter(n() >= 3) %>%
    pairwise_wilcox_test(as.formula(paste0("frequency ~ ", type)), p.adjust.method = method, exact = FALSE) %>%
    add_xy_position(x = "bins") %>%
    mutate(
      y.position = log10(y.position),
      p.adj = p.adjust(p, method = method),
      p.adj.signif = symnum(
        p.adj,
        cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
        symbols = c("****", "***", "**", "*", "ns")
      )
    )

  # Generate the plot
  fig <- ggplot(immdata_sum, aes(x = bins, y = m, fill = .data[[type]], group = .data[[type]], color = .data[[type]])) +
    geom_line() +
    geom_point() +
    geom_ribbon(aes(ymin = pmax(0, m - s), ymax = m + s, linetype = NA), alpha = 0.3) +
    geom_errorbar(aes(ymin = pmax(0, m - s), ymax = m + s), width = 0.1) +
    scale_x_discrete(na.translate = FALSE) +
    theme_minimal() +
    theme(
      text = element_text(size = 16),
      axis.text.x = element_text(angle = 75, vjust = 0.5, size = 12),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_blank(),
      axis.line = element_line(color = "black"),
      axis.ticks = element_line(color = "black"),
      axis.ticks.length = unit(0.25, "cm")
    ) +
    labs(title = title, y = ylab, x = "Days") +
    scale_fill_manual(values = colors) +
    scale_color_manual(values = colors)

  return(list(fig, stat_test, immdata_fig, immdata_sum)) # Return plot, statistical tests, and processed data
}


```


```{r, warning=FALSE, fig.width=8, fig.height = 5}


data_CD48_all_expand_adj <- data_CD48_all_expand
data_CD48_all_expand_adj$cell_type <- gsub(" ", "",data_CD48_all_expand_adj$cell_type )
data_CD48_all_expand_adj <- dplyr::full_join(data_CD48_all_expand_adj, clinical_data, by = "PatientID")%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1))%>%
  mutate(GVHD_GRADE_1 = factor(GVHD_GRADE_1, levels = c("No", "Mild", "Severe")))
# data_CD48_all_expand_adj <- data_CD48_all_expand_adj[data_CD48_all_expand_adj$GVHD_GRADE_1 != "Unclassified/Severe",]

All_by_grade <- avg_clone_fig(data_CD48_all_expand_adj, "GVHD_GRADE_1", title = "", num_cuts = 7, groupings = NA, colors = c("#09BB8C", "#005A8F", "#B85000", "grey"), method = "hochberg", legend_title = "", ylab = "")#clone_fig(data_CD48_all_expand_adj, 'GVHD_GRADE_1', "Expanded Alloreactive Clones", alpha = 0.1)
All_by_grade_fig <- All_by_grade[[1]] + theme(text = element_text(size = 20))+
  scale_color_manual(values = c("#09BB8C", "#005A8F", "#B85000", "grey"))+
  scale_fill_manual(values = c("#09BB8C", "#005A8F", "#B85000", "grey"))+
  labs(y = "Frequency", x = "Days", color = "GVHD Grade")

All_by_ptcy <- avg_clone_fig(data_CD48_all_expand_adj, "PTCy", title = "", num_cuts = 7, groupings = NA, colors = c("#E6BE60", "#A483AF"), method = "hochberg", legend_title = "", ylab = "")#clone_fig(data_CD48_all_expand_adj, 'PTCy', "Expanded Alloreactive Clones", alpha = 0.5)
All_by_ptcy_fig <- All_by_ptcy[[1]] + theme(text = element_text(size = 20))+
  scale_color_manual(values = c("#E6BE60", "#A483AF"))+
  labs(y = "Frequency", x= "Days", color = "PTCy")


plot(All_by_grade_fig)
plot(All_by_ptcy_fig)

```

```{r}
count_clones <- function(data, type, metric = "cum_freq", num_cuts = 5, group_by = c("PatientID","timepoint", "PTCy")){
  #' 
  #' @description Bins the clones over time.
  #' 
  #' @param data List of dataframes containing the data from immunoseq files.
  #' Dataframes are obtained through the function read_immunoseq
  #' @param type Grouping you would like to use
  #' @param metric diverseity metric you would like to use
  #' @param num_cuts, the number of bins you would like
  #' 
  #' 
  #' @usage count_clones(data, type, metric = "cum_freq", num_cuts = 5)
  #' @return updated: list(data_pre, data_post)
  #' 
  #' 
  data_proc <- data%>%
    group_by(across(all_of(c(group_by, type))))%>%
    dplyr::summarise(num_clones = sum(frequency), .groups = "keep")
  
  data_pre <- data_proc[grepl("MLR|unstim", data_proc$timepoint),] #Seperate timepoints before transplant
  
  data_proc <- data_proc %>%
    mutate(timepoint = strtoi(timepoint))
  data_proc <- data_proc[!is.na(data_proc$timepoint),]
  
  
  data_post <- data_proc[data_proc$timepoint <= 365 & data_proc$timepoint > 3,] %>%
    subset(!is.na(timepoint))%>%
    ungroup()
  
  qs <- matrix(nrow = length(unique(data_post[[type]])), ncol = num_cuts+1)
  for(i in seq_along(unique(data_post[[type]]))){
    t <- unique(data_post[[type]])[[i]]
    qs[i,] <- quantile(data_post[data_post[[type]] == t,]$timepoint, probs = 0:num_cuts/num_cuts)
  }

  qs_breaks <- apply(qs,2, max)
  
  breaks <- unique(quantile(data_post$timepoint, probs = 0:num_cuts/num_cuts))
  
  data_post <- data_post%>%
    mutate(bins = cut(timepoint,breaks = breaks, include.lowest = TRUE))
  
  return(list(data_pre, data_post))
}

num_clones <- function(data, type, metric = "cum_freq", num_cuts = 5, title, ylab = "Allo Frequency", text_size = 20, group_by = c("PatientID", "timepoint", "PTCy")) {
  #' Analyze number of clones and visualize results
  #'
  #' @description This function calculates the number of clones based on specific metrics, performs statistical comparisons between groups, and visualizes the results using boxplots with significance annotations.
  #'
  #' @param data A dataframe containing clone data, including a `SampleID` column.
  #' @param type A character string specifying the grouping variable for the analysis.
  #' @param metric A character string specifying the metric to use for clone frequency calculation (default is "cum_freq").
  #' @param num_cuts An integer indicating the number of bins to divide the timepoints into (default is 5).
  #' @param title A character string for the plot title.
  #' @param ylab A character string for the y-axis label in the plot (default is "Allo Frequency").
  #' @param text_size An integer specifying the size of text in the plot (default is 20).
  #' @param group_by A character vector specifying the grouping variables for clone counting (default is c("PatientID", "timepoint", "PTCy")).
  #' @usage num_clones(data, type, metric, num_cuts, title, ylab, text_size, group_by)
  #' @return A list containing the boxplot with significance annotations, the Wilcoxon test results, and the processed data.

  results <- count_clones(data, type, metric, num_cuts, group_by = group_by)
  data_post <- results[[2]]
  data_post$type_num <- as.numeric(factor(data_post[[type]]))

  # Filter data_post for non-zero clone numbers
  data_post_n0 <- data_post %>%
    filter(num_clones != 0)

  # Perform Wilcoxon tests and adjust p-values
  wilcox_results <- compare_means(as.formula(paste0("num_clones ~ ", type)), data = data_post, group.by = "bins")

  stat_test <- data_post %>%
    group_by(bins) %>%
    pairwise_wilcox_test(as.formula(paste0("num_clones ~ ", type)), p.adjust.method = "hochberg") %>%
    add_xy_position(x = "bins") %>%
    mutate(
      p.adj = p.adjust(p, method = "hochberg"),
      p.adj.signif = symnum(
        p.adj,
        cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
        symbols = c("****", "***", "**", "*", "ns")
      )
    )

  print(colnames(data_post)) # Print column names of data_post for debugging

  # Create boxplot with significance annotations and beeswarm overlay
  box_plot_post <- ggboxplot(data_post, x = "bins", y = "num_clones", fill = type) +
    stat_pvalue_manual(stat_test, label = "p.adj.signif", tip.length = 0.01, step.increase = 0.00, bracket.nudge.y = -1, size = 8) +
    geom_beeswarm(data = data_post_n0, position = "dodge", dodge.width = .75, na.rm = TRUE, aes(x = bins, y = num_clones, group = .data[[type]])) +
    scale_y_continuous(trans = 'log10', labels = scales::number_format(accuracy = 0.0001)) +
    theme_pubr() +
    theme(
      text = element_text(size = text_size),
      axis.text.x = element_text(angle = 45, hjust = 1),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_blank(),
      axis.line = element_line(color = "black"),
      axis.ticks.length = unit(0.25, "cm"),
      legend.text = element_markdown()
    ) +
    labs(y = ylab, title = title, x = "Days")

  # Prepare data for output
  data_post <- data_post %>%
    mutate(timepoint = as.character(timepoint))
  data <- bind_rows(data_post)

  return(list(box_plot_post, wilcox_results, data)) # Return boxplot, Wilcoxon test results, and processed data
}

```

```{r, warning=FALSE, fig.width=9.5, fig.height = 5}

data_CD48_all_expand_adj <- data_CD48_all_expand
data_CD48_all_expand_adj$cell_type <- gsub(" ", "",data_CD48_all_expand_adj$cell_type )
data_CD48_all_expand_adj <- dplyr::full_join(data_CD48_all_expand_adj, clinical_data, by = "PatientID")
data_CD48_all_expand_adj <- data_CD48_all_expand_adj%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1))

if(remove_pat_regx != ""){
  data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!grepl(remove_pat_regx,data_CD48_all_expand_adj$PatientID),]
}

data_CD48_all_expand_adj <- data_CD48_all_expand_adj%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("No|Mild", GVHD_GRADE_1), "No/Mild", "Severe"))#%>%
  #mutate(GVHD_GRADE_1 = factor(GVHD_GRADE_1, levels = c("No/Mild", "Severe")))

type <- "GVHD_GRADE_1"


result_NoPTCy <- num_clones(subset(data_CD48_all_expand_adj, data_CD48_all_expand_adj$PTCy == "No PTCy"), type, num_cuts = 4, title = "No PTCy No and Mild vs Severe", ylab = "Alloreactive \n Cumulative Frequency")
NoPTCy <- result_NoPTCy[[1]]+
  scale_fill_manual(name = "GVHD Grade", values=c("#058b8e", "#B85000"),
                    labels = c(
      `No/Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
test_NoPTCy <- result_NoPTCy[[2]]

result_PTCy <- num_clones(subset(data_CD48_all_expand_adj, data_CD48_all_expand_adj$PTCy == "PTCy"), type, num_cuts=4, title = "PTCy No and Mild vs Severe", ylab = "Alloreactive \n Cumulative Frequency")
PTCy <- result_PTCy[[1]]+
  scale_fill_manual(name = "GVHD Grade", values=c("#058b8e", "#B85000"),
                    labels = c(
      `No/Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
test_PTCy <- result_PTCy[[2]]

result_All_byPTCy <- num_clones(data_CD48_all_expand_adj, "PTCy", num_cuts = 4, title = "PTCy vs No PTCy", ylab = "Alloreactive \n Cumulative Frequency")
All_byPTCy <- result_All_byPTCy[[1]]+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"),
                    labels = c(
      `No PTCy` = "<span style='color:#E6BE60;'>No PTCy</span>",
      PTCy = "<span style='color:#A483AF;'>PTCy</span>"))
test_All_byPTCy <- result_All_byPTCy[[2]]

result_Grade_byPTCy <- num_clones(data_CD48_all_expand_adj, type, num_cuts = 4, title = "", ylab = "Alloreactive \n Cumulative Frequency")
Grade_byPTCy <- result_Grade_byPTCy[[1]]+
  scale_fill_manual(name = "GVHD Grade", values = c("#058b8e", "#B85000"),
                    labels = c(
      `No/Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
test_Grade_byPTCy <- result_Grade_byPTCy[[2]]

data_CD48_all_expand_adj_CD4 <- data_CD48_all_expand_adj%>%
  filter(grepl("CD4",cell_type) & !grepl("CD8", cell_type))

data_CD48_all_expand_adj_CD8 <- data_CD48_all_expand_adj%>%
  filter(!grepl("CD4",cell_type) & grepl("CD8", cell_type))

result_Grade_byPTCy <- num_clones(data_CD48_all_expand_adj_CD4, type, num_cuts = 2, title = "GVHD Grade CD4+", ylab = "Alloreactive \n Cumulative Frequency")
Grade_byPTCy_CD4 <- result_Grade_byPTCy[[1]]+
  scale_fill_manual(name = "GVHD Grade", values = c("#058b8e", "#B85000"),
                    labels = c(
      `No/Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
result_Grade_byPTCy <- num_clones(data_CD48_all_expand_adj_CD8, type, num_cuts = 2, title = "GVHD Grade CD8+", ylab = "Alloreactive \n Cumulative Frequency")
Grade_byPTCy_CD8 <- result_Grade_byPTCy[[1]]+
  scale_fill_manual(name = "GVHD Grade", values = c("#058b8e", "#B85000"),
                    labels = c(
      `No/Mild` = "<span style='color:#058b8e;'>No/Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))

result_CD4_byPTCy <- num_clones(data_CD48_all_expand_adj_CD4, "PTCy", num_cuts = 4, title = "PTCy vs No PTCy CD4+", ylab = "Alloreactive \n Cumulative Frequency")
CD4_byPTCy <- result_CD4_byPTCy[[1]]+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"),
                    labels = c(
      `No PTCy` = "<span style='color:#E6BE60;'>No PTCy</span>",
      PTCy = "<span style='color:#A483AF;'>PTCy</span>"))


result_CD8_byPTCy <- num_clones(data_CD48_all_expand_adj_CD8, "PTCy", num_cuts = 4, title = "PTCy vs No PTCy CD8+", ylab = "Alloreactive \n Cumulative Frequency")
CD8_byPTCy <- result_CD8_byPTCy[[1]]+
    scale_fill_manual(values = c("#E6BE60", "#A483AF"),
                    labels = c(
      `No PTCy` = "<span style='color:#E6BE60;'>No PTCy</span>",
      PTCy = "<span style='color:#A483AF;'>PTCy</span>"))




plot(Grade_byPTCy)
plot(Grade_byPTCy_CD4)
plot(Grade_byPTCy_CD8)
plot(NoPTCy)
plot(PTCy)
plot(All_byPTCy)

plot(CD4_byPTCy)
plot(CD8_byPTCy)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/Grade_byPTCy',type,'.png'), Grade_byPTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/Grade_byPTCy_CD4',type,'.png'), Grade_byPTCy_CD4, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/Grade_byPTCy_CD8',type,'.png'), Grade_byPTCy_CD8, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/all_NoPTCy',type,'.png'), NoPTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/all_PTCy',type,'.png'), PTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/all_All_byPTCy.png'), All_byPTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/CD4_byPTCy',type,'.png'), CD4_byPTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/CD8_byPTCy',type,'.png'), CD8_byPTCy, width = 9.5, height = 5, dpi = 600)


ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/Grade_byPTCy',type,'.pdf'), Grade_byPTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/Grade_byPTCy_CD4',type,'.pdf'), Grade_byPTCy_CD4, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/Grade_byPTCy_CD8',type,'.pdf'), Grade_byPTCy_CD8, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/all_NoPTCy',type,'.pdf'), NoPTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/all_PTCy',type,'.pdf'), PTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/all_All_byPTCy.pdf'), All_byPTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/CD4_byPTCy',type,'.pdf'), CD4_byPTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/CD8_byPTCy',type,'.pdf'), CD8_byPTCy, width = 9.5, height = 5, dpi = 600)


ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/Grade_byPTCy',type,'.svg'), Grade_byPTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/all_NoPTCy',type,'.svg'), NoPTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/Grade_byPTCy_CD4',type,'.svg'), Grade_byPTCy_CD4, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/Grade_byPTCy_CD8',type,'.svg'), Grade_byPTCy_CD8, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/all_PTCy',type,'.svg'), PTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/all_All_byPTCy.svg'), All_byPTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/CD4_byPTCy',type,'.svg'), CD4_byPTCy, width = 9.5, height = 5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3ex/CD8_byPTCy',type,'.svg'), CD8_byPTCy, width = 9.5, height = 5, dpi = 600)
```


```{r, warning=FALSE, fig.width=9.5, fig.height = 15}

fig_EF <- plot_grid(freq_Both_vio_ex, Grade_byPTCy, All_by_grade_fig_ex, ncol = 1, align = "v")
plot(fig_EF)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/fig_EF.pdf'), fig_EF, width = 9.5, height = 15, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/fig_EF.png'), fig_EF, width = 9.5, height = 15, dpi = 600)

```

```{r, warning=FALSE, fig.width=6, fig.height = 4}

# data_CD48_all$PTCy <- ifelse(grepl("PTCy", data_CD48_all$PTCy), "PTCy", "No PTCy")

data_CD48_all_expand_adj <- dplyr::full_join(data_CD48_all, clinical_data, by = 'PatientID')
data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!is.na(data_CD48_all_expand_adj$cdr3_amino_acid),]
data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!grepl("67",data_CD48_all_expand_adj$PatientID),]
data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!grepl("70",data_CD48_all_expand_adj$PatientID),]

data_CD48_all_expand_adj <- data_CD48_all_expand_adj%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", "No and Mild"))


if(remove_pat_regx != ""){
  data_CD48_all_expand_adj <- data_CD48_all_expand_adj[!grepl(remove_pat_regx,data_CD48_all_expand_adj$PatientID),]
}

data_CD48_all_expand_sel_PTCy <- subset(data_CD48_all_expand_adj, "PTCy" == data_CD48_all_expand_adj$PTCy)
data_CD48_all_expand_sel_noPTCy <- subset(data_CD48_all_expand_adj, "No PTCy" == data_CD48_all_expand_adj$PTCy)


BothPTCy_CD4_data <- subset(data_CD48_all_expand_adj,!grepl('CD4CD8',data_CD48_all_expand_adj$cell_type) & grepl('CD4',data_CD48_all_expand_adj$cell_type))

BothPTCy_CD8_data <- subset(data_CD48_all_expand_adj,!grepl('CD4CD8',data_CD48_all_expand_adj$cell_type) & grepl('CD8',data_CD48_all_expand_adj$cell_type))

BothPTCy_CD4_results <- num_clones(BothPTCy_CD4_data, "PTCy", num_cuts = 4, title = "Both PTCy CD4", ylab ="Alloreactive \n Cumulative Frequency")
BothPTCy_CD4 <- BothPTCy_CD4_results[[1]]+
  scale_fill_manual(values=c("#E6BE60", "#A483AF"))+
  # scale_color_manual(values=c("#E6BE60", "#A483AF"))+
  labs(title = "No PTCy vs PTCy CD4")
test_BothPTCy_CD4 <- BothPTCy_CD4_results[[2]]
BothPTCy_CD4_data_res <- BothPTCy_CD4_results[[3]]

BothPTCy_CD8_results <- num_clones(BothPTCy_CD8_data, "PTCy", num_cuts = 4, title = "Both PTCy CD8", ylab = "Alloreactive \n Cumulative Frequency")
BothPTCy_CD8 <- BothPTCy_CD8_results[[1]]+
  scale_fill_manual(values=c("#E6BE60", "#A483AF"))+
  # scale_color_manual(values=c("#E6BE60", "#A483AF"))+
  labs(title = "No PTCy vs PTCy CD8")
test_BothPTCy_CD8 <- BothPTCy_CD8_results[[2]]
BothPTCy_CD8_data_res <- BothPTCy_CD8_results[[3]]

type = "GVHD_GRADE_1"

BothPTCy_CD4_results_g <- num_clones(BothPTCy_CD4_data, type, num_cuts = 4, title = "No and Mild vs. Severe CD4", ylab = "Alloreactive \n Cumulative Frequency")
BothPTCy_CD4_g <- BothPTCy_CD4_results_g[[1]]+
  scale_fill_manual(values=c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
test_BothPTCy_CD4_g <- BothPTCy_CD4_results_g[[2]]
BothPTCy_CD4_data_g <- BothPTCy_CD4_results_g[[3]]

BothPTCy_CD8_results_g <- num_clones(BothPTCy_CD8_data, type, num_cuts = 4, title = "No and Mild vs. Severe CD8", ylab = "Alloreactive \n Cumulative Frequency")
BothPTCy_CD8_g <- BothPTCy_CD8_results_g[[1]]+
  scale_fill_manual(values=c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
test_BothPTCy_CD8_g <- BothPTCy_CD8_results_g[[2]]
BothPTCy_CD8_data_g <- BothPTCy_CD8_results_g[[3]]

NoPTCy_CD4_data <- subset(data_CD48_all_expand_sel_noPTCy,!grepl('CD4CD8',data_CD48_all_expand_sel_noPTCy$cell_type) & grepl('CD4',data_CD48_all_expand_sel_noPTCy$cell_type))
NoPTCy_CD8_data <- subset(data_CD48_all_expand_sel_noPTCy,!grepl('CD4CD8',data_CD48_all_expand_sel_noPTCy$cell_type) & grepl('CD8',data_CD48_all_expand_sel_noPTCy$cell_type))
PTCy_CD4_data <- subset(data_CD48_all_expand_sel_PTCy,!grepl('CD4CD8',data_CD48_all_expand_sel_PTCy$cell_type) & grepl('CD4',data_CD48_all_expand_sel_PTCy$cell_type))
PTCy_CD8_data <- subset(data_CD48_all_expand_sel_PTCy,!grepl('CD4CD8',data_CD48_all_expand_sel_PTCy$cell_type) & grepl('CD8',data_CD48_all_expand_sel_PTCy$cell_type))

num_cuts <- 4

NoPTCy_CD4_results <- num_clones(NoPTCy_CD4_data, type, num_cuts = num_cuts, title = "NoPTCy No and Mild vs. Severe CD4+", ylab = "Alloreactive \n Cumulative Frequency")
NoPTCy_CD4 <- NoPTCy_CD4_results[[1]]+
  labs(title = "No PTCy No and Mild vs. Severe CD4+")+
  scale_fill_manual(values=c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
test_NoPTCy_CD4 <- NoPTCy_CD4_results[[2]]
NoPTCy_CD4_data <- NoPTCy_CD4_results[[3]]

NoPTCy_CD8_results <- num_clones(NoPTCy_CD8_data, type, num_cuts = num_cuts, title = "NoPTCy No and Mild vs. Severe CD8+", ylab = "Alloreactive \n Cumulative Frequency")
NoPTCy_CD8 <- NoPTCy_CD8_results[[1]]+
  labs(title = "No PTCy No and Mild vs. Severe CD8+")+
  scale_fill_manual(values=c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
test_NoPTCy_CD8 <- NoPTCy_CD8_results[[2]]
NoPTCy_CD8_data <-NoPTCy_CD8_results[[3]]

PTCy_CD4_results <- num_clones(PTCy_CD4_data, type, num_cuts = num_cuts, title = "PTCy No and Mild vs. Severe CD4+", ylab = "Alloreactive \n Cumulative Frequency")
PTCy_CD4 <- PTCy_CD4_results[[1]]+
  labs(title = "PTCy No and Mild vs. Severe CD4+")+
  scale_fill_manual(values=c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))
test_PTCy_CD4 <- PTCy_CD4_results[[2]]
PTCy_CD4_data <- PTCy_CD4_results[[3]]


PTCy_CD8_results <- num_clones(PTCy_CD8_data, type, num_cuts = num_cuts, title = "PTCy No and Mild vs. Severe CD8+", ylab = "Alloreactive \n Cumulative Frequency")
test_PTCy_CD8 <- PTCy_CD8_results[[2]]
PTCy_CD8_data <- PTCy_CD8_results[[3]]

PTCy_CD8 <- PTCy_CD8_results[[1]]+
  labs(title = "PTCy No and Mild vs. Severe CD8+")+
  scale_fill_manual(values=c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))

plot(BothPTCy_CD4)
plot(BothPTCy_CD8)
plot(BothPTCy_CD4_g)
plot(BothPTCy_CD8_g)
plot(NoPTCy_CD4)
plot(NoPTCy_CD8)
plot(PTCy_CD4)
plot(PTCy_CD8)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/BothPTCy_CD4_all_',type,'.png'), BothPTCy_CD4,  width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/BothPTCy_CD8_all_',type,'.png'), BothPTCy_CD8, width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/BothPTCy_CD4_all_',type,'.png'), BothPTCy_CD4_g,  width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/BothPTCy_CD8_all_',type,'.png'), BothPTCy_CD8_g, width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/NoPTCy_CD4_all_',type,'.png'), NoPTCy_CD4, width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/NoPTCy_CD8_all_',type,'.png'), NoPTCy_CD8,  width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/PTCy_CD4_all_',type,'.png'), PTCy_CD4, width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/PTCy_CD8_all_',type,'.png'), PTCy_CD8, width = 6, height = 3, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/BothPTCy_CD4_all_',type,'.pdf'), BothPTCy_CD4,  width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/BothPTCy_CD8_all_',type,'.pdf'), BothPTCy_CD8, width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/BothPTCy_CD4_all_',type,'.pdf'), BothPTCy_CD4_g,  width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/BothPTCy_CD8_all_',type,'.pdf'), BothPTCy_CD8_g, width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/NoPTCy_CD4_all_',type,'.pdf'), NoPTCy_CD4, width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/NoPTCy_CD8_all_',type,'.pdf'), NoPTCy_CD8,  width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/PTCy_CD4_all_',type,'.pdf'), PTCy_CD4, width = 6, height = 3, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/PTCy_CD8_all_',type,'.pdf'), PTCy_CD8, width = 6, height = 3, dpi = 600)

#KEEP IN MIND
```

```{r}
div_clones <- function(data, type, div_metric = "shannons.entropy", num_cuts = 5){
  #' Divide clones based on timepoints and diversity metric
  #' 
  #' @description This function processes a dataset of clones to divide it into pre- and post-stimulation data based on specific timepoints. The function filters and bins data to analyze clonal diversity using a specified diversity metric.
  #'
  #' @param data A dataframe containing clone data, including a `SampleID` column which indicates timepoints or conditions.
  #' @param type A character string specifying the type of analysis or dataset being processed.
  #' @param div_metric A character string specifying the diversity metric to use (default is "shannons.entropy").
  #' @param num_cuts An integer indicating the number of bins to divide the timepoints into (default is 5).
  #' @usage div_clones(data, type, div_metric, num_cuts)
  #' @return A list containing two dataframes: `data_pre` for pre-stimulation and `data_post` for post-stimulation data.
  
  data_pre <- data[grepl("MLR|unstim", data$SampleID),]%>%
    dplyr::mutate(timepoint = SampleID) # Filter for pre-stimulation samples and mutate timepoint to be the same as SampleID
    
  data_proc <- data %>%
    dplyr::mutate(timepoint = strtoi(SampleID)) # Convert SampleID to an integer timepoint for processing
  data_proc <- data_proc[!is.na(data_proc$timepoint),] # Remove any rows where timepoint conversion was not possible (NA values)
  
  data_post <- data_proc[data_proc$timepoint <= 365 & data_proc$timepoint > 3,] %>%
    ungroup() %>%
    dplyr::mutate(bins = cut(timepoint, breaks = unique(quantile(timepoint, probs = 0:num_cuts/num_cuts)), include.lowest = TRUE)) # Filter for post-stimulation samples within a year and create bins based on timepoints
  
  return(list(data_pre, data_post)) # Return a list containing pre- and post-stimulation dataframes
}


div_cut <- function(data, type, div_metric = "shannons.entropy", num_cuts = 5, text_size = 20, ylab = "", title = ""){
  #' Analyze diversity metrics and visualize results
  #' 
  #' @description This function calculates diversity metrics for different bins of timepoints, performs pairwise Wilcoxon tests, and visualizes the results using boxplots with significance annotations.
  #'
  #' @param data A dataframe containing clone data, including a `SampleID` column.
  #' @param type A character string specifying the grouping variable for the analysis.
  #' @param div_metric A character string specifying the diversity metric to use (default is "shannons.entropy").
  #' @param num_cuts An integer indicating the number of bins to divide the timepoints into (default is 5).
  #' @param text_size An integer specifying the size of text in the plot (default is 20).
  #' @param ylab A character string for the y-axis label in the plot.
  #' @param title A character string for the plot title.
  #' @usage div_cut(data, type, div_metric, num_cuts, text_size, ylab, title)
  #' @return A list containing the boxplot with significance annotations, the Wilcoxon test results, and the processed data.

  results <- div_clones(data, type, div_metric, num_cuts)
  data_pre <- results[[1]]
  data_post <- results[[2]]

  print(unique(data_pre[[type]])) # Print unique values of the grouping variable in pre-stimulation data

  # Summarize post-stimulation data by bins and grouping variable
  data_post_sum <- data_post %>%
    group_by(across(all_of(c("bins", type)))) %>%
    summarise(n_each = n())

  # Perform pairwise Wilcoxon test and add positions for significance annotations
  stat_test <- data_post %>%
    group_by(bins) %>%
    pairwise_wilcox_test(as.formula(paste0(div_metric, " ~ ", type)), p.adjust.method = "hochberg") %>%
    rstatix::add_xy_position(x = "bins")

  # Create boxplot with beeswarm overlay and significance annotations
  box_plot_post <- ggboxplot(data_post, x = "bins", y = div_metric, fill = type, outlier = NA) +
    stat_pvalue_manual(stat_test, label = "p.adj.signif", tip.length = 0.01, step.increase = 0.001) +
    geom_beeswarm(position = "dodge", dodge.width = .75, aes(group = .data[[type]])) +
    scale_y_continuous(labels = scales::number_format(accuracy = 1))+
    theme_pubr() +
    theme(
      legend.text = element_markdown(),
      text = element_text(size = text_size),
      axis.text.x = element_text(angle = 45, hjust = 1),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.background = element_blank(),
      axis.line = element_line(color = "black"),
      axis.ticks = element_blank(),
      axis.ticks.length = unit(0.25, "cm")
    ) +
    labs(y = ylab, title = title)

  return(list(box_plot_post, stat_test, data_post)) # Return boxplot, statistical test results, and processed post-stimulation data
}

```

```{r, warning=FALSE, fig.width=8, fig.height = 4}

patient_data_allo_copy <- patient_data_all[patient_data_all$Rep_div == "allo", ]

patient_data_all_clinic <- dplyr::full_join(patient_data_allo_copy, meta_data_clin, by = "PatientID")%>%
  try(mutate(PTCy = PTCy.x))%>%
  try(select(-c(PTCy.x, PTCy.y)))%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", "No and Mild"))%>%
  mutate(GVHD_GRADE_1 = factor(GVHD_GRADE_1, levels = c("No and Mild", "Severe")))
patient_data_all_clinic <- patient_data_all_clinic[patient_data_all_clinic$GVHD_GRADE_1 != "Unclassified/Severe",]
patient_data_all_clinic <- patient_data_all_clinic[!is.na(patient_data_all_clinic$slope),]

if(remove_pat_regx != ""){
  patient_data_all_clinic <- patient_data_all_clinic[!grepl(remove_pat_regx,patient_data_all_clinic$PatientID),]
}


type <- "GVHD_GRADE_1"


result_Both <- div_cut(patient_data_all_clinic, type, num_cuts = 4, div_metric ="shannons.diversity" , ylab = "Shannon's Diversity",)
Both <- result_Both[[1]]+
  scale_fill_manual(name = "GVHD Grade", values = c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(x = "Days", title = "No and Mild vs Severe")
test_BothPTCy <- result_Both[[2]]


result_NoPTCy <- div_cut(subset(patient_data_all_clinic, patient_data_all_clinic$PTCy == "No PTCy"), type, num_cuts = 4, div_metric ="shannons.diversity" , ylab = "Shannon's Diversity",)
NoPTCy <- result_NoPTCy[[1]]+
  scale_fill_manual(name = "GVHD Grade",values = c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(x = "Days", title = "No PTCy: No and Mild vs Severe")
test_NoPTCy <- result_NoPTCy[[2]]


result_PTCy <- div_cut(subset(patient_data_all_clinic, patient_data_all_clinic$PTCy == "PTCy"), type, num_cuts = 4, div_metric ="shannons.diversity" , ylab = "Shannon's Diversity",)
PTCy <- result_PTCy[[1]]+
  scale_fill_manual(name = "GVHD Grade", values = c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(x = "Days", title = "PTCy: No and Mild vs Severe")
test_PTCy <- result_PTCy[[2]]

result_All_byPTCy <- div_cut(patient_data_all_clinic[patient_data_all_clinic$PatientID != "R070",], "PTCy", num_cuts = 4, div_metric ="shannons.diversity" , ylab = "Shannon's Diversity", title = "PTCy vs No PTCy")
All_byPTCy <- result_All_byPTCy[[1]]+
  scale_fill_manual(values = c("#E6BE60", "#A483AF"),
                    labels = c(
      `No PTCy` = "<span style='color:#E6BE60;'>No PTCy</span>",
      PTCy = "<span style='color:#A483AF;'>PTCy</span>"))+
  labs(x = "Days")

test_All_byPTCy <- result_All_byPTCy[[2]]



plot(Both)
plot(NoPTCy)
plot(PTCy)
plot(All_byPTCy)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/Grade_Shannon_',type,'.png'), Both, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/Grade_Shannon_', type,'.pdf'), Both, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/Grade_Shannon_', type,'.svg'), Both, width = 8, height = 4, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GradeNoPTCy_Shannon_',type,'.png'), NoPTCy,width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GradeNoPTCy_Shannon_', type,'.pdf'), NoPTCy,width = 8, height = 4, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GradePTCy_Shannon_',type,'.png'), PTCy, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/GradePTCy_Shannon_', type,'.pdf'), PTCy, width = 8, height = 4, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/ByPTCy_Shannon_',type,'.png'), All_byPTCy, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/ByPTCy_Shannon_', type,'.pdf'), All_byPTCy, width = 8, height = 4, dpi = 600)



```


```{r, fig.width = 8 , fig.height = 4}
patient_data_allo_CD4 <-bind_rows(all_allo_div_CD4)%>%
  subset(!is.na(slope))%>%
  mutate(cell_type = "CD4", PatientID = str_split_fixed(ID, "_", 3)[,1], SampleID = str_split_fixed(ID, "_", 3)[,2])
patient_data_allo_CD8 <-bind_rows(all_allo_div_CD8)%>%
  subset(!is.na(slope))%>%
  mutate(cell_type = "CD8", PatientID = str_split_fixed(ID, "_", 3)[,1], SampleID = str_split_fixed(ID, "_", 3)[,2])

patient_data_allo_CD48 <- bind_rows(patient_data_allo_CD8, patient_data_allo_CD4)
patient_data_allo_CD48 <- dplyr::inner_join(patient_data_allo_CD48, meta_data_clin, by = "PatientID")

if(remove_pat_regx != ""){
  patient_data_all_clinic <- patient_data_all_clinic[!grepl(remove_pat_regx,patient_data_all_clinic$PatientID),]
}
patient_data_allo_CD48_clinic <- patient_data_allo_CD48#[!grepl("70",patient_data_allo_CD48$PatientID),]


patient_data_allo_CD48_clinic <- patient_data_allo_CD48_clinic%>%
  filter(PatientID != "R070")%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", "No and Mild"))%>%
  mutate(GVHD_GRADE_1 = factor(GVHD_GRADE_1, levels = c("No and Mild", "Severe")))

patient_data_allo_CD48_clinic$GVHD_GRADE_2 <- factor(patient_data_allo_CD48_clinic$GVHD_GRADE_2, levels = c("Mild", "Moderate", "Severe"))
patient_data_allo_CD48_clinic_CD4 <- subset(patient_data_allo_CD48_clinic, grepl("4",patient_data_allo_CD48_clinic$cell_type))
patient_data_allo_CD48_clinic_CD8 <- subset(patient_data_allo_CD48_clinic, grepl("8",patient_data_allo_CD48_clinic$cell_type))
patient_data_allo_CD48_clinic_PTCy <- subset(patient_data_allo_CD48_clinic, "PTCy" == patient_data_allo_CD48_clinic$PTCy)
patient_data_allo_CD48_clinic_noPTCy <- subset(patient_data_allo_CD48_clinic, "PTCy" != patient_data_allo_CD48_clinic$PTCy)


n = 4


BothPTCy_CD4_results <- div_cut(patient_data_allo_CD48_clinic_CD4, "PTCy", num_cuts = n, ylab = "Shannon's Diversity", div_metric = "shannons.diversity")
BothPTCy_CD4 <- BothPTCy_CD4_results[[1]]+
  scale_fill_manual(values = c("#E6BE60", "#A483AF"),
                    labels = c(
      `No PTCy` = "<span style='color:#E6BE60;'>No PTCy</span>",
      PTCy = "<span style='color:#A483AF;'>PTCy</span>"))+
  labs(title = "No PTCy vs PTCy CD4", x = "Days")+
  theme(legend.position = "none")
test_BothPTCy_CD4 <- BothPTCy_CD4_results[[2]]
BothPTCy_CD4_data <- BothPTCy_CD4_results[[3]]

BothPTCy_CD8_results <- div_cut(patient_data_allo_CD48_clinic_CD8, "PTCy", num_cuts = n, ylab = "Shannon's Diversity", div_metric = "shannons.diversity")
BothPTCy_CD8 <- BothPTCy_CD8_results[[1]]+
  scale_fill_manual(values = c("#E6BE60", "#A483AF"),
                    labels = c(
      `No PTCy` = "<span style='color:#E6BE60;'>No PTCy</span>",
      PTCy = "<span style='color:#A483AF;'>PTCy</span>"))+
  labs(title = "No PTCy vs PTCy CD8", x = "Days")+
  theme(legend.position = "none")
test_BothPTCy_CD8 <- BothPTCy_CD8_results[[2]]
BothPTCy_CD8_data <- BothPTCy_CD8_results[[3]]

plot(BothPTCy_CD4)
plot(BothPTCy_CD8)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/BothPTCy_CD4.png'), BothPTCy_CD4, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/BothPTCy_CD8.png'), BothPTCy_CD8, width = 8, height = 4, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/BothPTCy_CD4.pdf'), BothPTCy_CD4, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/BothPTCy_CD8.pdf'), BothPTCy_CD8, width = 8, height = 4, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/BothPTCy_CD4.svg'), BothPTCy_CD4, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/BothPTCy_CD8.svg'), BothPTCy_CD8, width = 8, height = 4, dpi = 600)

type = "GVHD_GRADE_1"

num_cuts <- 2

BothPTCy_CD4_results <- div_cut(patient_data_allo_CD48_clinic_CD4, type, num_cuts = num_cuts, ylab = "Shannon's Diversity", div_metric = "shannons.diversity")
BothPTCy_CD4 <- BothPTCy_CD4_results[[1]]+
  scale_fill_manual(name = "GVHD Grade", values = c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(title = "No and Mild vs Severe CD4+", x = "Days")
test_BothPTCy_CD4 <- BothPTCy_CD4_results[[2]]
BothPTCy_CD4_data <- BothPTCy_CD4_results[[3]]

BothPTCy_CD8_results <- div_cut(patient_data_allo_CD48_clinic_CD8, type, num_cuts = num_cuts, ylab = "Shannon's Diversity", div_metric = "shannons.diversity")
BothPTCy_CD8 <- BothPTCy_CD8_results[[1]]+
  scale_fill_manual(name = "GVHD Grade", values = c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(title = "No and Mild vs Severe CD8+", x = "Days")
test_BothPTCy_CD8 <- BothPTCy_CD8_results[[2]]
BothPTCy_CD8_data <- BothPTCy_CD8_results[[3]]

NoPTCy_CD4_data <- subset(patient_data_allo_CD48_clinic_noPTCy,!grepl('CD4CD8',patient_data_allo_CD48_clinic_noPTCy$cell_type) & grepl('CD4',patient_data_allo_CD48_clinic_noPTCy$cell_type))

NoPTCy_CD8_data <- subset(patient_data_allo_CD48_clinic_noPTCy,!grepl('CD4CD8',patient_data_allo_CD48_clinic_noPTCy$cell_type) & grepl('CD8',patient_data_allo_CD48_clinic_noPTCy$cell_type))

PTCy_CD4_data <- subset(patient_data_allo_CD48_clinic_PTCy,!grepl('CD4CD8',patient_data_allo_CD48_clinic_PTCy$cell_type) & grepl('CD4',patient_data_allo_CD48_clinic_PTCy$cell_type))

PTCy_CD8_data <- subset(patient_data_allo_CD48_clinic_PTCy,!grepl('CD4CD8',patient_data_allo_CD48_clinic_PTCy$cell_type) & grepl('CD8',patient_data_allo_CD48_clinic_PTCy$cell_type))


NoPTCy_CD4_results <- div_cut(NoPTCy_CD4_data, type, num_cuts = num_cuts)
NoPTCy_CD4 <- NoPTCy_CD4_results[[1]]+
  scale_fill_manual(name = "GVHD Grade", values = c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(title = "No PTCy CD4", x = "Days")
test_NoPTCy_CD4 <- NoPTCy_CD4_results[[2]]
NoPTCy_CD4_data <- NoPTCy_CD4_results[[3]]

NoPTCy_CD8_results <- div_cut(NoPTCy_CD8_data, type, num_cuts = num_cuts, ylab = "Shannon's Diversity", div_metric = "shannons.diversity")
NoPTCy_CD8 <- NoPTCy_CD8_results[[1]]+
  scale_fill_manual(name = "GVHD Grade", values = c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(title = "No PTCy CD8", x = "Days")
test_NoPTCy_CD8 <- NoPTCy_CD8_results[[2]]
NoPTCy_CD8_data <-NoPTCy_CD8_results[[3]]

PTCy_CD4_results <- div_cut(PTCy_CD4_data, type, num_cuts = num_cuts, ylab = "Shannon's Diversity", div_metric = "shannons.diversity")
PTCy_CD4 <- PTCy_CD4_results[[1]]+
  scale_fill_manual(name = "GVHD Grade", values = c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(title = "PTCy CD4", x = "Days")
test_PTCy_CD4 <- PTCy_CD4_results[[2]]
PTCy_CD4_data <- PTCy_CD4_results[[3]]


PTCy_CD8_results <- div_cut(PTCy_CD8_data, type, num_cuts = num_cuts, ylab = "Shannon's Diversity", div_metric = "shannons.diversity")
PTCy_CD8 <- PTCy_CD8_results[[1]]+
  scale_fill_manual(name = "GVHD Grade", values = c("#058b8e", "#B85000", "grey"),
                    labels = c(
      `No and Mild` = "<span style='color:#058b8e;'>No and Mild</span>",
      Severe = "<span style='color:#B85000;'>Severe</span>"))+
  labs(title = "PTCy CD8", x = "Days")
test_PTCy_CD8 <- PTCy_CD8_results[[2]]
PTCy_CD8_data <- PTCy_CD8_results[[3]]

plot(BothPTCy_CD4)
plot(BothPTCy_CD8)
plot(NoPTCy_CD4)
plot(NoPTCy_CD8)
plot(PTCy_CD4)
plot(PTCy_CD8)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/BothPTCy_CD4_shannons',type,'.png'), BothPTCy_CD4, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/BothPTCy_CD8_shannons',type,'.png'), BothPTCy_CD8, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/NoPTCy_CD4_shannons',type,'.png'), NoPTCy_CD4, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/NoPTCy_CD8_shannons',type,'.png'), NoPTCy_CD8, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/PTCy_CD4_shannons',type,'.png'), PTCy_CD4, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/PTCy_CD8_shannons',type,'.png'), PTCy_CD8, width = 8, height = 4, dpi = 600)


ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/BothPTCy_CD4_shannons',type,'.pdf'), BothPTCy_CD4, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/BothPTCy_CD8_shannons',type,'.pdf'), BothPTCy_CD8, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/NoPTCy_CD4_shannons',type,'.pdf'), NoPTCy_CD4, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/NoPTCy_CD8_shannons',type,'.pdf'), NoPTCy_CD8, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/PTCy_CD4_shannons',type,'.pdf'), PTCy_CD4, width = 8, height = 4, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/PTCy_CD8_shannons',type,'.pdf'), PTCy_CD8, width = 8, height = 4, dpi = 600)

```

```{r}
data_CD48_all_expand_sel <- dplyr::full_join(data_CD48_all, clinical_data, by = 'PatientID')
data_CD48_all_expand_sel <- data_CD48_all_expand_sel[!is.na(data_CD48_all_expand_sel$cdr3_amino_acid),]

```


```{r}
div_fig_pat <- function(patient_data, div, type, group, ylim){
  #' Create diversity figures for patient data
  #' 
  #' @description This function creates two timeline plots for displaying diversity metrics
  #' over time, split into early (0-365 days) and later (365+ days) time points.
  #' 
  #' @param patient_data Dataframe containing patient data
  #' @param div Name of the column containing the diversity metric
  #' @param type Vector of Rep_div values to include
  #' @param group Name of the column to use for grouping and coloring
  #' @param ylim Vector of two numbers specifying y-axis limits
  #' @usage div_fig_pat(patient_data, div, type, group, ylim)
  #' @return list: list(fig_early, fig_later, max_tp)
  
  selected <- subset(patient_data, patient_data$Rep_div %in% type) %>%
    mutate(SampleID = strtoi(SampleID)) %>%
    filter(!is.na(SampleID), SampleID > 3, is.finite(.data[[div]]))
  
  selected$GVHD_GRADE_1 <- factor(selected$GVHD_GRADE_1, levels = c('No', 'Severe', 'Mild'))
  max_tp <- max(selected$SampleID)
  
  fig_early <- ggplot(selected, aes(x = SampleID, y = .data[[div]], group = PatientID, color = .data[[group]], shape = .data[[group]])) +
    geom_point(size = 4) +
    geom_line(size = 1.75) +
    geom_vline(data = selected, aes(xintercept = Onset, color = factor(PatientID)), linetype = 'dashed', show.legend = FALSE) +
    labs(x = "Days Post Transplantation", y = div) +
    scale_y_continuous(labels = scales::number_format(accuracy = 0.01))+
    scale_colour_discrete(na.translate = F) +
    scale_shape_discrete(na.translate = F) +
    coord_cartesian(xlim = c(0, 365), ylim = ylim, expand = F) +
    theme_pubr() +
    theme(axis.text.x = element_text(size = 30, angle = 75, vjust = 0.5),
          axis.text.y = element_text(size = 30),
          axis.title.x.bottom = element_text(hjust = 0.75),
          axis.line.x.bottom = element_line(linewidth = 1.3),
          axis.line.y = element_line(linewidth = 1.3),
          axis.text = element_text(size = 40))
  
  fig_later <- ggplot(selected, aes(x = SampleID, y = .data[[div]], group = PatientID, color = .data[[group]], shape = .data[[group]])) +
    geom_point(size = 4) +
    geom_line(size = 1.75) +
    geom_vline(data = selected, aes(xintercept = Onset, color = factor(PatientID)), linetype = 'dashed', show.legend = FALSE) +
    labs(x = "") +
    scale_y_continuous(labels = scales::number_format(accuracy = 0.01))+
    scale_colour_discrete(na.translate = F) +
    scale_shape_discrete(na.translate = F) +
    coord_cartesian(xlim = c(365, max_tp), ylim = ylim, expand = F) +
    theme_pubr() +
    theme(axis.text.x = element_text(size = 30, angle = 75, vjust = 0.5),
          axis.text.y = element_text(size = 30),
          axis.line.x.bottom = element_line(linewidth = 1.3),
          axis.title.y = element_blank(),
          axis.text.y.left = element_blank(),
          axis.line.y.left = element_blank(),
          axis.minor.ticks.y.left = element_blank(),
          axis.ticks.y.left = element_blank(),
          axis.text = element_text(size = 40),
          legend.position = "none")
  
  return(list(fig_early, fig_later, max_tp))
}

```


```{r, fig.width = 10, fig.height= 5}
#plot meta timeline
plot_metatime_cont <- function(immdata, meta_to_display, time_var, color = "PatientID", labels = c(""), max_tp){
  #' Plot meta timeline
  #' 
  #' @description This function creates two timeline plots for displaying metadata over time.
  #' One plot shows early time points (0-365 days) and the other shows later time points.
  #' 
  #' @param immdata Dataframe containing the immunosequencing data and metadata
  #' @param meta_to_display Vector of column names in immdata to be displayed as metadata
  #' @param time_var Name of the column in immdata that represents time
  #' @param color Name of the column to be used for coloring points (default: "PatientID")
  #' @param labels Vector of labels for the y-axis (should match length of meta_to_display)
  #' @param max_tp Maximum time point for the second plot
  #' @usage plot_metatime_cont(immdata, meta_to_display, time_var, color, labels, max_tp)
  #' @return list: list(meta_data_plot_early, meta_data_plot_later)
  
  # Reshape and prepare the data for plotting
  immdata_meta <- immdata %>%
    select(all_of(c(time_var, color, meta_to_display))) %>%
    pivot_longer(cols = meta_to_display, names_to = "Meta") %>%
    mutate(Meta = factor(Meta, levels = rev(meta_to_display)))
  
  # Create the first plot for early time points (0-365 days)
  meta_data_plot_early <- ggplot(immdata_meta, aes(x = value, y = Meta, color = !!sym(color))) +
    geom_point(shape = 18, size = 9) +
    scale_y_discrete(labels = rev(labels)) +
    theme_pubr() +
    coord_cartesian(xlim = c(0,365)) +
    scale_x_continuous(expand = expansion(mult = c(0,0))) +
    theme(legend.position="none",
          axis.line.x.bottom = element_blank(),
          axis.line.y.left = element_line(size = 1.5),
          axis.title.y.left = element_blank(), 
          axis.title.x.bottom = element_blank(), 
          axis.minor.ticks.x = element_blank(), 
          axis.ticks.x = element_blank(), 
          axis.text.x.bottom = element_blank(), 
          panel.grid.major.y = element_line(color = "#F0F0F0"))
  
  # Create the second plot for later time points (366 days to max_tp)
  meta_data_plot_later <- ggplot(immdata_meta, aes(x = value, y = Meta, color = !!sym(color))) +
    geom_point(shape = 18, size = 9) +
    theme_pubr() +
    coord_cartesian(xlim = c(366, max_tp)) +
    scale_x_continuous(expand = expansion(mult = c(0,0))) +
    theme(legend.position="none",
          axis.line.x.bottom = element_blank(), 
          axis.title.y.left = element_blank(), 
          axis.title.x.bottom = element_blank(),
          axis.text.y.left = element_blank(), 
          axis.line.y.left = element_blank(), 
          axis.minor.ticks.y.left = element_blank(), 
          axis.ticks.y.left = element_blank(), 
          axis.minor.ticks.x = element_blank(), 
          axis.ticks.x = element_blank(), 
          axis.text.x.bottom = element_blank(), 
          panel.grid.major.y = element_line(color = "#F0F0F0"))
  
  # Return both plots as a list
  return(list(meta_data_plot_early, meta_data_plot_later))
}
```

```{r, fig.height=10/1.1,fig.width = 18/1.1, warning = FALSE}
clinical_data <- read_excel("/Users/lingtingshi/Documents/GVHD_project/From_Michael/TCR_Project.xlsx", sheet = 'Patient Metadata cont.')

patient_data_all_clinic <- dplyr::inner_join(patient_data_all, clinical_data, by = "PatientID")%>%
  dplyr::inner_join(meta_data, by = "PatientID")%>%
  filter(strtoi(SampleID) > 3 | grepl("MLR|unstim", SampleID))%>%
  mutate(Onset = as.numeric(Onset), `EBV +` = as.numeric(`EBV +`),`Off Immunosuppresion` = as.numeric(`Off Immunosuppresion`), `CMV +` = as.numeric(`CMV +`),`HHV6 +` = as.numeric(`HHV6 +`), Systemic_Steroids = as.numeric(Systemic_Steroids) )
severe_cumulative <- patient_data_all_clinic[grepl('Severe',patient_data_all_clinic$GVHD_GRADE_1),]
mild_cumulative <- patient_data_all_clinic[grepl('Mild',patient_data_all_clinic$GVHD_GRADE_1),]
no_cumulative <- patient_data_all_clinic[grepl('No',patient_data_all_clinic$GVHD_GRADE_1),]
mild_cumulative <- mild_cumulative

no_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", no_cumulative$`Off Immunosuppresion`))
mild_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", mild_cumulative$`Off Immunosuppresion`))
severe_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", severe_cumulative$`Off Immunosuppresion`))

labels <- c("GHVD Onset", "Off Immunosuppresion", "EB Viremia", "CM Viremia", "HHV6", "Systemic Steroid")
##SEVERE

severe_cumulative_fig <- div_fig_pat(severe_cumulative, 'shannons.diversity', c('all'), 'PatientID', ylim = c(1,10))
severe_cumulative_fig_early <- severe_cumulative_fig[[1]]+
  labs(title = 'Severe GVHD', y = "Shannon's Diveristy")+
  theme(text = element_text(size = 36), plot.margin=unit(c(0,0,0,6.2), 'cm'))
severe_cumulative_fig_later <- severe_cumulative_fig[[2]]+
  theme(text = element_text(size = 36))
severe_meta_data_plot <- plot_metatime_cont(severe_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = severe_cumulative_fig[[3]])

severe_meta_data_plot[[1]] <- severe_meta_data_plot[[1]]+
  theme(text = element_text(size = 28))

severe_meta_data_plot[[2]] <- severe_meta_data_plot[[2]]+
  theme(text = element_text(size = 28))

  severe_comb <- plot_grid(plotlist = list(severe_cumulative_fig_early, severe_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

severe_meta_comb <- plot_grid(plotlist = list(severe_meta_data_plot[[1]], severe_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

Severe_meta_fig <- plot_grid(severe_comb, severe_meta_comb, nrow = 2, align = "hv", greedy = TRUE, rel_heights = c(1,.3))

##MIlD

mild_cumulative_fig <- div_fig_pat(mild_cumulative, 'shannons.diversity', c('all'), 'PatientID', ylim = c(1,10))
mild_cumulative_fig_early <- mild_cumulative_fig[[1]]+
  labs(title = 'Mild GVHD', y = "Shannon's Diveristy")+
  theme(text = element_text(size = 36), plot.margin=unit(c(0,0,0,6.2), 'cm'))
mild_cumulative_fig_later <- mild_cumulative_fig[[2]]+
  theme(text = element_text(size = 36))
mild_meta_data_plot <- plot_metatime_cont(mild_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = mild_cumulative_fig[[3]])

mild_meta_data_plot[[1]] <- mild_meta_data_plot[[1]]+
  theme(text = element_text(size = 28))

mild_meta_data_plot[[2]] <- mild_meta_data_plot[[2]]+
  theme(text = element_text(size = 28))

  mild_comb <- plot_grid(plotlist = list(mild_cumulative_fig_early, mild_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

 mild_meta_comb <- plot_grid(plotlist = list(mild_meta_data_plot[[1]], mild_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

Mild_meta_fig <- plot_grid(mild_comb, mild_meta_comb, nrow = 2, align = "hv", greedy = TRUE, rel_heights = c(1,.3))
  
##NO

no_cumulative_fig <- div_fig_pat(no_cumulative, 'shannons.diversity', c('all'), 'PatientID', ylim = c(1,10))
no_cumulative_fig_early <- no_cumulative_fig[[1]]+
  labs(title = 'No GVHD', y = "Shannon's Diveristy")+
  theme(text = element_text(size = 36), plot.margin=unit(c(0,0,0,6.2), 'cm'))
no_cumulative_fig_later <- no_cumulative_fig[[2]]+
  theme(text = element_text(size = 36))
no_meta_data_plot <- plot_metatime_cont(no_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = no_cumulative_fig[[3]])

no_meta_data_plot[[1]] <- no_meta_data_plot[[1]]+
  theme(text = element_text(size = 28))

no_meta_data_plot[[2]] <- no_meta_data_plot[[2]]+
  theme(text = element_text(size = 28))

  no_comb <- plot_grid(plotlist = list(no_cumulative_fig_early, no_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))
  
  no_meta_comb <- plot_grid(plotlist = list(no_meta_data_plot[[1]], no_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))
  
No_meta_fig <- plot_grid(no_comb, no_meta_comb, nrow = 2, align = "hv", greedy = TRUE, rel_heights = c(1,.3))

plot(Severe_meta_fig)
plot(Mild_meta_fig)
plot(No_meta_fig)

  
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/severe_cumulative.png'), Severe_meta_fig, width = 18/1.1, height = 10/1.1, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/mild_cumuluative.png'), Mild_meta_fig, width = 18/1.1, height = 10, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/no_cumuluative.png'), No_meta_fig, width = 18/1.1, height = 10/1.1, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/severe_cumulative.pdf'), Severe_meta_fig, width = 18/1.1, height = 10/1.1, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/mild_cumuluative.pdf'), Mild_meta_fig, width = 18/1.1, height = 10/1.1, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/no_cumuluative.pdf'), No_meta_fig, width = 18/1.1, height = 10/1.1, dpi = 600)

```

```{r, fig.height=10/1.1,fig.width = 18/1.1, warning = FALSE}
clinical_data <- read_excel("/Users/lingtingshi/Documents/GVHD_project/From_Michael/TCR_Project.xlsx", sheet = 'Patient Metadata cont.')

patient_data_all_clinic <- dplyr::inner_join(patient_data_all, clinical_data, by = "PatientID")%>%
  dplyr::inner_join(meta_data, by = "PatientID")%>%
  filter(strtoi(SampleID) > 3 | !grepl("MLR|unstim", SampleID))%>%
  mutate(Onset = as.numeric(Onset), `EBV +` = as.numeric(`EBV +`),`Off Immunosuppresion` = as.numeric(gsub("[^0-9]", "", `Off Immunosuppresion`)), `CMV +` = as.numeric(`CMV +`),`HHV6 +` = as.numeric(`HHV6 +`), Systemic_Steroids = as.numeric(Systemic_Steroids) )
severe_cumulative <- patient_data_all_clinic[grepl('Severe',patient_data_all_clinic$GVHD_GRADE_1),]
mild_cumulative <- patient_data_all_clinic[grepl('Mild',patient_data_all_clinic$GVHD_GRADE_1),]
no_cumulative <- patient_data_all_clinic[grepl('No',patient_data_all_clinic$GVHD_GRADE_1),]
mild_cumulative <- mild_cumulative

no_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", no_cumulative$`Off Immunosuppresion`))
mild_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", mild_cumulative$`Off Immunosuppresion`))
severe_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", severe_cumulative$`Off Immunosuppresion`))

labels <- c("GHVD Onset", "Off Immunosuppresion", "EBV Viremia", "CMV Viremia", "HHV6 Viremia", "Systemic Steroid")
##SEVERE

severe_cumulative_fig <- div_fig_pat(severe_cumulative,  'cum_freq', c('allo'), 'PatientID', ylim = c(0, 0.05))
severe_cumulative_fig_early <- severe_cumulative_fig[[1]]+
  labs(y = "Alloreactive Cumulative\n Frequency", title = 'Severe GVHD')+
  theme(text = element_text(size = 32), plot.margin=unit(c(0,0,0,5.2), 'cm'))
severe_cumulative_fig_later <- severe_cumulative_fig[[2]]+
  theme(text = element_text(size = 32))
severe_meta_data_plot <- plot_metatime_cont(severe_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = severe_cumulative_fig[[3]])

severe_meta_data_plot[[1]] <- severe_meta_data_plot[[1]]+
  theme(text = element_text(size = 26))+
  coord_cartesian(xlim = c(0,360))+
  scale_x_continuous(expand = expansion(mult = c(0,0)))

severe_meta_data_plot[[2]] <- severe_meta_data_plot[[2]]+
  theme(text = element_text(size = 26))+
  coord_cartesian(xlim = c(360,severe_cumulative_fig[[3]]))+
  scale_x_continuous(expand = expansion(mult = c(0,0)))

severe_comb <- plot_grid(plotlist = list(severe_cumulative_fig_early, severe_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))


severe_meta_comb <- plot_grid(plotlist = list(severe_meta_data_plot[[1]], severe_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

Severe_meta_fig <- plot_grid(severe_comb, severe_meta_comb, nrow = 2, greedy = TRUE, rel_heights = c(1,.3))


##Mild

mild_cumulative_fig <- div_fig_pat(mild_cumulative,  'cum_freq', c('allo'), 'PatientID', ylim = c(0, 0.05))
mild_cumulative_fig_early <- mild_cumulative_fig[[1]]+
  labs(y = "Alloreactive Cumulative\n Frequency", title = 'Mild GVHD')+
  theme(text = element_text(size = 32), plot.margin=unit(c(0,0,0,5.2), 'cm'))
mild_cumulative_fig_later <- mild_cumulative_fig[[2]]+
  theme(text = element_text(size = 32))
mild_meta_data_plot <- plot_metatime_cont(mild_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = mild_cumulative_fig[[3]])

mild_meta_data_plot[[1]] <- mild_meta_data_plot[[1]]+
  theme(text = element_text(size = 26))

mild_meta_data_plot[[2]] <- mild_meta_data_plot[[2]]+
  theme(text = element_text(size = 26))

mild_comb <- plot_grid(plotlist = list(mild_cumulative_fig_early, mild_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))
  

mild_meta_comb <- plot_grid(plotlist = list(mild_meta_data_plot[[1]], mild_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

Mild_meta_fig <- plot_grid(mild_comb, mild_meta_comb, nrow = 2, greedy = TRUE, rel_heights = c(1,.3))


##NO

no_cumulative_fig <- div_fig_pat(no_cumulative,  'cum_freq', c('allo'), 'PatientID', ylim = c(0, 0.05))
no_cumulative_fig_early <- no_cumulative_fig[[1]]+
  labs(y = "Alloreactive Cumulative\n Frequency", title = 'No GVHD')+
  theme(text = element_text(size = 32), plot.margin=unit(c(0,0,0,5.2), 'cm'))
no_cumulative_fig_later <- no_cumulative_fig[[2]]+
  theme(text = element_text(size = 32))
no_meta_data_plot <- plot_metatime_cont(no_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = no_cumulative_fig[[3]])

no_meta_data_plot[[1]] <- no_meta_data_plot[[1]]+
  theme(text = element_text(size = 26))

no_meta_data_plot[[2]] <- no_meta_data_plot[[2]]+
  theme(text = element_text(size = 26))

no_comb <- plot_grid(plotlist = list(no_cumulative_fig_early, no_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))
  

no_meta_comb <- plot_grid(plotlist = list(no_meta_data_plot[[1]], no_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

No_meta_fig <- plot_grid(no_comb, no_meta_comb, nrow = 2, greedy = TRUE, rel_heights = c(1,.3))


plot(Severe_meta_fig)
plot(Mild_meta_fig)
plot(No_meta_fig)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/severe_cumulative.png'), Severe_meta_fig, width = 18/1.1, height = 10/1.1, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/mild_cumuluative.png'), Mild_meta_fig, width = 18/1.1, height = 10/1.1, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/no_cumuluative.png'), No_meta_fig, width = 18/1.1, height = 10/1.1, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/severe_cumulative.pdf'), Severe_meta_fig, width = 18/1.1, height = 10/1.1, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/mild_cumuluative.pdf'), Mild_meta_fig, width = 18/1.1, height = 10/1.1, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/no_cumuluative.pdf'), No_meta_fig, width = 18/1.1, height = 10/1.1, dpi = 600)

write.csv(patient_data_all_clinic, "/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/PatientDataClinic.csv")


```


```{r,fig.height=10,fig.width = 18, warning = FALSE}
clinical_data <- read_excel("/Users/lingtingshi/Documents/GVHD_project/From_Michael/TCR_Project.xlsx", sheet = 'Patient Metadata cont.')



data_CD48_all_expand_adj <- data_CD48_all%>%
  filter(templates !=0)%>%
  group_by(PatientID,timepoint)%>%
  summarise(n_clone = n())
  


patient_data_all_clinic <- dplyr::inner_join(data_CD48_all_expand_adj, clinical_data, by = "PatientID")%>%
  dplyr::inner_join(meta_data, by = "PatientID")%>%
  filter(strtoi(timepoint) > 3 | !grepl("MLR|unstim", timepoint))%>%
  mutate(Onset = as.numeric(Onset), `EBV +` = as.numeric(`EBV +`),`Off Immunosuppresion` = as.numeric(gsub("[^0-9]", "", `Off Immunosuppresion`)), `CMV +` = as.numeric(`CMV +`),`HHV6 +` = as.numeric(`HHV6 +`), Systemic_Steroids = as.numeric(Systemic_Steroids))%>%
mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1), Rep_div = "allo", SampleID = timepoint)
severe_cumulative <- patient_data_all_clinic[grepl('Severe',patient_data_all_clinic$GVHD_GRADE_1),]
mild_cumulative <- patient_data_all_clinic[grepl('Mild',patient_data_all_clinic$GVHD_GRADE_1),]
no_cumulative <- patient_data_all_clinic[grepl('No',patient_data_all_clinic$GVHD_GRADE_1),]
mild_cumulative <- mild_cumulative

no_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", no_cumulative$`Off Immunosuppresion`))
mild_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", mild_cumulative$`Off Immunosuppresion`))
severe_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", severe_cumulative$`Off Immunosuppresion`))

labels <- c("GHVD Onset", "Off Immunosuppresion", "EBV Viremia", "CMV Viremia", "HHV6 Viremia", "Systemic Steroid")
##SEVERE

severe_cumulative_fig <- div_fig_pat(severe_cumulative,  'n_clone', c('allo'), 'PatientID', ylim = c(0, 2000))
severe_cumulative_fig_early <- severe_cumulative_fig[[1]]+
  labs(title = 'Severe GVHD', y = "# of Clones")+
  theme(text = element_text(size = 20), plot.margin=unit(c(0,0,0,5.1), 'cm'))+
  scale_y_continuous(trans = "log1p", labels = scales::number_format(accuracy = 0.01))
severe_cumulative_fig_later <- severe_cumulative_fig[[2]]+
  theme(text = element_text(size = 20))+
  scale_y_continuous(trans = "log1p", labels = scales::number_format(accuracy = 0.01))
severe_meta_data_plot <- plot_metatime_cont(severe_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = severe_cumulative_fig[[3]])

severe_meta_data_plot[[1]] <- severe_meta_data_plot[[1]]+
  theme(text = element_text(size = 20))+
  coord_cartesian(xlim = c(0,360))+
  scale_x_continuous(expand = expansion(mult = c(0,0)))

severe_meta_data_plot[[2]] <- severe_meta_data_plot[[2]]+
  theme(text = element_text(size = 20))+
  coord_cartesian(xlim = c(360,severe_cumulative_fig[[3]]))+
  scale_x_continuous(expand = expansion(mult = c(0,0)))

  severe_comb <- plot_grid(plotlist = list(severe_cumulative_fig_early, severe_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))
  

severe_meta_comb <- plot_grid(plotlist = list(severe_meta_data_plot[[1]], severe_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

Severe_meta_fig <- plot_grid(severe_comb, severe_meta_comb, nrow = 2, greedy = TRUE, rel_heights = c(1,.3))


##Mild

mild_cumulative_fig <- div_fig_pat(mild_cumulative,  'n_clone', c('allo'), 'PatientID', ylim = c(0, 2000))
mild_cumulative_fig_early <- mild_cumulative_fig[[1]]+
  labs(title = 'Mild GVHD', y = "# of Clones")+
  theme(text = element_text(size = 20), plot.margin=unit(c(0,0,0,5.1), 'cm'))+
  scale_y_continuous(trans = "log1p", labels = scales::number_format(accuracy = 0.01))
mild_cumulative_fig_later <- mild_cumulative_fig[[2]]+
  theme(text = element_text(size = 20))+
  scale_y_continuous(trans = "log1p", labels = scales::number_format(accuracy = 0.01))
mild_meta_data_plot <- plot_metatime_cont(mild_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = mild_cumulative_fig[[3]])

mild_meta_data_plot[[1]] <- mild_meta_data_plot[[1]]+
  theme(text = element_text(size = 20))

mild_meta_data_plot[[2]] <- mild_meta_data_plot[[2]]+
  theme(text = element_text(size = 20))

  mild_comb <- plot_grid(plotlist = list(mild_cumulative_fig_early, mild_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))
  

mild_meta_comb <- plot_grid(plotlist = list(mild_meta_data_plot[[1]], mild_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

Mild_meta_fig <- plot_grid(mild_comb, mild_meta_comb, nrow = 2, greedy = TRUE, rel_heights = c(1,.3))


##NO

no_cumulative_fig <- div_fig_pat(no_cumulative,  'n_clone', c('allo'), 'PatientID', ylim = c(0, 2000))
no_cumulative_fig_early <- no_cumulative_fig[[1]]+
  labs(title = 'No GVHD', y = "# of Clones")+
  theme(text = element_text(size = 20), plot.margin=unit(c(0,0,0,5.1), 'cm'))+
  scale_y_continuous(trans = "log1p", labels = scales::number_format(accuracy = 0.01))
no_cumulative_fig_later <- no_cumulative_fig[[2]]+
  theme(text = element_text(size = 20))+
  scale_y_continuous(trans = "log1p", labels = scales::number_format(accuracy = 0.01))
no_meta_data_plot <- plot_metatime_cont(no_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = no_cumulative_fig[[3]])

no_meta_data_plot[[1]] <- no_meta_data_plot[[1]]+
  theme(text = element_text(size = 20))

no_meta_data_plot[[2]] <- no_meta_data_plot[[2]]+
  theme(text = element_text(size = 20))




no_comb <- plot_grid(plotlist = list(no_cumulative_fig_early, no_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))
  

no_meta_comb <- plot_grid(plotlist = list(no_meta_data_plot[[1]], no_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

No_meta_fig <- plot_grid(no_comb, no_meta_comb, nrow = 2, greedy = TRUE, rel_heights = c(1,.3))


plot(Severe_meta_fig)
plot(Mild_meta_fig)
plot(No_meta_fig)

```

```{r, fig.height=10,fig.width = 18, warning = FALSE}
clinical_data <- read_excel("/Users/lingtingshi/Documents/GVHD_project/From_Michael/TCR_Project.xlsx", sheet = 'Patient Metadata cont.')

patient_data_all_clinic <- dplyr::inner_join(patient_data_all, clinical_data, by = "PatientID")%>%
  dplyr::inner_join(meta_data, by = "PatientID")%>%
  filter(strtoi(SampleID) > 3 | grepl("MLR|unstim", SampleID))%>%
  mutate(Onset = as.numeric(Onset), `EBV +` = as.numeric(`EBV +`),`Off Immunosuppresion` = as.numeric(`Off Immunosuppresion`), `CMV +` = as.numeric(`CMV +`),`HHV6 +` = as.numeric(`HHV6 +`), Systemic_Steroids = as.numeric(Systemic_Steroids) )
ptcy_cumulative <- patient_data_all_clinic[patient_data_all_clinic$PTCy == "PTCy",]
no_ptcy_cumulative <- patient_data_all_clinic[patient_data_all_clinic$PTCy == "No PTCy",]

no_ptcy_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", no_ptcy_cumulative$`Off Immunosuppresion`))
ptcy_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", ptcy_cumulative$`Off Immunosuppresion`))

labels <- c("GHVD Onset", "Off Immunosuppresion", "EB Viremia", "CM Viremia", "HHV6", "Systemic Steroid")
##ptcy

ptcy_cumulative_fig <- div_fig_pat(ptcy_cumulative, 'shannons.diversity', c('all'), 'PatientID', ylim = c(1,10))
ptcy_cumulative_fig_early <- ptcy_cumulative_fig[[1]]+
  labs(title = 'PTCy', y = "Shannon's Diveristy")+
  theme(text = element_text(size = 28), plot.margin=unit(c(0,0,0,7.3), 'cm'))
ptcy_cumulative_fig_later <- ptcy_cumulative_fig[[2]]+
  theme(text = element_text(size = 28))
ptcy_meta_data_plot <- plot_metatime_cont(ptcy_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = ptcy_cumulative_fig[[3]])

ptcy_meta_data_plot[[1]] <- ptcy_meta_data_plot[[1]]+
  theme(text = element_text(size = 28))

ptcy_meta_data_plot[[2]] <- ptcy_meta_data_plot[[2]]+
  theme(text = element_text(size = 28))

  ptcy_comb <- plot_grid(plotlist = list(ptcy_cumulative_fig_early, ptcy_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

ptcy_meta_comb <- plot_grid(plotlist = list(ptcy_meta_data_plot[[1]], ptcy_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

ptcy_meta_fig <- plot_grid(ptcy_comb, ptcy_meta_comb, nrow = 2, align = "hv", greedy = TRUE, rel_heights = c(1,.3))

##no_ptcy

no_ptcy_cumulative_fig <- div_fig_pat(no_ptcy_cumulative, 'shannons.diversity', c('all'), 'PatientID', ylim = c(1,10))
no_ptcy_cumulative_fig_early <- no_ptcy_cumulative_fig[[1]]+
  labs(title = 'No PTCy', y = "Shannon's Diveristy")+
  theme(text = element_text(size = 28), plot.margin=unit(c(0,0,0,7.3), 'cm'))
no_ptcy_cumulative_fig_later <- no_ptcy_cumulative_fig[[2]]+
  theme(text = element_text(size = 28))
no_ptcy_meta_data_plot <- plot_metatime_cont(no_ptcy_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = no_ptcy_cumulative_fig[[3]])

no_ptcy_meta_data_plot[[1]] <- no_ptcy_meta_data_plot[[1]]+
  theme(text = element_text(size = 28))

no_ptcy_meta_data_plot[[2]] <- no_ptcy_meta_data_plot[[2]]+
  theme(text = element_text(size = 28))

  no_ptcy_comb <- plot_grid(plotlist = list(no_ptcy_cumulative_fig_early, no_ptcy_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

 no_ptcy_meta_comb <- plot_grid(plotlist = list(no_ptcy_meta_data_plot[[1]], no_ptcy_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

no_ptcy_meta_fig <- plot_grid(no_ptcy_comb, no_ptcy_meta_comb, nrow = 2, align = "hv", greedy = TRUE, rel_heights = c(1,.3))


plot(ptcy_meta_fig)
plot(no_ptcy_meta_fig)

  
# ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure2ex/ptcy_cumulative.png'), ptcy_meta_fig, width = 18, height = 10, dpi = 600)
# ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure2ex/no_ptcy_cumuluative.png'), no_ptcy_meta_fig, width = 18, height = 10, dpi = 600)

# ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure2ex/ptcy_cumulative.pdf'), ptcy_meta_fig, width = 18, height = 10, dpi = 600)
# ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure2ex/no_ptcy_cumuluative.pdf'), no_ptcy_meta_fig, width = 18, height = 10, dpi = 600)

```

```{r, fig.height=10,fig.width = 18, warning = FALSE}
clinical_data <- read_excel("/Users/lingtingshi/Documents/GVHD_project/From_Michael/TCR_Project.xlsx", sheet = 'Patient Metadata cont.')

patient_data_all_clinic <- dplyr::inner_join(patient_data_all, clinical_data, by = "PatientID")%>%
  dplyr::inner_join(meta_data, by = "PatientID")%>%
  filter(strtoi(SampleID) > 3 | !grepl("MLR|unstim", SampleID))%>%
  mutate(Onset = as.numeric(Onset), `EBV +` = as.numeric(`EBV +`),`Off Immunosuppresion` = as.numeric(gsub("[^0-9]", "", `Off Immunosuppresion`)), `CMV +` = as.numeric(`CMV +`),`HHV6 +` = as.numeric(`HHV6 +`), Systemic_Steroids = as.numeric(Systemic_Steroids) )
ptcy_cumulative <- patient_data_all_clinic[patient_data_all_clinic$PTCy == "PTCy",]
no_ptcy_cumulative <- patient_data_all_clinic[patient_data_all_clinic$PTCy == "No PTCy",]

no_ptcy_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", no_ptcy_cumulative$`Off Immunosuppresion`))
ptcy_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", ptcy_cumulative$`Off Immunosuppresion`))

labels <- c("GHVD Onset", "Off Immunosuppresion", "EBV Viremia", "CMV Viremia", "HHV6 Viremia", "Systemic Steroid")
##ptcy

ptcy_cumulative_fig <- div_fig_pat(ptcy_cumulative,  'cum_freq', c('allo'), 'PatientID', ylim = c(0, 0.05))
ptcy_cumulative_fig_early <- ptcy_cumulative_fig[[1]]+
  labs(y = "Alloreactive Cumulative Frequency", title = 'PTCy')+
  theme(text = element_text(size = 20), plot.margin=unit(c(0,0,0,5.2), 'cm'))
ptcy_cumulative_fig_later <- ptcy_cumulative_fig[[2]]+
  theme(text = element_text(size = 20))
ptcy_meta_data_plot <- plot_metatime_cont(ptcy_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = ptcy_cumulative_fig[[3]])

ptcy_meta_data_plot[[1]] <- ptcy_meta_data_plot[[1]]+
  theme(text = element_text(size = 20))+
  coord_cartesian(xlim = c(0,360))+
  scale_x_continuous(expand = expansion(mult = c(0,0)))

ptcy_meta_data_plot[[2]] <- ptcy_meta_data_plot[[2]]+
  theme(text = element_text(size = 20))+
  coord_cartesian(xlim = c(360,ptcy_cumulative_fig[[3]]))+
  scale_x_continuous(expand = expansion(mult = c(0,0)))

ptcy_comb <- plot_grid(plotlist = list(ptcy_cumulative_fig_early, ptcy_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))


ptcy_meta_comb <- plot_grid(plotlist = list(ptcy_meta_data_plot[[1]], ptcy_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

ptcy_meta_fig <- plot_grid(ptcy_comb, ptcy_meta_comb, nrow = 2, greedy = TRUE, rel_heights = c(1,.3))


##no_ptcy

no_ptcy_cumulative_fig <- div_fig_pat(no_ptcy_cumulative,  'cum_freq', c('allo'), 'PatientID', ylim = c(0, 0.05))
no_ptcy_cumulative_fig_early <- no_ptcy_cumulative_fig[[1]]+
  labs(y = "Alloreactive Cumulative Frequency", title = 'No PTCy')+
  theme(text = element_text(size = 20), plot.margin=unit(c(0,0,0,5.2), 'cm'))
no_ptcy_cumulative_fig_later <- no_ptcy_cumulative_fig[[2]]+
  theme(text = element_text(size = 20))
no_ptcy_meta_data_plot <- plot_metatime_cont(no_ptcy_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = no_ptcy_cumulative_fig[[3]])

no_ptcy_meta_data_plot[[1]] <- no_ptcy_meta_data_plot[[1]]+
  theme(text = element_text(size = 20))

no_ptcy_meta_data_plot[[2]] <- no_ptcy_meta_data_plot[[2]]+
  theme(text = element_text(size = 20))

no_ptcy_comb <- plot_grid(plotlist = list(no_ptcy_cumulative_fig_early, no_ptcy_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))
  

no_ptcy_meta_comb <- plot_grid(plotlist = list(no_ptcy_meta_data_plot[[1]], no_ptcy_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

no_ptcy_meta_fig <- plot_grid(no_ptcy_comb, no_ptcy_meta_comb, nrow = 2, greedy = TRUE, rel_heights = c(1,.3))




plot(ptcy_meta_fig)
plot(no_ptcy_meta_fig)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/ptcy_cumulative.png'), ptcy_meta_fig, width = 18, height = 10, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1ex/no_ptcy_cumuluative.png'), no_ptcy_meta_fig, width = 18, height = 10, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/ptcy_cumulative.pdf'), ptcy_meta_fig, width = 18, height = 10, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure1/no_ptcy_cumuluative.pdf'), no_ptcy_meta_fig, width = 18, height = 10, dpi = 600)

write.csv(patient_data_all_clinic, "/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/PatientDataClinic.csv")


```


```{r,fig.height=10,fig.width = 18, warning = FALSE}
clinical_data <- read_excel("/Users/lingtingshi/Documents/GVHD_project/From_Michael/TCR_Project.xlsx", sheet = 'Patient Metadata cont.')



data_CD48_all_expand_adj <- data_CD48_all%>%
  filter(templates !=0)%>%
  group_by(PatientID,timepoint)%>%
  summarise(n_clone = n())
  


patient_data_all_clinic <- dplyr::inner_join(data_CD48_all_expand_adj, clinical_data, by = "PatientID")%>%
  dplyr::inner_join(meta_data, by = "PatientID")%>%
  filter(strtoi(timepoint) > 3 | !grepl("MLR|unstim", timepoint))%>%
  mutate(Onset = as.numeric(Onset), `EBV +` = as.numeric(`EBV +`),`Off Immunosuppresion` = as.numeric(gsub("[^0-9]", "", `Off Immunosuppresion`)), `CMV +` = as.numeric(`CMV +`),`HHV6 +` = as.numeric(`HHV6 +`), Systemic_Steroids = as.numeric(Systemic_Steroids))%>%
mutate(GVHD_GRADE_1 = ifelse(grepl("ptcy", GVHD_GRADE_1), "ptcy", GVHD_GRADE_1), Rep_div = "allo", SampleID = timepoint)
ptcy_cumulative <- patient_data_all_clinic[patient_data_all_clinic$PTCy == "PTCy",]
no_ptcy_cumulative <- patient_data_all_clinic[patient_data_all_clinic$PTCy == "No PTCy",]


no_ptcy_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", no_ptcy_cumulative$`Off Immunosuppresion`))
ptcy_cumulative$`Off Immunosuppresion` <- as.numeric(gsub("[^0-9]", "", ptcy_cumulative$`Off Immunosuppresion`))

labels <- c("GHVD Onset", "Off Immunosuppresion", "EBV Viremia", "CMV Viremia", "HHV6 Viremia", "Systemic Steroid")
##ptcy

ptcy_cumulative_fig <- div_fig_pat(ptcy_cumulative,  'n_clone', c('allo'), 'PatientID', ylim = c(0, 2000))
ptcy_cumulative_fig_early <- ptcy_cumulative_fig[[1]]+
  labs(title = 'PTCy', y = "# of Clones")+
  theme(text = element_text(size = 20), plot.margin=unit(c(0,0,0,5.1), 'cm'))+
  scale_y_continuous(trans = "log1p", labels = scales::number_format(accuracy = 1), breaks = c(0,10,100,1000))
ptcy_cumulative_fig_later <- ptcy_cumulative_fig[[2]]+
  theme(text = element_text(size = 20))+
  scale_y_continuous(trans = "log1p", labels = scales::number_format(accuracy = 1), breaks = c(0,10,100,1000))
ptcy_meta_data_plot <- plot_metatime_cont(ptcy_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = ptcy_cumulative_fig[[3]])

ptcy_meta_data_plot[[1]] <- ptcy_meta_data_plot[[1]]+
  theme(text = element_text(size = 20))+
  coord_cartesian(xlim = c(0,360))+
  scale_x_continuous(expand = expansion(mult = c(0,0)))

ptcy_meta_data_plot[[2]] <- ptcy_meta_data_plot[[2]]+
  theme(text = element_text(size = 20))+
  coord_cartesian(xlim = c(360,ptcy_cumulative_fig[[3]]))+
  scale_x_continuous(expand = expansion(mult = c(0,0)))

  ptcy_comb <- plot_grid(plotlist = list(ptcy_cumulative_fig_early, ptcy_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))
  

ptcy_meta_comb <- plot_grid(plotlist = list(ptcy_meta_data_plot[[1]], ptcy_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

ptcy_meta_fig <- plot_grid(ptcy_comb, ptcy_meta_comb, nrow = 2, greedy = TRUE, rel_heights = c(1,.3))


##no_ptcy

no_ptcy_cumulative_fig <- div_fig_pat(no_ptcy_cumulative,  'n_clone', c('allo'), 'PatientID', ylim = c(0, 2000))
no_ptcy_cumulative_fig_early <- no_ptcy_cumulative_fig[[1]]+
  labs(title = 'No PTCy', y = "# of Clones")+
  theme(text = element_text(size = 20), plot.margin=unit(c(0,0,0,5.1), 'cm'))+
  scale_y_continuous(trans = "log1p", labels = scales::number_format(accuracy = 1), breaks = c(0,10,100,1000))
no_ptcy_cumulative_fig_later <- no_ptcy_cumulative_fig[[2]]+
  theme(text = element_text(size = 20))+
  scale_y_continuous(trans = "log1p", labels = scales::number_format(accuracy = 1), breaks = c(0,10,100,1000))
no_ptcy_meta_data_plot <- plot_metatime_cont(no_ptcy_cumulative, c("Onset", "Off Immunosuppresion", "EBV +", "CMV +", "HHV6 +", "Systemic_Steroids"), 'SampleID', label = labels, max_tp = no_ptcy_cumulative_fig[[3]])

no_ptcy_meta_data_plot[[1]] <- no_ptcy_meta_data_plot[[1]]+
  theme(text = element_text(size = 20))

no_ptcy_meta_data_plot[[2]] <- no_ptcy_meta_data_plot[[2]]+
  theme(text = element_text(size = 20))

  no_ptcy_comb <- plot_grid(plotlist = list(no_ptcy_cumulative_fig_early, no_ptcy_cumulative_fig_later), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))
  

no_ptcy_meta_comb <- plot_grid(plotlist = list(no_ptcy_meta_data_plot[[1]], no_ptcy_meta_data_plot[[2]]), ncol = 2, nrow = 1, hjust = 5, align = "h", rel_widths = c(1, .2))

no_ptcy_meta_fig <- plot_grid(no_ptcy_comb, no_ptcy_meta_comb, nrow = 2, greedy = TRUE, rel_heights = c(1,.3))


plot(ptcy_meta_fig)
plot(no_ptcy_meta_fig)

```

```{r}
clinical_data <- read_excel("/Users/lingtingshi/Documents/GVHD_project/From_Michael/TCR_Project.xlsx", sheet = 'Patient Metadata cont.')

patient_data_all_clinic <- dplyr::inner_join(patient_data_all, clinical_data, by = "PatientID")%>%
  dplyr::inner_join(meta_data, by = "PatientID")%>%
  filter(strtoi(SampleID) > 3 | grepl("MLR|unstim", SampleID))%>%
  mutate(Onset = as.numeric(Onset), `EBV +` = as.numeric(`EBV +`),`Off Immunosuppresion` = as.numeric(gsub("[^0-9]", "", `Off Immunosuppresion`)), `CMV +` = as.numeric(`CMV +`),`HHV6 +` = as.numeric(`HHV6 +`), Systemic_Steroids = as.numeric(Systemic_Steroids))%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("Severe",GVHD_GRADE_1), "Severe", GVHD_GRADE_1))%>%
  # mutate(PTCy = ifelse(grepl("PTCy", PTCy), "PTCy", "No PTCy"))%>%
  filter(!is.na(Onset))%>%
  group_by(PatientID, Onset)%>%
  mutate(tp_closest_onset = abs(strtoi(SampleID) - Onset))%>%
  slice_min(tp_closest_onset)%>%
  ungroup()%>%
  select(-tp_closest_onset)%>%
  filter(Rep_div == "allo")

data_CD48_all_expand_adj <- dplyr::full_join(data_CD48_all, clinical_data, by = 'PatientID')
data_CD48_all_expand_adj_tp <- data_CD48_all_expand_adj%>%
  select(timepoint, PatientID)%>%
  distinct()

data_CD48_all_expand_adj <- data_CD48_all_expand_adj%>%
  filter(templates !=0)%>%
  group_by(PatientID,timepoint,Onset, GVHD_GRADE_1, PTCy)%>%
  # mutate(PTCy = ifelse(grepl("PTCy", PTCy), "PTCy", "No PTCy"))%>%
  summarise(n_clone = n())%>%
  inner_join(data_CD48_all_expand_adj_tp, by = c("PatientID", "timepoint"))%>%
  replace_na(list(n_clone = 0))%>%
  filter(!is.na(Onset))%>%
  group_by(PatientID, Onset)%>%
  mutate(tp_closest_onset = abs(strtoi(timepoint) - Onset))%>%
  slice_min(tp_closest_onset)%>%
  ungroup()%>%
  select(-tp_closest_onset)%>%
  mutate(GVHD_GRADE_1 = ifelse(grepl("Severe", GVHD_GRADE_1), "Severe", GVHD_GRADE_1))

onset_plot_freq <- ggplot(patient_data_all_clinic, aes(x = PTCy, y = cum_freq, group = PTCy, fill = PTCy))+
  geom_boxplot()+
  geom_beeswarm()+
  stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test", method.args= list(exact = F), alternative = "two.sided", label = "p.format", hide.ns = TRUE)+
  scale_fill_manual(values = c( "#005A8F","#B85000")) +
  theme_pubr()

onset_plot_div <- ggplot(patient_data_all_clinic, aes(x = PTCy, y = shannons.diversity, fill = PTCy))+
  geom_boxplot()+
  geom_beeswarm()+
  stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test", method.args= list(exact = F), alternative = "two.sided", label = "p.format", hide.ns = TRUE)+
  scale_fill_manual(values = c( "#005A8F","#B85000")) +
  theme_pubr()

onset_plot_n <- ggplot(data_CD48_all_expand_adj, aes(x = PTCy, y = n_clone, fill = PTCy))+
  geom_boxplot()+
  geom_beeswarm()+
  stat_compare_means(comparisons = list(c("PTCy", "No PTCy")), method = "wilcox.test", method.args= list(exact = F), alternative = "two.sided", label = "p.format", hide.ns = TRUE)+
  scale_fill_manual(values = c( "#005A8F","#B85000")) +
  theme_pubr()
#Add to suplementary

plot(onset_plot_freq)
plot(onset_plot_div)
plot(onset_plot_n)
```




# ```{r}
# reads <- c(119366.00, 693276.00, 111624.00, 91063.00, 399078.00, 
#           44523.00, 139858.00, 69680.00, 153603.00, 104247.00, 
#           66289.00, 27041.00, 267988.00, 951864.00, 1664944.00, 
#           365043.00, 346627.00, 436310.00, 681248.00, 844270.00)
# 
# donor_types <- c("Haplo", "Matched Unrelated Donor", "Haplo", "Matched Unrelated Donor", 
#                 "Haplo", "Haplo", "Haplo", "Matched Unrelated Donor", 
#                 "Haplo", "Haplo", "Haplo", "Haplo", "Haplo", "Haplo", 
#                 "Matched Unrelated Donor", "Mismatched Unrelated Donor", 
#                 "Matched Sibling Donor", "Matched Sibling Donor", 
#                 "Haplo", "Matched Unrelated Donor")
# ```
# ```{r}
# data <- data.frame(
#   ID = custom_order,
#   DonorType = donor_types,
#   Value = reads,
#   allo_clone_counts = ordered_counts$num_unique_tags,
#   Normalized_to_reads = ordered_counts$num_unique_tags/reads,
#   templates = ordered_sums$sum_template,
#   Normalized_to_templates = ordered_counts$num_unique_tags/ordered_sums$sum_template
# )
# ```
# ```{r}
# # Add binary Haplo column (TRUE if Haplo, FALSE otherwise)
# data$IsHaplo <- data$DonorType == "Haplo"
# 
# # Add a factor version for easier analysis and plotting
# data$HaploCategory <- ifelse(data$IsHaplo, "Haplo", "Not Haplo")
# data$HaploCategory <- factor(data$HaploCategory)
# ```
# 
# # ```{r}
# # data
# # ```
# 
# ```{r}
# library(ggplot2)
# 
# # Create the box plot with improved aesthetics
# p <- ggplot(data, aes(x = HaploCategory, y = Normalized_to_reads, fill = HaploCategory)) +
#   # Add boxplot with slightly transparent fill
#   geom_boxplot(alpha = 0.7, outlier.shape = NA) +
#   # Add data points with jitter to avoid overlapping
#   geom_jitter(width = 0.2, size = 2.5, alpha = 0.7, aes(color = DonorType), show.legend = TRUE) +
#   # Custom fill colors for Haplo vs Not Haplo
#   scale_fill_manual(values = c("Haplo" = "#C0C0C0", "Not Haplo" = "#80DEEA")) +
#   # Create a custom color scale for the donor types
#   scale_color_manual(values = c(
#     "Haplo" = "#666666",
#     "Matched Unrelated Donor" = "#0288D1",
#     "Matched Sibling Donor" = "#FFC107",
#     "Mismatched Unrelated Donor" = "#F44336"
#   )) +
#   # Improve labels
#   labs(
#     title = "Alloclone Counts by Donor Type",
#     subtitle = "Comparison between Haplo and Non-Haplo donors",
#     y = "Normalized to reads Alloclone Counts",
#     x = NULL,
#     color = "Donor Type"
#   ) +
#   # Use a clean theme
#   theme_minimal() +
#   # Customize theme elements
#   theme(
#     plot.title = element_text(hjust = 0.5, face = "bold", size = 14),
#     plot.subtitle = element_text(hjust = 0.5, size = 11, color = "gray40"),
#     axis.title.y = element_text(face = "bold", size = 12),
#     axis.text = element_text(size = 10),
#     legend.position = "right",
#     legend.title = element_text(face = "bold"),
#     panel.grid.major.x = element_blank(),
#     panel.border = element_rect(fill = NA, color = "gray80", linewidth = 0.5)
#   )
# 
# # For statistical comparison, perform a t-test
# t_test_result <- t.test(Normalized_to_reads ~ HaploCategory, data = data)
# print(t_test_result)
# 
# # Extract p-value and format it with improved logic
# p_val <- t_test_result$p.value
# p_val_formatted <- if(p_val < 0.001) {
#   "p < 0.001 ***"
# } else if(p_val < 0.01) {
#   paste0("p = ", sprintf("%.3f", p_val), " **")
# } else if(p_val < 0.05) {
#   paste0("p = ", sprintf("%.3f", p_val), " *")
# } else {
#   paste0("p = ", sprintf("%.3f", p_val), " (ns)")
# }
# 
# # Calculate a good position for the significance bar based on data
# y_max <- max(data$Normalized_to_reads, na.rm = TRUE)
# y_buffer <- y_max * 0.15
# 
# # Add significance bracket and p-value with better positioning
# p <- p + 
#   # Add bracket connecting the two groups
#   geom_segment(aes(x = 1, y = y_max + (y_buffer*0.5), 
#                   xend = 2, yend = y_max + (y_buffer*0.5)), 
#               linewidth = 0.7) +
#   # Add short vertical segments at the ends of the bracket
#   geom_segment(aes(x = 1, y = y_max + (y_buffer*0.4), 
#                   xend = 1, yend = y_max + (y_buffer*0.6)), 
#               linewidth = 0.7) +
#   geom_segment(aes(x = 2, y = y_max + (y_buffer*0.4), 
#                   xend = 2, yend = y_max + (y_buffer*0.6)), 
#               linewidth = 0.7) +
#   # Add p-value text above bracket
#   geom_text(aes(x = 1.5, y = y_max + (y_buffer*0.8)), 
#             label = p_val_formatted, size = 4) +
#   # Extend y-axis to fit the significance bracket
#   ylim(NA, y_max + y_buffer*1.2)
# 
# # Also calculate and display mean values for each group
# mean_haplo <- mean(data$Normalized_to_reads[data$HaploCategory == "Haplo"], na.rm = TRUE)
# mean_not_haplo <- mean(data$Normalized_to_reads[data$HaploCategory == "Not Haplo"], na.rm = TRUE)
# cat("Mean for Haplo group:", sprintf("%.2f", mean_haplo), "\n")
# cat("Mean for Not Haplo group:", sprintf("%.2f", mean_not_haplo), "\n")
# cat("Fold difference:", sprintf("%.2f", mean_not_haplo/mean_haplo), "\n")
# 
# # Display the plot
# print(p)
# 
# # Optional: Save the plot as a high-resolution image
# # ggsave("donor_comparison_plot.png", p, width = 8, height = 6, dpi = 300)
# ```
# ```{r}
# library(ggplot2)
# 
# # Create the box plot with improved aesthetics
# p <- ggplot(data, aes(x = HaploCategory, y = Normalized_to_templates, fill = HaploCategory)) +
#   # Add boxplot with slightly transparent fill
#   geom_boxplot(alpha = 0.7, outlier.shape = NA) +
#   # Add data points with jitter to avoid overlapping
#   geom_jitter(width = 0.2, size = 2.5, alpha = 0.7, aes(color = DonorType), show.legend = TRUE) +
#   # Custom fill colors for Haplo vs Not Haplo
#   scale_fill_manual(values = c("Haplo" = "#C0C0C0", "Not Haplo" = "#80DEEA")) +
#   # Create a custom color scale for the donor types
#   scale_color_manual(values = c(
#     "Haplo" = "#666666",
#     "Matched Unrelated Donor" = "#0288D1",
#     "Matched Sibling Donor" = "#FFC107",
#     "Mismatched Unrelated Donor" = "#F44336"
#   )) +
#   # Improve labels
#   labs(
#     title = "Alloclone Counts by Donor Type",
#     subtitle = "Comparison between Haplo and Non-Haplo donors",
#     y = "Normalized to template Alloclone Counts",
#     x = NULL,
#     color = "Donor Type"
#   ) +
#   # Use a clean theme
#   theme_minimal() +
#   # Customize theme elements
#   theme(
#     plot.title = element_text(hjust = 0.5, face = "bold", size = 14),
#     plot.subtitle = element_text(hjust = 0.5, size = 11, color = "gray40"),
#     axis.title.y = element_text(face = "bold", size = 12),
#     axis.text = element_text(size = 10),
#     legend.position = "right",
#     legend.title = element_text(face = "bold"),
#     panel.grid.major.x = element_blank(),
#     panel.border = element_rect(fill = NA, color = "gray80", linewidth = 0.5)
#   )
# 
# # For statistical comparison, perform a t-test
# t_test_result <- t.test(Normalized_to_templates ~ HaploCategory, data = data)
# print(t_test_result)
# 
# # Extract p-value and format it with improved logic
# p_val <- t_test_result$p.value
# p_val_formatted <- if(p_val < 0.001) {
#   "p < 0.001 ***"
# } else if(p_val < 0.01) {
#   paste0("p = ", sprintf("%.3f", p_val), " **")
# } else if(p_val < 0.05) {
#   paste0("p = ", sprintf("%.3f", p_val), " *")
# } else {
#   paste0("p = ", sprintf("%.3f", p_val), " (ns)")
# }
# 
# # Calculate a good position for the significance bar based on data
# y_max <- max(data$Normalized_to_templates, na.rm = TRUE)
# y_buffer <- y_max * 0.15
# 
# # Add significance bracket and p-value with better positioning
# p <- p + 
#   # Add bracket connecting the two groups
#   geom_segment(aes(x = 1, y = y_max + (y_buffer*0.5), 
#                   xend = 2, yend = y_max + (y_buffer*0.5)), 
#               linewidth = 0.7) +
#   # Add short vertical segments at the ends of the bracket
#   geom_segment(aes(x = 1, y = y_max + (y_buffer*0.4), 
#                   xend = 1, yend = y_max + (y_buffer*0.6)), 
#               linewidth = 0.7) +
#   geom_segment(aes(x = 2, y = y_max + (y_buffer*0.4), 
#                   xend = 2, yend = y_max + (y_buffer*0.6)), 
#               linewidth = 0.7) +
#   # Add p-value text above bracket
#   geom_text(aes(x = 1.5, y = y_max + (y_buffer*0.8)), 
#             label = p_val_formatted, size = 4) +
#   # Extend y-axis to fit the significance bracket
#   ylim(NA, y_max + y_buffer*1.2)
# 
# # Also calculate and display mean values for each group
# mean_haplo <- mean(data$Normalized_to_templates[data$HaploCategory == "Haplo"], na.rm = TRUE)
# mean_not_haplo <- mean(data$Normalized_to_templates[data$HaploCategory == "Not Haplo"], na.rm = TRUE)
# cat("Mean for Haplo group:", sprintf("%.2f", mean_haplo), "\n")
# cat("Mean for Not Haplo group:", sprintf("%.2f", mean_not_haplo), "\n")
# cat("Fold difference:", sprintf("%.2f", mean_not_haplo/mean_haplo), "\n")
# 
# # Display the plot
# print(p)
# 
# # Optional: Save the plot as a high-resolution image
# # ggsave("donor_comparison_plot.png", p, width = 8, height = 6, dpi = 300)
# ```


```{r}
# Calculate ratio of allo to nonallo frequency for each patient with timepoint 2, 3, or unstimulated
calculate_ratios <- function(data, patient_clinical_data, timepoints_to_include = c("2", "3")) {
  # Get unique patients with the specified timepoints
  patients <- unique(data$PatientID[data$timepoint %in% timepoints_to_include])
  timepoints_to_include = c("2", "3", "unstimulated")
  
  # Initialize results dataframe
  results <- data.frame(
    PatientID = character(),
    timepoint = character(),  # Changed to character to accommodate "unstimulated"
    allo_frequency = numeric(),
    nonallo_frequency = numeric(),
    ratio = numeric(),
    DonorType = character(),
    Disease = character(),
    PTCy = character(),
    GVHD_GRADE_1 = character(),
    stringsAsFactors = FALSE
  )
  
  # Calculate ratio for each patient and timepoint
  for (patient in patients) {
    for (tp in timepoints_to_include) {
      # Check if this patient has data for this timepoint
      if (nrow(data[data$PatientID == patient & data$timepoint == tp, ]) > 0) {
        # Calculate allo frequency
        allo_data <- data %>%
          filter(type == "allo" & timepoint == tp & PatientID == patient)
        allo_f <- sum(allo_data$frequency)
        
        # Calculate nonallo frequency
        nonallo_data <- data %>%
          filter(type == "nonallo" & timepoint == tp & PatientID == patient)
        nonallo_f <- sum(nonallo_data$frequency)
        
        # Calculate ratio (handle division by zero)
        if (nonallo_f > 0) {
          ratio <- allo_f / nonallo_f
        } else {
          ratio <- NA  # or could use Inf if you prefer
        }
        
        # Get clinical data for this patient
        patient_data <- patient_clinical_data[patient_clinical_data$PatientID == patient, ]
        
        # If patient exists in clinical data, add all information
        if (nrow(patient_data) > 0) {
          results <- rbind(results, data.frame(
            PatientID = patient,
            timepoint = tp,
            allo_frequency = allo_f,
            nonallo_frequency = nonallo_f,
            ratio = ratio,
            DonorType = patient_data$DonorType,
            Disease = patient_data$Disease,
            PTCy = patient_data$PTCy,
            GVHD_GRADE_1 = patient_data$GVHD_GRADE_1,
            stringsAsFactors = FALSE
          ))
        } else {
          # If patient doesn't exist in clinical data, add results without clinical info
          results <- rbind(results, data.frame(
            PatientID = patient,
            timepoint = tp,
            allo_frequency = allo_f,
            nonallo_frequency = nonallo_f,
            ratio = ratio,
            DonorType = NA,
            Disease = NA,
            PTCy = NA,
            GVHD_GRADE_1 = NA,
            stringsAsFactors = FALSE
          ))
        }
      }
    }
  }
  
  return(results)
}

# Use the function to get ratios with clinical information
ratio_results <- calculate_ratios(data_all_squished,meta_data_clin)

# Display results
print(ratio_results)

```
```{r}
ratio_results$timepoint_modified <- ifelse(ratio_results$timepoint %in% c("2", "3"), "2/3", as.character(ratio_results$timepoint))

ratio_results$grade_modified <- ifelse(ratio_results$GVHD_GRADE_1 %in% c("Unclassified/Severe"), "Severe", as.character(ratio_results$GVHD_GRADE_1))
ratio_results$timepoint_modified <- factor(ratio_results$timepoint_modified, 
                                          levels = c("unstimulated", "2/3"))

early_plot_avg_freq <- ggplot(ratio_results[ratio_results$PatientID != "R090", ], aes(x = timepoint_modified, y = ratio, shape = PatientID))+
    geom_line(aes(group = PatientID, color = grade_modified), linewidth = 1.5)+
    geom_point(aes(color = GVHD_GRADE_1), size = 6)+
    scale_y_continuous(labels = scales::number_format(accuracy = 0.0001))+
    stat_compare_means(comparisons = list(c("unstimulated","2/3")), method = "wilcox.test", paired = TRUE, label = "p.signif", hide.ns = FALSE, group.by = "PatientID", method.args = list(alternative ="less"), size = 6, label.y = quantile(ratio_results[ratio_results$PatientID != "R090", ]$ratio, probs = c(0.90)))+
    scale_color_manual(values = c("#005A8F", "#B85000", "grey"))+
    # expand_limits(y = max(df_timepoint$cum_freq)+.05)+
    labs(x = "Timepoint", y = "Accumulative Frequency", title = "allo/nonallo", color = "GVHD Grade")+
    # coord_cartesian(ylim = c(0,0.4))+
    theme_minimal()+
    theme_pubr()+
    theme(legend.position = "none")
plot(early_plot_avg_freq)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/allovsnonallo.png'), early_plot_avg_freq,  height = 6, width=5, dpi = 600)

ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/allovsnonallo.pdf'), early_plot_avg_freq,  height = 6, width=5, dpi = 600)
ggsave(paste0('/Users/lingtingshi/Documents/GVHD_project/From_Michael/MLR_Pipeline/Main_fig/Figure3/allovsnonallo.svg'), early_plot_avg_freq,  height = 6, width=5, dpi = 600)

```
```{r}

ratio_results_filtered <- ratio_results[ratio_results$timepoint %in% c('2','3'),]
ratio_results_filtered_un <- ratio_results[ratio_results$timepoint %in% c('unstimulated'),]
```
```{r}
ratio_results_filtered

```

```{r}
ratio_results_filtered_un
```
```{r}
# Assuming ratio_results_filtered contains the stimulated samples (timepoint 2 or 3)
# and ratio_results_filtered_un contains the matching unstimulated samples

# Calculate fold change
fold_change <- ratio_results_filtered$ratio / ratio_results_filtered_un$ratio

# Calculate log2 fold change
log2_fold_change <- log2(fold_change)

# Or in one step:
log2_fold_change <- log2(ratio_results_filtered$ratio / ratio_results_filtered_un$ratio)
median(log2_fold_change)
```
```{r}
log2_fold_change
```

